\documentclass[11pt,a4paper,oneside]{book}	%fleqn,
\usepackage{etex}

\usepackage{longtable,rotating, amsmath, epic, graphics, bm}

\usepackage{multirow, setspace, fullpage, graphicx, amssymb}

\usepackage{epsfig, moreverb}

\usepackage{natbib, color}

\usepackage{tocbibind}

\usepackage{easymat}

\usepackage{tabulary}

\usepackage[pdftex,
            pdfauthor={Your Name},
            pdftitle={Thesis},
            pdfsubject={Dissertation},
            pdfkeywords={microarray gene expression statistical models regression},
            pdfproducer={Latex with hyperref, or other system},
            pdfcreator={pdflatex, or other tool}]{hyperref}

% pandoc
\usepackage[T1]{fontenc}
\usepackage{lmodern}
\usepackage{amssymb,amsmath}
\usepackage{ifxetex,ifluatex}
\usepackage{booktabs}
\usepackage{fixltx2e} % provides \textsubscript
% use upquote if available, for straight quotes in verbatim environments
\IfFileExists{upquote.sty}{\usepackage{upquote}}{}
\ifnum 0\ifxetex 1\fi\ifluatex 1\fi=0 % if pdftex
  \usepackage[utf8]{inputenc}
\else % if luatex or xelatex
  \ifxetex
    \usepackage{mathspec}
    \usepackage{xltxtra,xunicode}
  \else
    \usepackage{fontspec}
  \fi
  \defaultfontfeatures{Mapping=tex-text,Scale=MatchLowercase}
  \newcommand{\euro}{â‚¬}
\fi
% use microtype if available
\IfFileExists{microtype.sty}{\usepackage{microtype}}{}
\usepackage{color}
\usepackage{fancyvrb}
\newcommand{\VerbBar}{|}
\newcommand{\VERB}{\Verb[commandchars=\\\{\}]}
\DefineVerbatimEnvironment{Highlighting}{Verbatim}{commandchars=\\\{\}}
% Add ',fontsize=\small' for more characters per line
\newenvironment{Shaded}{}{}
\newcommand{\KeywordTok}[1]{\textcolor[rgb]{0.00,0.44,0.13}{\textbf{{#1}}}}
\newcommand{\DataTypeTok}[1]{\textcolor[rgb]{0.56,0.13,0.00}{{#1}}}
\newcommand{\DecValTok}[1]{\textcolor[rgb]{0.25,0.63,0.44}{{#1}}}
\newcommand{\BaseNTok}[1]{\textcolor[rgb]{0.25,0.63,0.44}{{#1}}}
\newcommand{\FloatTok}[1]{\textcolor[rgb]{0.25,0.63,0.44}{{#1}}}
\newcommand{\CharTok}[1]{\textcolor[rgb]{0.25,0.44,0.63}{{#1}}}
\newcommand{\StringTok}[1]{\textcolor[rgb]{0.25,0.44,0.63}{{#1}}}
\newcommand{\CommentTok}[1]{\textcolor[rgb]{0.38,0.63,0.69}{\textit{{#1}}}}
\newcommand{\OtherTok}[1]{\textcolor[rgb]{0.00,0.44,0.13}{{#1}}}
\newcommand{\AlertTok}[1]{\textcolor[rgb]{1.00,0.00,0.00}{\textbf{{#1}}}}
\newcommand{\FunctionTok}[1]{\textcolor[rgb]{0.02,0.16,0.49}{{#1}}}
\newcommand{\RegionMarkerTok}[1]{{#1}}
\newcommand{\ErrorTok}[1]{\textcolor[rgb]{1.00,0.00,0.00}{\textbf{{#1}}}}
\newcommand{\NormalTok}[1]{{#1}}

\settocname{Table of Contents}


\bibliographystyle{natbib} % natbib

\bibpunct{[}{]},a{},

% TODO: Remove all unused macros

\renewcommand\baselinestretch{1.5} %line spacing is 1.5

\newcommand{\var}{{\rm var}}

\newcommand{\cov}{{\rm cov}}

\def\eqd{\,{\buildrel d \over =}\,}

%\newcommand{\R}{{\textbf{\textsf{R}}}}}


\setlength{\oddsidemargin}{5mm}

\setlength{\evensidemargin}{5mm}

\setlength{\textwidth}{150mm}

\setlength{\marginparsep}{0pt}

\setlength{\marginparwidth}{0pt}


\setlength{\topmargin}{0pt}

\setlength{\headsep}{0pt}

\setlength{\textheight}{225mm}

\setlength{\footskip}{1.5cm}

\setlength{\parskip}{4pt}


%\definecolor{sapgreen}{rgb}{.1,.4,.1}

\definecolor{brick}{rgb}{.5,.0,.0}

\definecolor{darkblue}{rgb}{.0,.0,.5}

%\definecolor{black}{rgb}{0,1,0}

\definecolor{gray}{gray}{.5}


\pagestyle{plain}

\setcounter{tocdepth}{1}

\begin{document}


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%% Starting

\title{\huge \bf The statistical analysis of high-throughput assays
for studying DNA methylation.\\}

\author{{\bf \Large Peter Francis Hickey}\\\\
Submitted in total fulfilment of the requirements\\
of the degree of Doctor of Philosophy\\\\
November 2014\\\\
Department of Mathematics and Statistics\\
The University of Melbourne\\\\}

\date{The Walter and Eliza Hall Institute of Medical Research}

\maketitle

\frontmatter

\chapter{Abstract}

% should be between 300 - 500 words

\textbf{TODO}


\chapter{Declaration}

{\bf This is to certify that}

\begin{itemize}

\item [(i)] the thesis comprises only my original work towards

the PhD except where indicated in the Preface,

\item [(ii)] due acknowledgement has been made in

the text to all other material used,

\item [(iii)] the thesis is less than 100,000 words in length,

exclusive of tables, maps, bibliographies and appendices.

\end{itemize}



\vspace{2cm}


{\large\bf Signed,\\\\\\\\}



\chapter{Preface}

* Some data in my thesis are mapped to hg18 and some to hg19.
* Include summary of notation?


The data sets used in this thesis were provided by the investigators listed below.\\\\

\chapter{Acknowledgments}

I am grateful to the following people for their support during my PhD...


\tableofcontents

\listoffigures

\listoftables


\chapter{List of Abbreviations}

\begin{tabular}{ll}

bp & base pair \\
kb & kilobase = $1000$ bp \\
Mb & megabase = $1,000,000$ bp \\
Gb & gigabase = $1,000,000,000$ bp \\
WGBS & whole genome bisulfite sequencing \\
RRBS & reduced representation bisulfite sequencing \\
FDR & false discovery rate \\
DMR & differentially methylated region \\
DMC & differentially methylated cytosine \\
HMM & hidden Markov model \\
CNV & copy number variant \\
SE & single-end sequencing \\
PE & paired-end sequencing \\
indel & insertion or deletion \\

\end{tabular}

\chapter{List of Symbols}

\begin{tabular}{ll}

i.e. & that is\\

\end{tabular}

\mainmatter

% each chapter is in a separate .tex file

\include{introduction} % introduction.tex

\include{biology_background} % biology_background.tex

\include{wgbs_bioinformatics_analysis} % wgbs_bioinformatics_analysis.tex

\include{datasets} % datasets.tex

\include{wgbs_statistical_framework} % wgbs_statistical_framework.tex

\include{wgbs_downstream_analyses} % wgbs_downstream_analyses.tex

\include{comethylation_review} % comethylation_review.tex

\include{comethylation} % comethylation.tex

\include{methsim} % methim.tex

\backmatter

\chapter{Concluding Comments}\label{chapter:Conclusions}


% 2 - 4 pages
\textbf{TODO}

\include{appendix} % appendix.tex

\bibliography{phd_thesis}

\end{document}
