---
title: "Appendix"
author: "Peter Hickey"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  pdf_document:
    fig_caption: yes
    includes:
      in_header: ../latex/header.tex
    keep_tex: no
  html_document:
    keep_md: yes
bibliography: ../latex/phd_thesis.bib
---

\chapter{Appendix}\label{chap:appendix}

\section{Chapter \ref{chap:co-methylation_review}}\label{app:co-methylation_review}

\subsection{The probability that two dependent Bernoulli random variables are identical}\label{app:bernoulli_equal}

\cite{Lindqvist:1978fm} wrote a brief note on Bernoulli trials with dependence.
Building on earlier work by \cite{Klotz:1973if}, Lindqvist parameterises the
Bernoulli process $X_{1}, X_{2}, \ldots$ on $\{0, 1\}$ by the parameters
$p = Pr(X_{i} = 1)$ and $c = cor(X_{i - 1}, X_{i})$ and shows that the
transition matrix is given by

\begin{equation*}
\label{eq:transition_matrix_id}
\Pi =
 \begin{pmatrix}
  (1 - p) + cp & p (1 - c) \\
  (1 - p) (1 - c) & p + c (1 - p)
 \end{pmatrix}
\end{equation*}

provided that $max(1 - \frac{1}{p}, 1 - \frac{1}{1 - p}) \leq c \leq 1$.

From this we can compute the joint distribution,

\begin{equation*}
Pr(X_{1} = x_{1}, \ldots, X_{n} = x_{n}) = Pr(X_{1} = x_{1}) Pr(X_{2} = x_{2} | X_{2} = x_{2}) \cdots Pr(X_{n} = x_{n} | X_{n - 1} = x_{n - 1})
\end{equation*}

In particular, in the case $n = 2$ we can compute the probability that two
dependent and identically distributed Bernoulli random variables are equal.

To extend the above result to the probability that two dependent and
__non-identically__ distributed Bernoulli random variables are equal, simply
requires that we derive the appropriate transition matrix. Switching notation
to that used in Chapter \ref{chap:co-methylation_review}, let
$Z_{h, i} \eqd Bernoulli(p_{h})$ and $Z_{h', i'} \eqd Bernoulli(p_{h'})$. The
transition matrix,
$\Pi = \bigl(\begin{smallmatrix}
Pr(Z_{h', i'} = z_{h', i'} | Z_{h, i} = z_{h, i})
\end{smallmatrix} \bigr)$, is given by

\begin{equation*}
\label{eq:transition_matrix_nid}
\Pi =
 \begin{pmatrix}
  (1 - p_{h'}) + c p_{h} & p_{h'} - c p_{h} \\
  \big [ (1 - p_{h'}) + c p_{h} \big ] (1 - p_{h}) & \big [ p_{h'} + c (1 - p_{h}) \big ] p_{h}
 \end{pmatrix}
\end{equation*}

We can then compute the desired probability

\begin{align*}
Pr(Z_{h, i} = Z_{h', i'}) &= Pr(Z_{h, i} = 0, Z_{h', i'} = 0) + Pr(Z_{h, i} = 1, Z_{h', i'} = 1) \\
                  &= Pr(Z_{h', i'} = 0 | Z_{h, i} = 0) Pr(X_{0} = 0) +
                     Pr(Z_{h', i'} = 1 | Z_{h, i} = 1) Pr(X_{0} = 1) \\
                  &= \big [ (1 - p_{h'}) + c p_{h} \big ] (1 - p_{h}) +
                     \big [ p_{h'} + c ( 1 - p_{h}) \big ] p_{h} \\
                  &= (1 - p_{h}) (1 - p_{h'}) + c p_{h} \big [1 - p_{h}] +
                     p_{h} p_{h'} + c p_{h} (1 - p_{h}) \\
                  &= (1 - p_{h}) (1 - p_{h'}) + p_{h} p_{h'} + 2 c p_{h} (1 - p_{h})
\end{align*}

__TODO: What is the general restriction on $c$ when $p_{h, i} \neq p_{h', i'}$?__

\subsection{Estimate of number of CpGs used by \citet{Lacey:2013iy}}\label{app:lacey}

I want to estimate the number of CpGs used by \citet{Lacey:2013iy} to estimate the correlation of methylation levels. They used the chromosome 11 data from an RRBS experiment on a normal myotube cell line, _MTCTL2_. Unfortunately, there is no URL provided to these data. \citet{Lacey:2013iy} cite their data as having come from \citet{Tsumagari:2013gn}, which includes a sample "Normal Human Skeletal Muscle Myoblasts", _HSMM_ ([http://genome-mirror.duhs.duke.edu/ENCODE/cellTypes.html](http://genome-mirror.duhs.duke.edu/ENCODE/cellTypes.html), [http://hgdownload.cse.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeHaibMethylRrbs/](http://hgdownload.cse.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeHaibMethylRrbs/)). I believe this to be the same cell line. In any case, for my purposes I simply needed some comparable RRBS data, not necessarily the exact same sample as  _MTCTL2_.

There are approximately $60,000$ CpGs on chromosome 11 in these comparable RRBS data. The following R code shows how I computed this using three RRBS replicates of the _HSMM_ sample.

Firstly, I download the data and unzip it:
```{r, cache = FALSE, message = FALSE}
library(R.utils)
library(data.table)
dl_dir <- "../data/appendix/"
url <- 'http://hgdownload.cse.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeHaibMethylRrbs/'
files <- paste0('wgEncodeHaibMethylRrbsHsmmtubefshdDukeSitesRep',
                1:3, '.bed.gz')
dests <- paste0(dl_dir, files)
x <- mapply(function(f, d) {
  if (!file.exists(gsub(".gz", "", paste0(dl_dir, f)))) {
    download.file(url = paste0(url, f), d)
    gunzip(filename = d, overwrite = TRUE)
    }
  fread(gsub(pattern = '.gz', replacement = '', x = d), sep = '\t',
        verbose = FALSE)
  }, files, dests, SIMPLIFY = FALSE)
```

Then I count the number of CpGs on chromosome 11 in each replicate:
```{r}
y <- rbindlist(x)
y[V1 == "chr11", .N, by = list(V4)]
```

\section{Computing details}\label{app:computing_details}

All computational work was performed on one of the Bioinformatics Division's HP Blade servers. These are shared-use, shared-memory machines. The basic specifications are shown in Table \ref{tab:unices}.

\begin{table}[h]
\centering
\begin{tabular}{@{}lllll@{}}
\toprule
Machine name & Processors & Cores per processor & Total number of cores & RAM    \\ \midrule
unix88       & $4\times$ Intel Xeon X7350 @ 2.93GHz & 4                   & 16                    & 128 GB \\
unix301      & $4\times$ AMD Opteron 8435 @ 2.6GHz  & 6                   & 24                    & 256 GB \\
unix302      & $4\times$ AMD Opteron 6174 @ 2.2GHz  & 12                  & 48                    & 512 GB \\
unix303      & $4\times$ AMD Opteron 6176 @ 2.3GHz  & 12                  & 48                    & 512 GB \\
unix305      & $4\times$ AMD Opteron 6276 @ 2.3GHz  & 16                  & 64                    &        \\ \bottomrule
\end{tabular}
\caption{Bioinformatics Division server specifications.}
\label{tab:unices}
\end{table}

\section{Software details}\label{app:software_details}

The analyses described in my thesis made extensive use of R (`R Under development (unstable) (2014-10-29 r66891)` and `R version 3.2.0 (2015-04-16)`) and Python (`v2.6.6`). All analyses were run on one of the servers described above. The scripts used to prepare the results for each chapter are available from \url{https://github.com/PeteHaitch/phd_thesis_analyses}. Further details of software used are available in Appendix \ref{app:software_details}.

I developed several pieces of software during my PhD. These are listed in below, along with the version used in for analyses in my thesis:

- `methtuple` (`v1.4.0`)
- `GenomicTuples` (`v1.2.1`)
- `MethylationTuples` (`v0.3.0.9007`)
- `methsim` (`v0.5.0.9013`)

In addition, I made use of the R packages listed in Table \ref{tab:R_packages} and gratefully acknowledge the developers. 

```{r R_packages, echo = FALSE, eval = TRUE, message = FALSE, results = "asis"}
library(xtable)
x <- read.table("../data/r_packages.txt",  header = TRUE,
                stringsAsFactors = FALSE, sep = "\t")
# Drop packages with "*" column set, which indicates the package is a
# "dependency of a dependency".
x <- x[x[, 2] != "*", ]
x <- x[, -2]
xt <- xtable(x, caption = c("R packages used in thesis as reported by \texttt{devtools::session\\_info()}", "R packages used in thesis"),
             label = "tab:R_packages")
align(xt) <- rep("c", ncol(x) + 1)
print(xt, booktabs = TRUE, comment = FALSE, tabular.environment = "longtable",
      floating = FALSE, include.rownames = FALSE, caption.placement = "top")
```
