---
title: "Biology background"
author: "Peter Hickey"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  pdf_document:
    fig_caption: yes
    includes:
      in_header: ../latex/header.tex
  html_document:
    keep_md: true
bibliography: ../latex/phd_thesis.bib
---

\section{Chapter Overview}\label{sec:biology_background_chapter_overview}

- [x] DONE

In this chapter I summarise the basic biology necessary for
understanding my thesis. I then introduce DNA methylation, describe some of its biological roles and discuss the types of questions asked in experiments studying DNA methylation. Finally, I describe assays for studying DNA methylation, with a particular focus on the whole-genome bisulfite-sequencing assay known as _methylC-seq_.

\section{From DNA to genomes}\label{ref:from_dna_to_genomes}

- [x] DONE

The genome is the genetic material of an organism. In most organisms, the genome is encoded by deoxyribonucleic acid, _DNA_. __FIGURE__ beautifully illustrates the hierarchy that takes us from DNA through to the genome.

In eukaryotes[^eukaryotes], which includes plants, fungi and animals, the DNA is wound around repeating units of eight _histone_ protein cores to form _nucleosomes_, which are the fundamental unit of _eukaryotic chromatin_. Chromatin compactly packages the DNA into _chromosomes_, so that the organism's complete nuclear DNA, its nuclear genome, might fit into the nucleus of its cells[^extra_nuclear_genomes].

[^eukaryotes]: Eukaryotes are organisms composed of one or more cells with a distinct nucleus and cytoplasm. Most of a eukaryotic cell's DNA is contained in the nucleus \cite{Alberts:2007vj}.

[^extra_nuclear_genomes]: Not all of an organism's genome is present in the nucleus of a cell. Important exceptions are mitochondrial DNA (mtDNA) found in animals, plants and fungi and chloroplast DNA (ctDNA) found in plants.

\subsection{DNA}\label{sec:DNA}

The double helix is the most common, and most famous, structure of DNA (__REFERENCE__). In the double helix, the two strands of DNA run in opposite directions to each other and are therefore _anti-parallel_. One strand is called the 5' strand (pronounced "5 prime strand") - also known as the sense strand, Crick strand or top strand - and the other strand called the 3' strand (pronounced "3 prime strand") - also known as the antisense strand, Watson strand or bottom strand. Along each strand of the double helix are the four DNA nucleobases (_bases_): adenine (_A_), cytosine (_C_), guanine (_G_) and thymine (_T_). These bases form _complementary_ base pairings, _A_ with _T_ and _C_ with _G_, along the DNA double helix.

[__FIGURE OF DNA BASE  PAIRING__](http://upload.wikimedia.org/wikipedia/commons/4/4c/DNA_Structure%2BKey%2BLabelled.pn_NoBB.png).

A gene is a sequence of DNA that is _transcribed_ to produce a functional product called ribonucleic acid or _RNA_ (Source: [http://sandwalk.blogspot.com.au/2007/01/what-is-gene.html](http://sandwalk.blogspot.com.au/2007/01/what-is-gene.html)). RNA may in turn be translated into a protein sequence or perform other roles in the regulation of gene expression (__CITE__). It is important to note here than not all DNA is transcribed into RNA. That is not to say that untranscribed DNA is unimportant. For instance, there are untranscribed _regulatory sequences_ of DNA that determine whether a nearby gene is transcribed. There is also _junk DNA_, _garbage DNA_, __etc.__ (__SOURCE__). Conversely, not all transcribed DNA is a gene. DNA transcription is "permissive" and there are many DNA sequences that are transcribed by accident or in error.

DNA is able to self-replicate so that, for example, eukaryotic cells created during _mitosis_ contain the same DNA as the "parent" cell. During DNA replication, the two strands are separated and each strand's _complementary_ DNA sequence is copied by an enzyme called _DNA polymerase_. It is because the two strands of DNA are complementary that ensures the daughter cell contains the same DNA sequence as the parent cell[^dna_replication].

[^dna_replication]: This of course ignores errors in the replication process. Such errors are very rare events but because DNA replication happens __HOW FREQUENTLY__ these events do occur. There are error-correcting processes that reduce the chance that such an error is retained in the daughter sequence, however, these are not perfect. Hence errors in DNA replication are one source of what are known as _mutations_ in the DNA.

\subsection{Nucleosomes and chromatin}\label{sec:nucleosome_and_chromatin}

Basic descriptions of nucleosomes and chromatin are ripe with analogies. In one such popular analogy, the histones are the 'spool' around which the DNA 'thread' is wrapped to form nucleosomes (__CITE__). These nucleosomes themselves have the appearance of "beads on a string" when viewed under an electron microscope (__CITE__).

The core of a nucleosome consists of four pairs of histones, H2A, H2B, H3 and H4, which are consequently known as the _core histones_. A fifth histone, H1/H5, is known as the _linker histone_. Each of these histones has a 'tail' consisting of a string of amino acids. These tails can undergo _post-translational modifications_, such as methylation, acetylation and phosphorylation, which can alter their interactions with DNA and nuclear proteins (__CITE__) (see below Section \ref{epigenetics} on epigenetics).

The nucleosomes are interconnected by _linker DNA_ to form the macromolecule  called chromatin. The linker sequences are between $20-60$ nucleotides in length, while approximately 147 nucleotides are wrapped around each nucleosome and a further 20 nucleotides wrapped around each additional H1/H5 histone (__CITE: http://www.nature.com/scitable/topicpage/dna-packaging-nucleosomes-and-chromatin-310#__).

Chromatin is frequently described as either _heterochromatin_ or _euchromatin_. Heterochromatin, or closed chromatin, is more tightly packed than euchromatin, or open chromatin. Heterochromatin is associated with transcriptionally repressed regions of the genome because the machinery required to translate DNA to RNA is unable to physically access the DNA. In contrast, euchromatin is associated with transcriptionally active regions of the genome since its more open nature allows easier access for the translational machinery.

The same region of an individual's genome may vary between heterochromatin and euchromatin states at different stages of the organism's life and in different cells of the organism. This is one mechanism which by gene expression is regulated.

Chromatin is further coiled up into various literally-named structures, such as the 30-nanometre and 250-nanometre fibres, and, ultimately, packaged into _chromosomes_. The set of chromosomes makes up an individual's _genome_.

\subsection{Genomes}\label{sec:genomes}

As is clear from the above description, the genome is a complex three-dimensional structure. Nonetheless, in bioinformatics and computational biology, the genome is mostly considered as a single-stranded, one-dimensional string of the nucleobases _A_, _C_, _G_ and _T_.

While eukaryotic genomes share the above-described features, and indeed share many regions of common DNA sequence, eukaryotic genomes come in many shapes and sizes. I only discuss the genomes of three organisms relevant to my thesis: Human (_Homo sapiens_), Mouse (_Mus musculus_) and Arabidopsis (_Arabidopsis thaliana_).

\subsubsection{Human genome}\label{sec:human_genome}

Humans are _diploid_ organisms, meaning that we have two copies of each chromosome in a typical cell[^somatic_cell]. We inherit one chromosome of each pair from our mother and one from our father. A typical human cell has 23 pairs of nuclear chromosomes, 22 _autosomes_ and 1 pair of _sex chromosomes_, as well as hundreds or thousands of copy of the small mitochondrial chromosome[^mtDNA].

[^somatic_cell]: A sperm or egg cell is haploid and has a single (recombined) copy of each chromosome.

[^mtDNA]: The mitochondrial DNA is maternally inherited.

The length of a chromosome is typically reported as the number of DNA base pairs (bp) in a single copy of that chromosome and the _haploid_ length of a genome is the sum of these chromosome lengths. The haploid human genome is approximately 3.3 billion base pairs long.

A _reference genome_ was completed by the Human Genome Project in 2003 (__CITE__). This reference genome does not represent the genome of any one  human since it uses DNA donated by several different people. Rather, a reference genome is a kind of map or scaffold that can be used to identify similarities and differences between individual genomes.

The human genome has obvious uses in medical research and biotechnology, but is also used to learn about evolution and human history, such as migration. Every person, except genetically identical twins, has their own unique genome. However, the genomes of any two randomly selected people are identical at __WHAT PERCENTAGE__ of sites. The vast majority, __WHAT PERCENTAGE__, of the human genome is made up of non-coding DNA. This is similar to other organisms with similarly large genomes (__CITE__). Much of the non-coding DNA is junk DNA. Indeed, over __WHAT PERCENTAGE__ of the human genome is comprised of repetitive DNA.

\subsubsection{Mouse genome}\label{sec:mouse_genome}

Mice are also diploid organisms, but have 19 pairs of autosomes, one pair of sex chromosomes and a mitochondrial chromosome. The mouse genome is slightly smaller than the human genome, at 2.8 billion base pairs long. Like the human genome, there is a mouse reference genome (__CITE__). It is based on several female mice from the _C57BL/6J_ strain, an important strain of mouse widely used in medical research.

Mouse strains used in medical research are highly inbred due to years of concerted mating programs. This reduced genetic variability, and the control that researchers have over it, make these mice a very powerful tool in identifying the biological cause of a diverse range of phenotypes.

\subsubsection{Arabidopsis Thaliana genome}\label{sec:human_genome}

Arabidopsis is a small flowering plant. The most widely studied species is _A. thaliana_, which is diploid and has one of the smallest genomes known in the plant kingdom (__CITE__). The genome of _A. thaliana_ is 135 million base pairs long and is packaged as five pairs of chromosomes (__CITE__). The small size of its genome, coupled with its small physical size, rapid development and its tolerance to both cross-pollination and self-pollination[^cross_and_self_pollination], make _A. thaliana_ an ideal organism for studying plant development.

[^cross_and_self_pollination]: The advantage in genetic studies of using a plant that is tolerant of self-pollination is that recessive mutations can be quickly made homozygous and thus expressed. Cross-pollination allows the creation of new strains with multiple mutations.

\subsection{Genetic variation}\label{sec:genetic_variation}

A key question in biology, perhaps _the_ key question, is what determines an organism's _phenotype_. An individual's phenotype is "the set of observable characteristics of an individual resulting from the interaction of its genotype with the environment" (__CITE: currently using google's definition__). Two simple examples of phenotypes are height and weight. Both have a genetic component, e.g., the offspring of tall parents are on average taller than those offspring of short parents, but there are also environmental components, such as the contribution of diet to weight. In medical research, a person's phenotype might be whether she is affected by a particular disease. It might also be some proxy, such as her blood pressure or the expression levels of particular genes.

Variation between individual's genomes, be it at single nucleotides[^snvs_snps] or across larger regions[^indels], is one important source of phenotypic variation. Importantly, genetically-driven phenotypic variation is frequently _heritable_, meaning that phenotypes can be passed on from one generation to the generation via the genome.

[^snvs_snps]: An individual nucleotide that is variable in the population is called a single nucletoide variant (_SNV_). A SNV that is frequently variable in the population, say at least $1\%$ of the population are variable at that site, is called a single nucleotide polymorphism (SNP). All SNPs are SNVs but the converse is not true.

[^indels]: __TODO: Define__

Environmental variation has clear influence on certain phenotypes. However, it often can be difficult to determine whether phenotypic variation is driven by genetic varation or environment variation, particularly in humans where genetically similar individuals typically grow up in similar environments.

The above discussion has been about phenotypic variation in a population. But there is also phenotypic variation within the individual. If that sounds strange, consider the fact that in your body a neuron, a leukocyte (a white blood cell) and a cone cell (a photoreceptor in the retina) all have identical genomes. In fact, all cells in an organism, excluding the gametes, have an identical genome yet play very different roles. This is due to different genes being active in different cells.

_Epigenetics_, which I describe in the next section, plays a role in determining both the between-individual phenotypic variation as well as the within-individual phenotypic variation.

\section{Epigenetics}\label{sec:epigenetics}

- [x] DONE

"Epigenetics" is a real Humpty Dumpty phrase; each author seems to believe that, "When I use [the] word, ... it means just what I choose it to mean - neither more nor less" \cite{carroll1897through}. I am sure to continue in this grand tradition.

Conrad Waddington coined the phrase in 1942 as a portmanteau of the words "epigenesis" and "genetics" \cite{Waddington:2012bu}. Waddington meant epigenetics as the study of how "processes involved in the mechanism by which the genes of the genotype bring about phenotypic effects" \cite{Waddington:2012bu}.

A popular contemporary definition of epigenetics is attributed to the epigeneticist Art Riggs that epigenetics is "the study of mitotically and/or meiotically heritable changes in gene function that cannot be explained by changes in DNA sequence[^mitotic_meiotic]" \cite[p1]{russo1996epigenetic}. The _epi_ prefix, derived from the Greek word for 'upon', 'near to', or 'in addition', emphasises the idea that epigenetics encodes information 'on top of' the DNA sequence. However, it is quite different to Waddington's original definition.

[^mitotic_meiotic]: Mitotically heritable means heritable during cell division and meitotically heritable means heritable during sexual reproduction.

More recently, the definition of epigenetics has taken on a "more biochemical flavour" \cite{Daxinger:2010bu} to include marks whose heritability is yet to be established. The heritability, or lack thereof, of histone modifications means that many epigeneticists do not consider these to be truly epigenetic \cite{Berger:2009ex}, and describing them as such is a sure-fire way to annoy a good percentage of your audience \cite{Ledford:2008ew}.

Sir Adrian Bird, an esteemed British geneticist and someone who is in favour of the broader definition, attempts to unite these definitions in the following \cite{Bird:2007em}:

> The following could be a unifying definition of epigenetic events: the structural adaptation of chromosomal regions so as to register, signal or perpetuate altered activity states. This definition is inclusive of chromosomal marks [e.g. histone marks], because transient modifications associated with both DNA repair or cell-cycle phases and stable changes maintained across multiple cell generations qualify.

Regardless of which definition you subscribe to, an _epigenetic mark_ is the modification that causes this "epigenetic change". Actually, "causes" may be too strong a phrase as much of current epigenetics research is in identifying associations rather than causations and the question of whether the epigenetic mark is a cause or consequence of the ascribed function is oft-debated.

The _epigenome_ of a cell is the set of epigenetic marks present on the cell's genome. In contrast to the genome, which is identical between cells within an individual, the epigenome is highly variable between cells within an individual. Indeed, we can identify variation for a single epigenetic mark within cells of the same cell type from the same individual (__TODO: cite examples or upcoming chapters__). Epigenomics is the study of the epigenome, analogous to genomics being the study of the genome. However, one can rarely study the epigenome in isolation from the underlying genetic sequence, as there is much evidence that genetic variation can drive epigenetic variation (__CITE__).

One mark that most authors agree is an epigenetic mark is DNA methylation, which I describe in the next section and whose study is the focus of my thesis.

\section{DNA methylation}\label{sec:dna_methylation}

- [x] DONE

DNA methylation is a chemical modification of DNA that can impart information on top of the DNA sequence. It is heritable during mitotic cell division, which means that it is faithfully copied across to the daughter cell during cell division[^dnam_copying]. It therefore fits into Art Riggs' afore-mentioned definition of an epigenetic modification \cite[p1]{russo1996epigenetic}. DNA methylation is found in bacteria, fungi, plants and animals.


[^dnam_copying]: In practice this copying is not as faithful as, say, the copying of DNA from the parent to the daughter strand. Furthermore, the faithfulness of this copying will be different in different conditions, such as in a healthy, well-differentiated liver cell compared to a cancerous liver cell. Nevertheless, the copying of DNA methylation is faithful enough for most biologists to consider it as a mitotically heritable mark, most of the time (__TODO__: Is this just true of 5mC in mammals or more generally?)

A major reason that DNA methylation is studied is that it is essential for normal development in mammals \cite{Li:1992bp}. It is also involved in many key biological processes related to human development and health, such as the regulation of gene expression (__CITE__), silencing of transposable elements (__CITE__), X chromosome inactivation (__CITE__) and tumorigenesis (__CITE__). However, DNA methylation is not found in all organisms, for example, it not detectable in yeast, and so is not essential for life. Furthermore, the level of DNA methylation varies widely amongst different organisms, with many having very low levels of methylation (__TODO: examples and citations__).

When people speak of DNA methylation they are generally referring to methylation of the cytosine nucleobase. Even more specifically, they are referring to the molecule _5-methylcytosine_. 5-methylcytosine, abbreviated as _5mC_, is by far the most common form of DNA methylation in the animal and plant kingdoms[^other_DNAm]. However, I will continue to use the terms _DNA methylation_ and _cytosine methylation_ interchangeably, as is standard in the literature. The importance of 5-methylcytosine in mammalian genomes is such it has been dubbed the "fifth base" of the DNA code \cite{@Lister:2009boa}.

[^other_DNAm]: Two additional forms of DNA methylation are N6-methyladenine (m6A) and N4-methylcytosine (m4C). N6-methyladenine is a methylated form of adenine, which is found in mRNA \cite[e.g.,]{Adams:1975ki} and DNA, although the latter only in bacterial DNA \cite{Ratel:2006dv}. N4-methylcytosine has also been detected in bacterial DNA \cite[e.g.,]{Ehrlich:1985bv, Ratel:2006dv}.

A German chemist, W.G. Ruppel, first identified a methylated nucleic acid in 1898. Ruppel was studying _tuberculinic acid_, the poison of _Mycobacterium tuberculosis_[^tuberculosis] , and discovered that it contained a methylated nucleotide (__TODO: Can only find first page of this paper and it's in German. Need to verify__). In 1925, Johnson and Coghill isolated 5-methylcytosine as a product of hydrolysis of tuberculinic acid, the nucleic acid of _Mycobacterium tuberculosis_ \cite{Johnson:1925js}. However, Johnson and Coghill's results were disputed for over twenty years by other researchers who were unable to replicate the original findings \cite{@VISCHER:1949ty}.

[^tuberculosis]: _Mycobacterium tuberculosis_ was then known as _Tubercle bacillus_.

In 1945, Hotchkiss ultimately proved Johnson and Coghill correct when he isolated _5-methylcytosine_ from nucleic acid prepared from cow thymus \cite{HOTCHKISS:1948va}. Using paper chromatography, Hotchkiss demonstrated that methylated cytosine existed and was distinct from conventional cytosine and uracil.

\subsection{DNA methylation in mammals}\label{sec:dna_methylation_in_mammals}

- [x] Done

DNA methylation is essential for normal development in mammals, such as human and mouse (__REFERENCE__). My thesis focuses on DNA methylation in humans but much of the material in this section is true for other mammals.

The typical site of DNA methylation is at the C5 carbon position of a cytosine nucleotide. This modified cytosine is called _5-Methylcytosine_ or _5mC_. [__FIGURE__](http://en.wikipedia.org/wiki/File:5-Methylcytosine.svg) shows the structure of 5mC. It is common to describe a cytosine as being "methylated" or "unmethylated", however, care must be taken in defining these terms.

At the lowest level, the level of single-stranded DNA, methylation is a binary event: a cytosine is either methylated or unmethylated. Double-stranded DNA, at least at _palindromic_ methylation loci[^palindromic_sequence], is generally symmetrically methylated, i.e., the loci on each strand are both methylated or both unmethylated. However, _hemimethylation_, where the methylation loci on one strand is methylated and its partner on the opposite strand is unmethylated, can and does occur.

[^palindromic_sequence]: "A palindromic sequence is a nucleic acid sequence (DNA or RNA) that is the same whether read 5' (five-prime) to 3' (three prime) on one strand or 5' to 3' on the complementary strand with which it forms a double helix" (__TODO: current source is wikipedia, find and properly cite__).

Within a diploid cell, a particular cytosine may be unmethylated or methylated on both homologous chromosomes or methylated on one chromosome and unmethylated on the other. While the former that is more common, examples of the latter case, such as _allele-specific methylation_ and _genomic imprinting_, are of considerable biological interest (__TODO: sources and examples__).

In mammalian genomes, most cytosines are unmethylated except for those at _CpG dinucleotides_. A _CpG dinucleotide_, or more simply a _CpG_, is a cytosine followed by a guanine in the linear DNA sequence (__FIGURE__). The "p" stands for the phosphate backbone of DNA and some authors omit it in favour of simply calling it CG methylation. A CpG is a palindromic sequence and is generally symmetrically methylated. Approximately 60-80% of CpGs are methylated in mammals, meaning that the cytosine in the dinculeotide is a 5-methylcytosine (__REFERENCE__).

\subsection{CpG dinucleotides}\label{cpg_dinucleotides}

- [x] Done

CpGs are underrepresented in the human genome. The GC-content of the human genome, which is defined as the percentage of nucleotides that are either guanines or cytosines, is approximately __WHAT %__. Under a random sampling model of nucleotides we would expect approximately __WHAT %__ of the human genome to be comprised of CpGs. In fact, only __WHAT %__ of the dinucleotides are CpGs. __FIGURE__ shows the frequency of each dinucleotide in the human reference genome (hg19)[^cpg_frequency].

[^cpg_frequency]: See __APPENDIX__ for the `R` code used to compute these values for the human reference genome (build hg19).

One reason for this relative scarcity of CpGs is that methylated cytosines can spontaneously deaminate to thymines \cite{@Scarano:1967ue}. Thus, over time, many methylated CpGs will become TpGs, leading to a genome-wide reduction in the proportion of CpGs and a genome-wide increase in the proportion of TpGs (see __FIGURE__).

There are many other evolutionary pressures on the distribution of nucleotides in a genome. One effect of this is that the distribution of CpGs is far from uniform. In fact, CpGs tend to form clusters, which are termed CpG islands.

\subsection{CpG islands and other sandy metaphors}\label{sec:cpg_islands}

- [ ] Done

One way to explore the distribution of CpGs in the human genome is to look at the distribution of distances from one CpG to the next. __FIGURE__ is a kernel density plot of the distances between neighbouring CpGs  and __FIGURE__ is the empirical cumulative distribution function of the distances between neighbouring CpGs for the human reference genome (hg19)[^cpg_distance].

[^cpg_distance]: See __APPENDIX__ for the `R` code used to generate these figures.

As can be seen from __FIGURE__, there are a subset of CpGs that are within __HOW MANY__ bases of the next CpG.

The 20-40% of CpGs that are methylated in mammalian genomes are mostly found in clusters called _CpG islands_ (_CGIs_). Again, there are evolutionary arguments for why these CGIs form (__TODO: sources__).

CGIs are important regulatory elements in the genome and are where most differences in DNA methylation between different cell types are found (__TODO: sources__). The classical definition of a CpG island, given by \cite{GardinerGarden:1987we} and used by, for example, the UCSC genome browser, is a region of the genome where:

1. The moving average of GC-content is greater than $50\%$, and
2. The observed-to-expected ratio of CpGs is greater than $0.6$, and
3. The region is longer than 200 bp.

This definition was refined by \cite{Takai:2002jb} to exclude _Alu_-repetitive elements, which otherwise are misclassified as _bona fide_ CpG islands. More recently, \cite{@Wu:2010do} developed a hidden Markov model to predict CpG islands based on the CpG density and GC-content of the region.

An alternative definition, and one that pre-dates the definition of \cite{GardinerGarden:1987we}, is based on the identification of unmethylated regions of the genome, which are typically CpG-dense. Such regions were previously called HpaII tiny fragment islands, or _HTF islands_, and named after the restriction enzyme used to identify them \cite{Cooper:1983ev, Bird:1985gj}.

The "sandy/beachy" metaphors have been continued (i.e., stretched to breaking point) with various authors defining CpG island shores, CpG island shelves, CpG canyons and CpGs in the "open sea":

- CpG island shores: regions within 2 kb of CpG islands \cite{Irizarry:2008hg}. CpG island shores have been demonstrated to have an increased variability of CpG methylation \cite{Irizarry:2008hg}.
- CpG island shelves have been defined as regions within 2 kb of a CpG island shore \cite{Bibikova:2011fw}.

__UP TO HERE__

- CpG canyons are __WHAT__ and the
- "open sea" is __WHAT AND HOW DOES THIS DIFFER FROM CpG DESERTS__.
- methylation deserts: 100 Kbp windows with the lowest 1% methylation level in the sample of interest

Another metaphor, based on methylation levels rather than CpG density, are methylation deserts. These regions are sample-dependent and were defined by \cite{Li:2012gg} as 100 Kbp windows with the lowest 1% methylation level in the given sample.

\subsection{Non-CpG methylation}\label{sec:non_cpg_methylation}

- [ ] Done

In humans, cytosine methylation in most cell types is found almost exclusively at CpGs. There are, however, certain cell types with widespread non-CpG methylation. Non-CpG methylation is often classified as CHH methylation or CHG methylation, where H stands for any nucleotide except G (__SOURCE IUPAC__). The rule-of-thumb is that non-CpG methylation is rare in somatic cells but common in pluripotent cells. Of course, there are exceptions to every rule, especially in biology.

To give a few examples, [@Lister:2009hy] found that in a somatic cell line (_fibroblasts_ or skin cells) $99.98\%$ of methylcytosines occured at CpGs whereas as in an embryonic stem cell line $24.5\%$ of methylcytosines occured in a non-CpG context. A subsequent paper from the same group ([@Lister:2011kg]) extended this result when they reported that, more generally, non-CpG methylation accounts for $20-30\%$ of methylcytosines in _pluripotent_ cell lines (which includes embryonic stem cell lines along with induced pluripotent stem cell lines, see __SECTION___). An exception to the rule is provided by [@Lister:2013et], where they found that neurons, a somatic and not a pluripotent type of cell, also have non-CpG methylation, albeit at a lower level ($1.3-1.5\%$ of all non-CpG cytosines were methylated).

Non-CpG methylation is less well studied than CpG methylation. This is partly due to the design of popular assays for studying cytosine methylation; for example the popular Illumina 27k and 450k beadchips (see __SECTION__) only measure cytosine methylation at CpGs __CHECK__. However, recent technological advances (see __SECTION__) mean that cytosine methylation can be routinely assayed regardless of the sequence context. The biological role of non-CpG methylation is less well understood than that of CpG methylation (__SOURCE__).

\subsection{Modifications of a modification}\label{sec:modifications_of_a_modification}

- [ ] Done

Methylation is not the only chemical modification of cytosine nucleobase, although it is by far the most common. Further modifications, listed approximately from most frequent to least frequent, are 5-hydroxymethylcytosine (5hmC), 5-formylcytosine (5fC) and 5-carboxylcytosine (5caC) (__TODO:__ cite 10.1038/nrg3772).

The biological significance of these marks is still being determined, in part because the assays for studying these are still in development (__SOURCE__) and because their relatively scarcity means that experiments to detect these modifications are more difficult and expensive.

One genome-wide study of 5hmC found that less than $ 1\%$ of all assayed cytosines in mouse fetal cortex and adult cortex cells were hydroxymethylated [@Lister:2013et]. Most of the 5hmC was detected in the CpG context and, although the genome-wide level of 5hmC was low, the authors reported significant levels of 5hmC at particular cytosines in the genome.

[@Kriaucionis:2009bm] and [@Tahiliani:2009kl] discovered that the TET enzymes can convert 5mC to 5hmC, 5hmC to 5fC and 5fC to hcaC. This suggested a role for 5hmC, 5fC and 5caC in the process of removing 5mC marks, as is discussed in the next section.

\subsection{Writers, readers and erasers}\label{sec:writers_reads_and_erasers}


A common analogy used in describing DNA methylation is that of "writers", "readers" and "erasers" [@Moore:2013in]. Writers catalyse the methyl group onto the DNA, readers recognise methylated DNA and erasers remove the methyl group from the DNA.

In mammalian cells, the writers are the DNA _methyltransferase_ (DNMT) enzymes. These DNMTs are commonly split into two groups, namely the maintainence methyltransferase (thought to be DNMT1) and the _de novo_ methyltransferases (thought to be DNMT3a and DNMT3b)[^DNMT]. DNA methylation is not preserved by the DNA replication machinery and so it is the role of DNMT1 to restore the methylation pattern on the daughter strand of DNA following DNA replication. In contrast, DNMT3a and DNMT3b lay down new methylation marks and are particularly active during development when there are widespread changes in DNA methylation (__SOURCE__). DNMT1 and DNMT3b appear to be essential for mammalian development as mouse knockouts[^KO] for either gene is embryonically lethal (__SOURCE: http://labs.genetics.ucla.edu/fan/papers/npp2012112a.pdf__) whereas mouse knockouts for DNMT3a are runted but survive for ~4 weeks after birth (__SOURCE: http://labs.genetics.ucla.edu/fan/papers/npp2012112a.pdf__).

[^DNMT]: DNMT2, now known as TRDMT1, was once thought to be a DNA methyltransferase, however, __http://www.sciencemag.org/content/311/5759/395__ showed that it in fact methylates a small RNA and not DNA. Another protein, DNMT3L, is homologous to DNMT3a and DNMT3b but does contain catalytic domain that is necessary for methyltransferase activity. Instead, DNMT3L is thought to stimulate the activity of DNMT3a and DNMT3b (__SOURCE__).

[^KO]: A knockout mouse for gene $X$ is a mouse that has been genetically engineered to remove or otherwise inactivate gene $X$. Mouse knockouts can be either heterozygous knockouts (one copy still of the gene is still present/active) or homozygous knockouts (both copies of gene absent/inactive).

The readers of DNA methylation recognise methylated DNA. These readers can recruit additional proteins to the site of the methylated cytosine to perform a variety of functions related to gene expression. For example, the _MBD_ (methyl-CpG-binding domain) group of proteins bind to DNA containing a methylated CpG which then suppress gene expression by preventing transcription factor binding at that site (__SOURCE__). Another group, the _UHRF_ (ubiquitin-like, containing PHD and RING finger domain) proteins, help DNMT1 methylate _hemimethylated_ DNA, which is DNA where one strand is methylated and the other is not (e.g. the parent strand compared to the daughter strand following DNA replication).

The removal or erasure of DNA methylation, _demethylation_, is characterised as _passive_ loss or _active_ removal. Passive loss occurs when the maintainence methyltransferases do not efficiently perform their role of restoring DNA methylation following cell division. This leads to a gradual, stochastic and genome-wide loss of DNA methylation after multiple cell divisions. This form of passive demethylation, sometimes called replication-dependent demethylation, cannot explain observations of precise targeted demethylation (__EXAMPLE__) nor the two stages of rapid global demethylation that occur during development [@Wu:2014gw].

Active demethylation is currently a hot topic in epigenetics research. Multiple mechanisms have been proposed, and it is indeed likely that there are multiple ways to achieve active demethylation. These mechanisms were recently reviewed by [@Wu:2014gw], which I now briefly summarise:

1. The direct removal of the methyl group from 5mC is considered unlikely due to the strong carbon-carbon bond between the methyl group and the cytosine.
2. There is evidence that the DNA repair machinery can be co-opted to remove a methylated base or the surrounding region. The excised base or region is then repaired with unmethylated cytosines replacing 5mCs.
3. 5mC oxidation-dependent active DNA demethylation. This follows from the observation that the TET enzymes can iteratively oxidate a 5mC -> 5hmC -> 5fC -> 5caC reaction. The removal of 5hmC, 5fC or 5caC is biochemically "easier" than the removal of 5mC and could occur via a more efficient form of replication-dependent demethylation, direct removal of the oxidized methyl group or through the DNA repair machinery (__FIGURE OF TET-CYCLE__).

One question raised by point (3) is whether 5hmC, 5fC and 5caC are simply intermediate products in an active demethylation cycle or if they themselves are bona fide epigenetic marks.

\subsection{Function of DNA methylation: "normal", cancer and development}\label{sec:function_of_dna_methylation}

- [ ] Done

\subsection{DNA methylation in other organisms}\label{sec:dna_methylation_in_other_organisms}

- [ ] Done

__TODO: Don't want too much detail here__

\subsubsection{Insect}

- [ ] Done

* Honeybees
* Fruit fly? DNA methylation's existance in other organisms, such as the common fruit fly (_Drosophila Melanogaster_), is an area of contention although recent studies suggest it is present albeit at very small quantities and only in very limited contexts [@Takayama:2014jp].
* General

\subsubsection{Plants}

- [ ] Done


* Arabadopsis
* Non-CG methylation

\subsubsection{Fungi}

- [ ] Done

\subsubsection{Bacteria}

- [ ] Done

\subsubsection{Archaea}

- [ ] Done

\section{Assays for studying DNA methylation}\label{sec:dna_methylation_assays}

- [ ] DONE

A challenge to measuring DNA methylation is that it is erased by standard molecular biology techniques, such as _polymerase chain reaction_ (PCR) and bacterial cloning, and it is not revealed by DNA hybridization assays [@Laird:2010iv]. Therefore, almost all assays of DNA methylation require one of the following _pre-treatments_ of the DNA:

1. _Enzyme digestion_
2. _Affinity enrichment_
3. _Sodium bisulfite conversion_

Each of these pre-treatments can then be combined with a second step of amplification or hybridisation. For example,

1. Gel-based analysis
2. Sanger sequencing
3. Microarray hybridisation
4. Massively parallel sequencing

An exception to this classification scheme are a new class of assay that seek to directly "read" whether a position is methylated or unmethylated without requiring a pre-treatment of the DNA. For example, both [@Laszlo:2013kf, Schreiber:2013in] measure the change in current as a DNA molecule passes through a nanopore to infer whether a cytosine was methylated while __PACBIO AND ONT DO WHAT?__.

Almost all assays of DNA methylation measure a "population average" from a pool of hundreds or thousands of cells. For a diploid organism, this is an average over several different levels: the two DNA strands, the two homologous chromosomes within a diploid cell and the hundreds or thousand of cells used in the assay (see __FIGURE__). Hundreds or thousands of cells are required in order to have sufficient material as input for the assay. Assays that require only a single cell as input do exist but are still in development and not in widespread use (__SOURCE/EXAMPLES__).

The _resolution_ and _throughput_ are two key variables when choosing which assay to use for an experiment. The resolution of an assay is the scale on which DNA methylation can be measured[^resolution]. For example, a high resolution assay allows a researcher to quantify the level of DNA methylation at a single base whereas a low resolution assay might only allow for qualitative assessment (i.e. presence or absence) of DNA methylation at larger regions, such as CpG islands.

[^resolution]: Depending on the experiment and its aims, the resolution of an assay might instead be defined as the scale on which DNA methylation can be quantified or that which allows inference to address a specific hypothesis.

The throughput of an assay can be quantified in two ways. The first is per-sample throughput, which is how many measurements of DNA methylation are made per-sample[^throughput1]. This is typically what people mean when they describe an assay as being "high-throughput" or "low-throughput" and is the definition I use in the title of my thesis. Depending on your definition of "high", a high-througput assay will produce on the order of tens of thousands to billions of measurements per sample. The second definition of throughput is per-dollar, that is, "how many samples can I process within my budget?". To avoid confusion, __I should refer to the latter definition as something else__.

[^throughput1]: This might reasonably be argued as being a definition of resolution.

The choice of which assay to use for an experiment is a trade-off between resolution, per-sample throughput and per-dollar throughput. Experiments that use an assay with high-resolution and high per-sample throughput generally have fewer samples than experiments using a lower resolution assay or an assay with lower per-sample throughput.

In this section I will describe each of the pre-treatments but will focus on the bisulfite-conversion assays. In particular, I will describe in detail the "gold standard" assay of DNA methylation, _whole-genome bisulfite-sequencing_ (WGBS), that combines the sodium bisulfite conversion pre-treatment with massively parallel sequencing to produce whole-genome maps of DNA methylation at base pair resolution.


\subsection{Enzyme digestion assays}\label{sec:enzyme_digestion_assays}

- [ ] Done

_Restriction endonucleases_ are an important technique in molecular biology. These enzymes can preferrentially "cut/cleave/digest" DNA at particular sequence motifs. The motif at which a restriction enzyme cleaves DNA is called the _recognition motif_ or _restriction sequence_. The methylation of a position in the recognication motif can inhibit a restriction enzyme from cleaving the DNA. This can be used to design an assay to infer the methylation state of a DNA fragment.

For example, the recognition site of the restriction enzyme _HpaII_ is `CCGC`. However, _HpaII_ will only digest DNA when the second cytosine in the motif is unmethylated. The _HELP_ (_HpaII_ tiny fragment enrichment by ligation-mediated PCR) assay compares DNA digested by _HpaII_ to one digested with another restriction enzyme that has the same recognition motif but is methylation-insensitive (_MspI_) to identify _hypomethylated_ regions of a genome (__SOURCE__).

Another popular restriction enzyme for studying DNA methylation is _McrBC_, "an enzyme with the unusual and desirable property of cutting methylated DNA promiscuously (recognition sequence $R^{m} C(N)_{55– 103} R^{m} C$)" [@Irizarry:2008hg].

Assays based on restriction enzyme were some of the first developed for studying DNA methylation. They were initially developed for studying particular loci although they have been extended to genome-scale analysis approaches [@Laird:2010iv]. Restriction enzyme assays have a relatively low resolution (__REASONS/SOURCE__) but may also be used as a first-pass to enrich a sample for methylated or unmethylated regions which are then assayed using a higher resolution technique.

\subsection{Affinity enrichment assays}\label{sec:affinity_enrichment_assays}

- [ ] Done

Affinity enrichment assays compare measurements between an "enriched" version and an "input" (control) version of the same sample to infer the presence or absence of DNA methylation. This is very similar to the idea behind chromatin immunoprecipitation followed by microarray hybridisation (ChIP-chip) or sequencing (ChIP-seq). Enrichment may be done using a methylation-specific antibody, such as __EXAMPLE__, or a __unmethylation-specific(?)__ antibody, such as __EXAMPLE__.

Some examples of affinity enrichment assays for DNA methylation are the microarray-based MeDIP, mDIP and mCIP (all based on methylated DNA immunoprecipitation) and their sequencing-based relatives, MeDIP-seq and mDIP-seq. These are all low resolution assays since they are based on the enrichment of regional differences between the enriched and input samples. Furthermore, the bioinformatic analysis of data from these assays is complicated by the varying CpG density along the genome, which leads to different enrichment affinities for different regions of the genome. However, these assays are can provide a relatively cheap and efficient genome-wide assessment of DNA methylation [@Laird:2010iv}].

\subsection{Sodium bisulfite conversion assays}\label{sec:bs_assays}

- [ ] Done

In the 1980s, __Hayatsu OR Wang et al.__ discovered that when _denatured_ (__DEFINE?__) DNA is treated with sodium bisulfite (__CHEMICAL SYMBOL__), unmethylated cytosines deaminate to uracils much faster than do methylated cytosines (__SOURCE__). The methylated cytosines are said to be "protected" from conversion to uracils. When the bisulfite treated is amplified by PCR the uracils are subsequently converted to thymines.

This discovery led to the development of assays for studying studying cytosine methylation based on the pre-treatment of DNA with sodium bisulfite, which are sometimes referred to as bisulfite-conversion assays (__FROMMER; CLARK__). These assays are all based on the idea of comparing the sequence of the untreated DNA to the sequence of the bisulfite treated DNA to infer the methylation state of all cytosines in the sequence by whether or not they were converted to uracil (thymine) following the bisulfite treatment (__see FIGURE__).

\subsubsection{Analysis platforms combined with bisulfite conversion}\label{sec:bs_platforms}

- [ ] Done

Initial experiments based on the sodium bisulfite pre-treatment of DNA used Sanger sequencing of cloned PCR products, a very laborious task, which restricted experiments to studying a limited number of short segments of DNA (__HOW MANY & LONG?__). Although subsequent enhancements in the automation of Sanger sequencing improved the throughput of these assays, it was never going to be able to deliver a cost-effective, genome-scale assay of DNA methylation. The development of hybridisation microarrays provided an alternative way of analysing bisulfite-treated DNA that could provide cheap, genome-wide measurements of DNA methylation(__SOURCE__).

Microarrays contain thousands, even millions, of short _oligonucleotide_ probes. Each probe is is designed to hybridise to a particular DNA sequence and emits a signal, such as flurousence, that can be measured to infer the strength of the hybridisation. Therefore, an (idealised) way to analyse DNA methylation with a microarray is to hybridise bisulfite-converted DNA to a microarray that contains probes for both the methylated and unmethylated versions of all sequences of interest. The relative methylation of each sequence can be inferred from the relative intensities of the "methylated probe" to the "unmethylated probe". Such an idealised experiment brushes over many complications including [@Laird:2010iv]:

* The reduced complexity of bisulfite-converted DNA (from a 4-base alphabet to a largely 3-base alphabet) leads to decreased hybridisation specificity
* Sequences containing multiple cytosines require multiple probe versions in order to assay all possible methylation patterns
* This approach requires the design of organism-specific microarrays

The Illumina 450k (__INSERT FULL NAME__) provides a modern implementation of the idea of bisulfite pre-treatment of DNA followed by microarray hybridisation for studying human samples. This array assays __HOW MANY__ CpGs (__IN WHAT TYPE OF REGIONS__) and uses two different probe designs (__WHY?__). There are still substantial bioinformatics challenges in analysing data from these assays (__CITE: SNPs in probes, normalisation, etc.__).

As mentioned in __SECTION__, genomics research was revolutionised by the development of cheap high-throughput sequencing technology and the study of DNA methylation was no exception. In 2008, two papers were published describing methods for whole-genome shotgun sequencing of bisulfite-converted DNA using the nascent Illumina sequencing technology [@Cokus:2008fc, Lister:2008bh]. [@Cokus:2008fc] termed their approach _BS-seq_ while [@Lister:2008bh] called theirs _methylC-seq_. These techniques, along with methods that these with targeting of particular regions of the genome, are described further in __SECTION__.

One final approach to mention is the use of mass spectrometry to analyse bisulfite-treated DNA, as implemented in Sequenom's EpiTYPER (__SOURCE__). This platform can provide quantitative measurements of CpG methylation across hundreds of loci and multiple samples and is frequently used to validate findings discovered using other platforms (__SOURCE__).

\subsection{DNA kinetics assays}\label{sec:dna_kinetics_assays}

- [ ] Done

\subsection{Pros and cons of bisulfite-conversion assays}\label{sec:pros_and_cons_of_bs_assays
}

- [ ] Done

Bisulfite-conversion assays are considered the "gold standard" for studying DNA methylation. Currently, multiple cells are required as input although the minimum number required is being lowered and efforts are well produce single-cell, whole-genome DNA methylation maps using bisulfite-treated DNA (__SOURCE__).

A key advantage of bisulfite-conversion assays over enzyme digest and affinity enrichment assays is that it provide base pair resolution of DNA methylation. In fact, single-molecule, base pair resolution is even possible for short DNA sequences when bisulfite-treated DNA is analysed with sequencing[^microarrays].

[^microarrays]: Microarray hybridisation can provide base pair resolution but not single molecule resolution. The signal from a microarray-based experiment a sample-wide average since, for a given locus, the signal is relative to the proportion of DNA fragments in the sample methylated at that locus.

A recently discovered disadvantage is that bisulfite-conversion assays are unable to distinguish between 5mC and 5hmC because both modifications similarly protect cytosine from deamination to uracil (__SEE [@Yu:2012ej] for references__). In effect, the detection of 5mC, and all subsequent inference, is confounded with that of 5hmC. For most experiments this isn't much of a problem - most cells have very low levels of 5hmC and so there is little confounding - however, in certain experiments this needs a more careful approach. To address this issue, [@Booth:2012fh] developed _oxidative bisulfite-sequencing_ (oxBS-seq) and [@Yu:2012ej] developed _Tet-assisted bisulfite sequencing_ (TAB-seq) to separate 5mC and 5hmC detection.

oxBS-seq specifically measures 5mC. The input DNA is oxidated by potassium perruthenate (KRuO$_{4}$), which converts 5hmC to 5fC, prior to bisulfite-treatment. Only 5mC is protected from conversion during the bisulfite-treatment, which effectively means that only 5mC remains to be detected at the sequencing stage. The level of 5hmC can be estimated by performing traditional bisulfite-sequencing and then "subtracting" the oxBS-seq signal (5mC) from the bisulfite-sequencing signal (5mC + 5hmC).

TAB-seq takes the opposite approach to oxBS-seq by specifically measuring 5hmC. The input DNA is treated with a $\beta$-glucosyltransferase, which converts 5hmC to $\beta$-glucosyl-5-hydroxymethylcytosine (5gmC), followed by TET oxidation. Only 5gmC is protected from TET oxidation, which effectively means that only 5hmC remains to be detected at the sequencing stage. The level of 5mC can be estimated by performing traditional bisulfite-sequencing and then "subtracting" the TAB-seq signal (5hmC) from the bisulfite-sequencing signal (5mC + 5hmC).

A disadvantage of bisulfite-conversion assays is that they require knowledge of the underyling DNA sequence in order to infer the methylation states of cytosines. This requires either a parallel experiment to sequence the target region(s) or reliance on a reference genome. When relying on a reference genome, the inference of the methylation state can be confounded by the DNA sequence of the sample (__SOURCE__). __FIGURE__ describes such an example.

The bisulfite-treatment of DNA can introduce biases and other problems. Three examples are _PCR-bias_, _incomplete bisulfite conversion_ and _DNA degradation_. _PCR-bias_ is the difference in amplification efficiency of methylated and unmethylated versions of the same DNA sequence [@Warnecke:1997eh]. _Incomplete bisulfite-conversion_ leads to cytosines being incorrectly inferred as 5mC (__SOURCE__). Finally, bisulfite-treatment can lead to _DNA degredation_, which, as the name suggets, leads to a loss in DNA quality or quantity and can make it particularly hard to sequence long DNA molecules (__SOURCE: Anecdotal__).

\section{Whole-genome bisulfite-sequencing}\label{sec:wgbs}

- [ ] DONE

\subsection{Targeted BS-seq, e.g. RRBS, NimbleGen and Agilent's capture kits, MRE-seq}\label{sec:targeted_bsseq}

- [ ] Done

* "RRBS’s ability to interrogate a locus is dependent on its MspI-cut-site (CCGG) density and consequently measures 10%–15% of the CpGs in the human genome (Bock et al. 2010; Harris et al. 2010)" \cite{Stevens:2013hv}
* See \cite{Stevens:2013hv} for MRE-seq details and a good summary of DNA methylation assays



## General TODOs
* Nucleotide, residue, (nucleo)base; what's the difference and which term should I use?
	* A nucleotide is composed of a nucleobase (also termed a nitrogenous base), a five-carbon sugar (either ribose or 2-deoxyribose), and one or more phosphate groups \[Coghill, Anne M.; Garson, Lorrin R., ed. (2006). The ACS style guide: effective communication of scientific information (3rd ed.). Washington, D.C.: American Chemical Society. p. 244. ISBN 978-0-8412-3999-9.)\]
* Should I give more detail on Rafa's CGI HMM?
* Define eukaryotes, prokaryotes and viruses
* Hyphenation or not of "bisulfite-sequencing"?
  - Yes, hyphenate.
* General term for methylC-seq, BS-seq, etc.
  - "whole-genome bisulfite-sequencing (WGBS)"
* Is it most accurate to say that restriction enzymes "cleave", "cut" or "digest" DNA?
* Define/describe DNA sequencing and the genomics revolution in the section describing DNA (or reference genome section?)
	* Define "library"
* Fix up reference formating
* Note that when I refer to a read I mean either a single-end read or a paired-end read. I will use to read_1 and read_2 if I want to refer a particular read that makes up a paired-end read.
* \cite{Zhang:2013uu} credit Laurent et al. (2010) with BS-seq and Cokus with WGBS.
