---
author: "Peter Hickey"
output:
  pdf_document:
    fig_caption: yes
    includes:
      in_header: ../latex/header.tex
  html_document:
    keep_md: true
bibliography: ../latex/phd_thesis.bib
---

\chapter{Co-methylation}\label{chap:co-methylation}

\begin{chapabstract}
I describe two measures of within-sample co-methylation, _within-fragment co-methylation_ and _correlation of $\beta$-values_. The exploration of within-fragment co-methylation leads to a study of methods for estimating meaningful summaries of dependence in sparse $2 \times 2 \times K$ contingency tables.

Using estimates of both within-fragment co-methylation and correlation of $\beta$-values, I explore how co-methylation is affected by the distance between methylation loci and the genomic context of the methylation loci. I apply these methods to 40 whole-genome bisulfite-sequencing samples to study the between-sample variability of co-methylation.

Throughout this chapter I focus on co-methylation of CpGs, although the methods described are also applicable to non-CpG methylation loci.
\end{chapabstract}

\section{Correlations of $\beta$-values}\label{subsec:beta_cor}

Chapter \ref{chap:co-methylation_review} documented several previous analyses based on 'correlations' of aggregate methylation levels, such as $\beta$-values. What was lacking from these were clear descriptions of the proposed methods and software implementations. In this chapter I try to rectify the former and with the `MethylationTuples` I make available a software implementation of these methods.

\subsection{Methods}

Here I describe two simple methods based on correlations of $\beta$-values. Both of these are implemented in the `methLevelCor` function that is part of the `MethylationTuples` software.

The aim of this analysis is to address the question of 'how are aggregate methylation levels correlated as a function of the distance between loci?'. There are two steps in this analysis:

1. Define the pairs of methylation loci.
2. For each $IPD$, compute the correlation of $\beta$-values for all pairs of methylation loci separated by that $IPD$.

Step (2) may be further stratified by the genomic context of the pair of CpGs, such as whether they are in a CpG island, or other variables of interest.

The same analysis could be performed using $\mathcal{M}$-values rather than $\beta$-values, and this is available as an option in the `methLevelCor` function.

\subsubsection{Constructing the pairs}\label{sec:beta_pair_construction}

I consider two different strategies for creating pairs of CpGs. Both strategies are based on CpGs in the reference genome after having filtered out those reference-specific CpGs. That is, I begin with $\mathcal{I}^{ref}$ and, based on the `Bis-SNP` output, I filter out loci that are not CpGs in the sample's genome. Rather than actually remove these sites, I simply set the corresponding $\beta$-value to `NA`. I also set the $\beta$-values of 'missing' CpGs, those without sufficient sequencing coverage to call a $\beta$-value, to `NA` so that these are still included in the set of CpGs.

The first strategy creates all pairs of adjacent CpGs on the same chromosome and strand, which are then stratified by $IPD$ and any secondary variables, such as genomic context. For a chromosome with $N$ CpGs there are $(N - 1)$ pairs. I call this the 'adjacent pairs' or $NIL = 0$ strategy, since all pairs have no intervening loci ($NIL$ = number of intervening loci)^[nil0].

[^nil0]: Recall that this is with respect to the reference genome and not the sample genome.

The second strategy uses all pairs of CpGs on the same chromosome and strand, which are then stratified by $IPD$ and any secondary variables, such as genomic context. In practice, the second strategy only uses pairs from a set of $IPD$s, e.g., $IPD = 2 - 2000$, otherwise the number of pairs becomes unwieldy. This is more like a traditional autocorrelation analysis of $\bm{\beta} = (\beta_{1}, \ldots, \beta_{N_{loci}})$, albeit of a vector of unequally spaced observations and with care taken to only use pairs on the same chromosome and strand. I refer to this as the $NIL \geq 0$ strategy. Using the $NIL \geq 0$ strategy, two pairs of CpGs with an identical $IPD$ may have a very different number of intervening CpGs.

The interpretation of correlations computed using the $NIL \geq 0$ is complicated because the set of pairs are more heterogeneous than those under the $NIL = 0$ strategy. For this reason I prefer the $NIL = 0$ strategy, however, it is useful to consider and contrast the two strategies, which is what I have done in my thesis.

\subsubsection{Stratifying by a genomic feature}\label{sec:pair_stratification}

Even when using the $NIL = 0$ strategy, two pairs of CpGs with an identical $IPD$ may come from two regions of the genome that have very different methylation dynamics. In other words, the vector of $\beta$-values for a given sample is highly non-stationary --- the correlation between two loci depends on more than just the $IPD$. An obvious thing to investigate is how strong an influence the genomic context has on these correlations of $\beta$-values. To do this, we can stratify our analysis by a genomic feature. While this does not eliminate this non-stationarity, it may help identify factors that drive some of the heterogeneity.

For any genomic feature, a pair of loci may be inside, outside or spanning the boundary of the feature. Figure \ref{fig:feature_status} illustrates the 'feature status' of some pairs of loci. Note that a pair where each loci is in a different CpG island is declared to be inside the feature, i.e., elements need not be in the same feature but rather in the same type of feature. The genomic feature should partition the genome so that each loci is either inside or outside of the feature. Since the number of 'spanning pairs' is substantially fewer than those inside or outside of the feature, and the boundaries of the feature are oftentimes fuzzy, I have excluded these 'spanning pairs' in the results shown below.

\begin{figure}[!htb]
\centering
\includegraphics[width=0.8\textwidth]{../figures/feature_status.pdf}
\caption{Schematic illustrating the 'feature status' of four pairs of methylation loci. The circles represent loci and the grey bars a genomic feature, such as CpG islands. Each pair's 'feature status' is determined by whether both loci are outside of the feature (outside), one locus is inside and one outside of the feature (spanning), or both inside the feature (inside). Note that the rightmost pair is 'inside' even though the two loci that comprise the pair are in different elements of the genomic feature.}
\label{fig:feature_status}
\end{figure}

\subsubsection{Choice of correlation coefficient}

Three commonly used correlation coefficients are Pearson's $r$ \citep{Pearson:1895vt}, Spearman's $\rho$ \citep{Spearman:1904eg} and Kendall's $\tau$ \citep{KENDALL:1938bv}. Briefly, Pearson's product-moment correlation is designed to detect linear relationships between two variables. However, it is not robust to outliers, which motivates the use of Spearman's and Kendall's rank-based correlation coefficients. Spearman's $\rho$ is simply the Pearson product-moment correlation of the ranks of the data while Kendall's $\tau$ is based on the relative frequency of concordantly ranked pairs.

Owing to the size of whole-genome bisulfite-sequencing data, I have chosen to only compute Pearson's $r$ and Spearman's $\rho$, which are much faster to compute than Kendall's $\tau$. The `methLevelCor` function in the `MethylationTuples` will compute a confidence interval, $95\%$ by default, for the Pearson correlation. This feature is not yet available when using Spearman's or Kendall's correlation coefficient.

\subsubsection{Interpretation of correlation}\label{sec:interpretation_of_correlation}

Before computing correlations and trying to interpret these, it is a good idea to look at the data from which these correlations are computed. Figure \ref{fig:ADS_beta_scatterplots} shows kernel density smoothed scatterplots for pairs of $\beta$-values separated by $IPD = 2, 20, 200, 2000$ with $NIL = 0$ or $NIL \geq 0$ from the _ADS_ sample. It highlights two major problems.

Firstly, the bimodality of the distribution of $\beta$-values means that the points concentrate in the corners of the plot --- $(0, 0), (0, 1), (1, 0)$ and $(1, 1)$. A correlation coefficient is not well suited to summarising these distributions of points. Secondly, when using the $NIL = 0$ strategy, as the $IPD$ increases, the number of points in each scatterplot decreases and so the correlations are very unstable.

In summary, the correlations of pairs of $\beta$-values should be cautiously interpreted, bearing in mind that these are rather crude summaries of the underyling scatterplots that hide many details.

\begin{figure}[!htb]
\centering
\includegraphics[width=\textwidth]{../figures/ADS_beta_scatterplots.pdf}
\caption{Kernel density smoothed scatterplots for pairs of CpG $\beta$-values, $(\beta_{1}, \beta_{2})$, from the ADS sample. $\beta$-values are computed using strand-collapsed counts. $n$ is the number of pairs in each scatterplot.  The Spearman correlation of each scatterplot is also reported. Individual plots created using the \texttt{smoothScatter} function in R.}
\label{fig:ADS_beta_scatterplots}
\end{figure}

\FloatBarrier
\subsection{Results}

For all 40 samples in the EPISCOPE, Lister, Seisenberger and Ziller datasets, I compute the Pearson and Spearman correlations using both the $NIL = 0$ and $NIL \geq 0$ strategies. I compute these genome-wide and after stratifying CpG pairs by whether they are in a CpG island. These analyses are performed separately for each strand and also for strand-collapsed $\beta$-values.

\subsubsection{Can we collapse by strand?}

Figures \ref{fig:EPISCOPE_spearman_beta_cors_min_cov_5}, \ref{fig:Lister_spearman_beta_cors_min_cov_5}, \ref{fig:Seisenberger_spearman_beta_cors_min_cov_5} and \ref{fig:Ziller_spearman_beta_cors_min_cov_5} show the Spearman correlations of $\beta$-values for $NIL = 0$ pairs of CpGs stratified by strand. Since the correlations are very similar for the two strands, I decided to use the strand-collapsed $\beta$-values in all that follows.

\begin{figure}[!htb]
\centering
\includegraphics[width=\textwidth]{../figures/EPISCOPE_spearman_beta_cors_min_cov_5.pdf}
\caption{Correlations of $\beta$-values for strand-specific pairs of CpGs with $NIL = 0$ for samples from the EPISCOPE dataset.}
\label{fig:EPISCOPE_spearman_beta_cors_min_cov_5}
\end{figure}

\begin{figure}[!htb]
\centering
\includegraphics[width=\textwidth]{../figures/Lister_spearman_beta_cors_min_cov_5.pdf}
\caption{Correlations of $\beta$-values for strand-specific pairs of CpGs with $NIL = 0$ for samples from the Lister dataset.}
\label{fig:Lister_spearman_beta_cors_min_cov_5}
\end{figure}

\begin{figure}[!htb]
\centering
\includegraphics[width=\textwidth]{../figures/Seisenberger_spearman_beta_cors_min_cov_5.pdf}
\caption{Correlations of $\beta$-values for strand-specific pairs of CpGs with $NIL = 0$ for samples from the Seisenberger dataset.}
\label{fig:Seisenberger_spearman_beta_cors_min_cov_5}
\end{figure}

\begin{figure}[!htb]
\centering
\includegraphics[width=\textwidth]{../figures/Ziller_merged_spearman_beta_cors_min_cov_5.pdf}
\caption{Correlations of $\beta$-values for strand-specific pairs of CpGs with $NIL = 0$ for samples from the Ziller dataset.}
\label{fig:Ziller_spearman_beta_cors_min_cov_5}
\end{figure}

\FloatBarrier
\subsubsection{Choice of correlation coefficient}

Figures \ref{fig:EPISCOPE_pearson_vs_spearman_beta_cors_min_cov_10_strand_collapsed}, \ref{fig:Lister_pearson_vs_spearman_beta_cors_min_cov_10_strand_collapsed}, \ref{fig:Seisenberger_pearson_vs_spearman_beta_cors_min_cov_10_strand_collapsed} and \ref{fig:Ziller_pearson_vs_spearman_beta_cors_min_cov_10_strand_collapsed} compares the Pearson correlation coefficient to the Spearman correlation coefficient. The two correlation coefficient give qualitatively similar results, however, the Spearman correlation is consistently smaller in magnitude than the Pearson correlation.

I have elected to use Spearman's correlation coefficient in what follows because it is more robust to outliers than Pearson's correlation \citep{Kraemer:2006te}. However, this does not remove the general limitations of using correlation coefficients (discussed in Section \ref{sec:interpretation_of_correlation}).

\begin{figure}[!htb]
\centering
\includegraphics[width=\textwidth]{../figures/EPISCOPE_pearson_vs_spearman_beta_cors_min_cov_10_strand_collapsed.pdf}
\caption{Pearson correlation versus Spearman correlations of $\beta$-values for strand-collapsed pairs of CpGs with $NIL = 0$ for samples from the EPISCOPE dataset.}
\label{fig:EPISCOPE_pearson_vs_spearman_beta_cors_min_cov_10_strand_collapsed}
\end{figure}

\begin{figure}[!htb]
\centering
\includegraphics[width=\textwidth]{../figures/Lister_pearson_vs_spearman_beta_cors_min_cov_10_strand_collapsed.pdf}
\caption{Pearson correlation versus Spearman correlations of $\beta$-values for strand-collapsed pairs of CpGs with $NIL = 0$ for samples from the Lister dataset.}
\label{fig:Lister_pearson_vs_spearman_beta_cors_min_cov_10_strand_collapsed}
\end{figure}

\begin{figure}[!htb]
\centering
\includegraphics[width=\textwidth]{../figures/Seisenberger_pearson_vs_spearman_beta_cors_min_cov_10_strand_collapsed.pdf}
\caption{Pearson correlation versus Spearman correlations of $\beta$-values for strand-collapsed pairs of CpGs with $NIL = 0$ for samples from the Seisenberger dataset.}
\label{fig:Seisenberger_pearson_vs_spearman_beta_cors_min_cov_10_strand_collapsed}
\end{figure}

\begin{figure}[!htb]
\centering
\includegraphics[width=\textwidth]{../figures/Ziller_merged_pearson_vs_spearman_beta_cors_min_cov_10_strand_collapsed.pdf}
\caption{Pearson correlation versus Spearman correlations of $\beta$-values for strand-collapsed pairs of CpGs with $NIL = 0$ for samples from the Ziller dataset.}
\label{fig:Ziller_pearson_vs_spearman_beta_cors_min_cov_10_strand_collapsed}
\end{figure}

\FloatBarrier
\subsubsection{$NIL = 0$}

Figures \ref{fig:EPISCOPE_spearman_beta_cors_min_cov_10_strand_collapsed_CI}, \ref{fig:Lister_spearman_beta_cors_min_cov_10_strand_collapsed_CI}, \ref{fig:Seisenberger_spearman_beta_cors_min_cov_10_strand_collapsed_CI} and \ref{fig:Ziller_spearman_beta_cors_min_cov_10_strand_collapsed_CI} show the Spearman correlations of strand-collapsed $\beta$-values for $NIL = 0$ pairs of CpGs. Two things immediately stand out. Firstly, for all samples the trend is that the strength of the correlation decays as a function of $IPD$ --- pairs of CpGs separated larger distances are on average less correlated than pairs of CpGs separated by shorter distances. Secondly, the estimated correlations violently jump around for larger $IPD$s. The latter is simply due to the aforementioned lack of $NIL = 0$ pairs at larger $IPD$s with which to estimate these correlations. Therefore, it is the trend, rather than the individual values, that should be analysed in these plots.

The are some consistent patterns in comparing these trends across samples. The correlations very quickly drop well below $0.5$, within approximately $50$ bp for all the EPISCOPE and Seisenberger samples and most of the Lister and Ziller samples. Samples that are exceptions to this trend --- _ADS_, _ADS-adipose_, _ADS-iPSC_, _FF_, _IMR90\_r1_, _IMR90\_r2_, _IMR90\_cell\_line_ and _HepG2\_cell\_line_ --- have a slower decay in these correlations and remain more highly correlated over $IPD = 2, \ldots, 1500$. These same samples are also notable because they have significant amounts of intermediate methylation (Section \ref{sec:genome-wide_distribution_of_beta_values}). I believe this is what drives the stronger and slowly decaying correlations. The scatterplots of $\beta$-values for these samples will have more $\beta$-values in the middle of the scatterplot, and fewer in the corners, hence the stronger correlations.

Most samples have a correlation close to $1$ for neighbouring CpGs ($IPD = 2$). The notable exceptions are the embryonic stem cell replicates, _H1\_r1_ and _H1\_r2_ (Lister dataset), which begin with a maximum correlation of only around $0.5$. This is not observed in the other embryonic stem cells _H9_, _H9\_Laurent_, _HSF1_ and _J1\_1_), which makes it difficult to know whether $\beta$-values of embryonic stem cells are truly less correlated than other cell types.

The trend of the correlations for all these samples plateaus out close to zero, slightly higher for those samples with significant intermediate methylation. However, it should not be concluded from this that the methylation of CpGs separated by more than a few hundred base pairs are uncorrelated. Recall that these results are only for $NIL = 0$ pairs of CpGs, which are increasingly rare for larger $IPD$s. These results tell us about the 'first-order' correlations of $\beta$-values, which decay very quickly but whose importance also decreases quite quickly owing to the vast majority of CpGs being within $100$ bp of another CpG (e.g., Figure \ref{fig:IPD_ECDF_hg19} for humans).

One final aspect of these genome-wide plots bears mentioning. There are hints of an approximately $150-200$ bp periodicity in these plots. This is a similar scale to the previously reported periodicities in correlations of $\beta$-values (see Chapter \ref{chap:co-methylation_review}). While this periodicity can readily be detected by eye, it is somewhat more difficult to verify using traditional methods based on Fourier analysis of the signal. The chief reason is that the data are very noisy beyond the second or third putative cycle due to the limited data at these larger $IPD$s.

To address this problem, we might try smoothing the raw correlations, for example, by using LOESS \citep{Cleveland:1979fu}, and then looking for periodicities in the smoothed values. However, this is only helpful when the data are corrupted by a small amount of noise, certainly not the extreme noise we see for $IPD > 1000$ in most samples. Furthermore, smoothing introduces problems of its own, such as choosing the smoothing bandwidth in such a way as to not introduce or destroy real periodicities. Even if we can automatically choose these smoothing parameters in an unbiased manner, we are still stuck with only observing a handful of periods of this size from $NIL = 0$ pairs. So while these periodicities are 'obvious' in the plots, they are rather more difficult to detect without visualising the raw data.

\begin{figure}[!htb]
\centering
\includegraphics[width=\textwidth]{../figures/EPISCOPE_spearman_beta_cors_min_cov_10_strand_collapsed_CI.pdf}
\caption{Spearman correlations of $\beta$-values as a function of $IPD$ for strand-collapsed pairs of CpGs with $NIL = 0$ for samples from the EPISCOPE dataset.}
\label{fig:EPISCOPE_spearman_beta_cors_min_cov_10_strand_collapsed_CI}
\end{figure}

\begin{figure}[!htb]
\centering
\includegraphics[width=\textwidth]{../figures/Lister_spearman_beta_cors_min_cov_10_strand_collapsed_CI.pdf}
\caption{Spearman correlations of $\beta$-values as a function of $IPD$ for strand-collapsed pairs of CpGs with $NIL = 0$ for samples from the Lister dataset.}
\label{fig:Lister_spearman_beta_cors_min_cov_10_strand_collapsed_CI}
\end{figure}

\begin{figure}[!htb]
\centering
\includegraphics[width=\textwidth]{../figures/Seisenberger_spearman_beta_cors_min_cov_10_strand_collapsed_CI.pdf}
\caption{Spearman correlations of $\beta$-values as a function of $IPD$ for strand-collapsed pairs of CpGs with $NIL = 0$ for samples from the Seisenberger dataset.}
\label{fig:Seisenberger_spearman_beta_cors_min_cov_10_strand_collapsed_CI}
\end{figure}

\begin{figure}[!htb]
\centering
\includegraphics[width=\textwidth]{../figures/Ziller_merged_spearman_beta_cors_min_cov_10_strand_collapsed_CI.pdf}
\caption{Spearman correlation of $\beta$-values as a function of $IPD$ for strand-collapsed pairs of CpGs with $NIL = 0$ for samples from the Ziller dataset.}
\label{fig:Ziller_spearman_beta_cors_min_cov_10_strand_collapsed_CI}
\end{figure}

The above are genome-wide plots, all pairs of CpGs with a given $IPD$ are treated identically. However, there are good reasons to expect that the correlations of $\beta$-values, like the $\beta$-values themselves, might be influenced by their genomic context. To investigate this, we can stratify pairs of CpGs pairs by a genomic feature and repeat the analysis. The obvious feature to stratify on is CpG islands.

Figures \ref{fig:EPISCOPE_spearman_beta_cors_min_cov_10_strand_collapsed_CI}, \ref{fig:Lister_spearman_beta_cors_min_cov_10_strand_collapsed_CI}, \ref{fig:Seisenberger_spearman_beta_cors_min_cov_10_strand_collapsed_CI} and \ref{fig:Ziller_spearman_beta_cors_min_cov_10_strand_collapsed_CI} are plots of Spearman correlations of $\beta$-values for pairs of CpGs stratified by whether the pair is inside or outside of a CpG island[^spanning]. Again, the substantial noise in these plots is due to a lack of $NIL = 0$ pairs with large $IPD$s, particularly inside CpG islands.

The message from these plots is clear, $\beta$-values of $NIL = 0$ pairs within CpG islands are much more correlated than those pairs outside of islands. Stratifying by CpG island status also reveals that the correlation of the methylation levels of CpGs inside islands is less influenced by $IPD$. In fact, for some samples it appears that the most correlated $NIL = 0$ pairs inside islands are not those with the minimum $IPD = 2$, but those with a slightly larger $IPD$. From these plots we can now see that the apparent sharp decline in correlations over the first $50$ bp from the genome-wide data is really driven by a shift from pairs that are frequently inside CpG islands to pairs that are mostly outside of islands.

[^spanning]: Pairs spanning the boundary are excluded. There are few of these pairs.

\begin{figure}[!htb]
\centering
\includegraphics[width=\textwidth]{../figures/EPISCOPE_spearman_beta_cors_min_cov_10_strand_collapsed_CGI.pdf}
\caption{Spearman correlations of $\beta$-values as a function of $IPD$ stratified by CpG island status for strand-collapsed pairs of CpGs with $NIL = 0$ for samples from the EPISCOPE dataset.}
\label{fig:EPISCOPE_spearman_beta_cors_min_cov_10_strand_collapsed_CGI}
\end{figure}

\begin{figure}[!htb]
\centering
\includegraphics[width=\textwidth]{../figures/Lister_spearman_beta_cors_min_cov_10_strand_collapsed_CGI.pdf}
\caption{Spearman correlations of $\beta$-values as a function of $IPD$ stratified by CpG island status for strand-collapsed pairs of CpGs with $NIL = 0$ for samples from the Lister dataset.}
\label{fig:Lister_spearman_beta_cors_min_cov_10_strand_collapsed_CGI}
\end{figure}

\begin{figure}[!htb]
\centering
\includegraphics[width=\textwidth]{../figures/Seisenberger_spearman_beta_cors_min_cov_10_strand_collapsed_CGI.pdf}
\caption{Spearman correlations of $\beta$-values as a function of $IPD$ stratified by CpG island status for strand-collapsed pairs of CpGs with $NIL = 0$ for samples from the Seisenberger dataset.}
\label{fig:Seisenberger_spearman_beta_cors_min_cov_10_strand_collapsed_CGI}
\end{figure}

\begin{figure}[!htb]
\centering
\includegraphics[width=\textwidth]{../figures/Ziller_merged_spearman_beta_cors_min_cov_10_strand_collapsed_CGI.pdf}
\caption{Spearman correlations of $\beta$-values as a function of $IPD$ stratified by CpG island status for strand-collapsed pairs of CpGs with $NIL = 0$ for samples from the Ziller dataset.}
\label{fig:Ziller_spearman_beta_cors_min_cov_10_strand_collapsed_CGI}
\end{figure}

\FloatBarrier
\subsubsection{$NIL \geq 0$}\label{sec:nil_geq_0_beta_cors}

Figures \ref{fig:EPISCOPE_spearman_beta_cors_all_min_cov_10_strand_collapsed_CI}, \ref{fig:Lister_spearman_beta_cors_all_min_cov_10_strand_collapsed_CI}, \ref{fig:Seisenberger_spearman_beta_cors_all_min_cov_10_strand_collapsed_CI} and \ref{fig:Ziller_spearman_beta_cors_all_min_cov_10_strand_collapsed_CI} are the plots of CpG $\beta$-value correlations against $IPD$ for $NIL \geq 0$ pairs. In contrast to the $NIL = 0$ results, the correlations here are relatively stable, at least for $IPD = 2, \ldots, 2000$[^ipd_choice]. This is because there are many more $NIL \geq 0$ pairs than there are $NIL = 0$ pairs.

[^ipd_choice]: These correlations could be computed for $IPD > 2000$. However, constructing all $NIL \geq 0$ pairs requires much more time and memory than constructing all $NIL = 0$ pairs because there are many more such pairs. For this reason, and for compatability with the $NIL = 0$ results, I have focused on $IPD = 2, \ldots, 2000$ for the $NIL \geq 0$ pairs.

When we aggregate over as many factors as we are in these genome-wide $NIL \geq 0$ plots, we find that there is little difference in the correlation curves between samples. All samples have a very high correlation, close to $1$, for small $IPD$s that steadily decays to a lower correlation of $0.2-0.4$ by $IPD = 2000$.

\begin{figure}[!htb]
\centering
\includegraphics[width=\textwidth]{../figures/EPISCOPE_spearman_beta_cors_all_min_cov_10_strand_collapsed_CI.pdf}
\caption{Spearman correlation of $\beta$-values for strand-collapsed pairs of CpGs with $NIL \geq 0$ for samples from the EPISCOPE dataset.}
\label{fig:EPISCOPE_spearman_beta_cors_all_min_cov_10_strand_collapsed_CI}
\end{figure}

\begin{figure}[!htb]
\centering
\includegraphics[width=\textwidth]{../figures/Lister_spearman_beta_cors_all_min_cov_10_strand_collapsed_CI.pdf}
\caption{Spearman correlation of $\beta$-values for strand-collapsed pairs of CpGs with $NIL \geq 0$ for samples from the Lister dataset.}
\label{fig:Lister_spearman_beta_cors_all_min_cov_10_strand_collapsed_CI}
\end{figure}

\begin{figure}[!htb]
\centering
\includegraphics[width=\textwidth]{../figures/Seisenberger_spearman_beta_cors_all_min_cov_10_strand_collapsed_CI.pdf}
\caption{Spearman correlation of $\beta$-values for strand-collapsed pairs of CpGs with $NIL \geq 0$ for samples from the Seisenberger dataset.}
\label{fig:Seisenberger_spearman_beta_cors_all_min_cov_10_strand_collapsed_CI}
\end{figure}

\begin{figure}[!htb]
\centering
\includegraphics[width=\textwidth]{../figures/Ziller_merged_spearman_beta_cors_all_min_cov_10_strand_collapsed_CI.pdf}
\caption{Spearman correlation of $\beta$-values for strand-collapsed pairs of CpGs with $NIL \geq 0$ for samples from the Ziller dataset.}
\label{fig:Ziller_spearman_beta_cors_all_min_cov_10_strand_collapsed_CI}
\end{figure}

Comparing CpG islands to non-islands, Figures \ref{fig:EPISCOPE_spearman_beta_cors_all_min_cov_10_strand_collapsed_CGI}, \ref{fig:Lister_spearman_beta_cors_all_min_cov_10_strand_collapsed_CGI}, \ref{fig:Seisenberger_spearman_beta_cors_all_min_cov_10_strand_collapsed_CGI} and \ref{fig:Ziller_spearman_beta_cors_all_min_cov_10_strand_collapsed_CGI}, we again see that the methylation levels of pairs of CpGs within islands are consistently more correlated than those outside of islands[^hepg2].

[^hepg2]: The minor exception is the _HepG2\_cell\_line_ where the CpG island and non-island lines intersect by $IPD = 1500$, but we have already seen that this sample has an unusual methylome compared to the other samples.

\begin{figure}[!htb]
\centering
\includegraphics[width=\textwidth]{../figures/EPISCOPE_spearman_beta_cors_all_min_cov_10_strand_collapsed_CGI.pdf}
\caption{Spearman correlations of $\beta$-values as a function of $IPD$ stratified by CpG island status for strand-collapsed pairs of CpGs with $NIL \geq 0$ for samples from the EPISCOPE dataset.}
\label{fig:EPISCOPE_spearman_beta_cors_all_min_cov_10_strand_collapsed_CGI}
\end{figure}

\begin{figure}[!htb]
\centering
\includegraphics[width=\textwidth]{../figures/Lister_spearman_beta_cors_all_min_cov_10_strand_collapsed_CGI.pdf}
\caption{Spearman correlations of $\beta$-values as a function of $IPD$ stratified by CpG island status for strand-collapsed pairs of CpGs with $NIL \geq 0$ for samples from the Lister dataset.}
\label{fig:Lister_spearman_beta_cors_all_min_cov_10_strand_collapsed_CGI}
\end{figure}

\begin{figure}[!htb]
\centering
\includegraphics[width=\textwidth]{../figures/Seisenberger_spearman_beta_cors_all_min_cov_10_strand_collapsed_CGI.pdf}
\caption{Spearman correlations of $\beta$-values as a function of $IPD$ stratified by CpG island status for strand-collapsed pairs of CpGs with $NIL \geq 0$ for samples from the Seisenberger dataset.}
\label{fig:Seisenberger_spearman_beta_cors_all_min_cov_10_strand_collapsed_CGI}
\end{figure}

\begin{figure}[!htb]
\centering
\includegraphics[width=\textwidth]{../figures/Ziller_merged_spearman_beta_cors_all_min_cov_10_strand_collapsed_CGI.pdf}
\caption{Spearman correlations of $\beta$-values as a function of $IPD$ stratified by CpG island status for strand-collapsed pairs of CpGs with $NIL \geq 0$ for samples from the Ziller dataset.}
\label{fig:Ziller_spearman_beta_cors_all_min_cov_10_strand_collapsed_CGI}
\end{figure}

The $150-200$ bp periodicity is also evident from the $NIL \geq 0$ data and is particularly clear for the CpG pairs outside of CpG islands. This may be due to  different regulation of DNA methylation outside of CpG islands or simply because CpG islands have such consistently high correlations that these periodicities are not evident.

\subsection{Summary}

__TODO: Summarise the results for correlation of $\beta$-values.__

A limitation to analysing co-methylation using correlations of $\beta$-values is that the analysis is based on correlations of measurements that are aggregates from many thousands of DNA molecules. We are not measuring co-methylation on the scale at which the biological process of DNA methylation acts, namely at the level of individual DNA molecules. As noted by \citet{Ball:2009kh}, "the clonal feature of Illumina sequencing [allows us] to investigate whether co-methylation occurs at the single-molecule level". Before doing so, I first explore the challenge of estimating meaningful measures of dependence in sparse $2 \times 2 \times K$ contingency tables. While perhaps not yet apparent, this will be essential for the analysis of within-fragment co-methylation.

\FloatBarrier

\section{Estimating dependence in sparse $2 \times 2 \times K$ contingency tables}

We will begin with an overview of the simplest case of estimating dependence in $2 \times 2$ contingency tables, i.e., $K = 1$. We will then progressively introduce the complications of sparsity, $K > 1$, and heterogeneity across the $K$ tables. We conclude with a simulation study highlighting the performance of several difference estimators and determining which is most appropriate for a study of within-fragment co-methylation.

\subsection{$2 \times 2$ contingency tables}

Consider two binary random variables, $X$ and $Y$. Let $\pi_{ij}, i, j = 1, 2$ denote the probability that a subject has response $i$ for variable $X$ and response $j$ for variable $Y$. We wish to estimate the _odds ratio_, which is defined as $\psi = \frac{\pi_{11} \pi_{22}}{\pi_{12} \pi_{21}}$. There are many reasons why we might wish to estimate the odds ratio (__CITE__), but our interest here is __WHAT__. Provided that all $\pi_{ij} > 0$, the odds ratio, $\psi$, can be interpreted as follows:

- $\psi = 1$: $X$ is independent of $Y$.
- $\psi > 1$: Subjects are more likely to have the same level for both $X$ and $Y$, e.g., subjects with $(X, Y) = (1, 1)$ or $(X, Y) = (2, 2)$ are more likely than subjects with $(X, Y) = (1, 2)$ or $(X, Y) = (2, 1)$.
- $\psi < 1$: Subjects are more likely to have the the opposite level for $X$ than they do for $Y$, e.g., subjects with $(X, Y) = (1, 2)$ or $(X, Y) = (2, 1)$ are more likely than subjects with $(X, Y) = (1, 1)$ or $(X, Y) = (2, 2)$.

The possible values of $\psi$ are highly skewed, with $0 < \psi < 1$ and $1 < \psi < \infty$ corresponding to the two distinct dependence scenarios. Rather than estimating $\psi$ directly, we will instead focus our attention on the _log odds ratio_, $\phi = \log{\psi}$. The log odds ratio is symmetric about 0, which makes statistical inference somewhat simpler. I will routinely switch between discussing the estimation of $\psi$ and $\phi$, noting that we can always convert one to the other by appropriate exponentiation or taking of logarithms.

Suppose we observe $(X, Y)$ for $n$ samples. Denote by $n_{ij}, i, j = 1, 2$ the number of subjects with response $i$ for variable $X$ and response $j$ for variable $Y$. We use the 'dot' notation to denote the sum over the index it replaces, e.g., $n_{1 \cdot} = n_{11} + n_{12}$. The general form of this $2 \times 2$ contingency table is shown in Table \ref{tab:2x2}.

\begin{table}[h]
\centering
\begin{tabular}{@{}cccc@{}}
\cmidrule(l){2-4}
 & \multicolumn{3}{c}{Column} \\
 \cmidrule(l){2-4}
\multirow{3}{*}{Row} & & & \\
 & $n_{11}$       & $n_{12}$       & $n_{1 \cdot}$ \\
 & $n_{21}$       & $n_{22}$       & $n_{2 \cdot}$ \\
 & $n_{\cdot 1}$  & $n_{\cdot 2}$  & $n_{\cdot \cdot} = n$ \\ \cmidrule(l){2-4}
\end{tabular}
\caption{Notation for a $2 \times 2$ contingency table.}
\label{tab:2x2}
\end{table}

The simplest estimator of $\phi$ is the unconditional maximum likelihood estimator, $\widehat{\phi_{U}} = \log{\frac{n_{11} n_{22}}{n_{12} n_{21}}}$. The asymptotic standard error of $\widehat{\phi_{U}}$ is given by $\hat{\sigma}(\widehat{\phi_{U}}) = \sqrt{\frac{1}{n_{11}} + \frac{1}{n_{12}} + \frac{1}{n_{21}} + \frac{1}{n_{22}}}$. The asymptotic distribution of $\widehat{\phi_{U}}$ as $n \to \infty$ is $Normal(\phi, \hat{\sigma}(\phi)^{2})$.

A problem with the estimator $\widehat{\phi_{U}}$ is that it is $0$ or $\infty$ if any $n_{ij} = 0$ and undefined if either the row or column sums are zero, events that have positive probabilities. We can avoid such problems by adding $\frac{1}{2}$ to each of the $n_{ij}$ and defining a modified estimator $\widehat{\phi_{0.5}} = \log{\frac{(n_{11} + 0.5) (n_{22} + 0.5)}{(n_{12} + 0.5) (n_{21} + 0.5)}}$ with corresponding asymptotic standard error $hat{\sigma}(\widehat{\phi_{0.5}}) = \sqrt{\frac{1}{n_{11}} + \frac{1}{n_{12}} + \frac{1}{n_{21}} + \frac{1}{n_{22}}}$. \citet{Haldane:1956wq} and \citet{Anscombe:1956te} showed that $\widehat{\phi_{U}}$ and $\widehat{\phi_{0.5}}$ have the same asymptotic distribution around $\phi$ as $n \to \infty$ but that $\widehat{\phi_{0.5}}$ has reduced first-order bias. The modified estimator, $\widehat{\phi_{0.5}}$, is therefore generally recommended.

Another alternative to $\widehat{\phi_{U}}$ is the _conditional_ maximum likelihood estimator, $\widehat{\phi_{C}}$. This is computed by first conditioning on ${n_{1 \cdot}, n_{\cdot 1}}$. Some algebra then leads to the conditional distribution of $n_{11}$, $f(n_{11}; n_{1 \cdot}, n_{\cdot, 1, \psi})$, which is a hypergeometric distribution \citep{Fisher:1935ug}. The maximum likelihood estimator of $\psi$, $\widehat{psi_{C}}$ is solved using iterative methods.

The conditional estimator, $\widehat{\phi_{C}}$, works better than the unconditional estimator, $\widehat{\phi_{U}}$, when the sample size, $n$, is small \citep{Agresti:2007tf}. Estimation of the odds ratio is made more difficult when the contingency table is _sparse_. We say that a contingency table is sparse when some of the $n_{ij}$ are small. Sparsity can occur even when the sample size, $n$, is large. Sparsity becomes more of a problem as we move from a single $2 \times 2$ table to $K$ $2 \times 2$ tables or, equivalently, a $2 \times 2 \times K$ table.

\subsection{$2 \times 2 \times K$ contingency tables}

Suppose we now measure the same two binary variables $X$ and $Y$ in $K$ different strata. We summarise these data in a $2 \times 2 \times K$ contingency table, $\{n_{ijk} \}$, where the $k^{th}$ 'slice' of the table corresponds to the $2 \times 2$ table for the $k^{th}$ stratum. For now, we assume that the odds ratio, $\psi = \frac{\pi_{11k} \pi_{22k}}{\pi_{12k} \pi_{21k}}$, remains constant across the $K$ tables but we allow the marginal probabilities, $\pi_{\cdot 1k} = 1 - \pi_{\cdot 2k}, \pi_{1 \cdot k} = 1 - \pi_{2 \cdot k}$ to vary.

We can again compute an unconditional maximum likelihood estimator, $\widehat{phi_{U}}$, a '0.5-adjusted' unconditional maximum likelihood estimator, $\widehat{phi_{0.5}}$, and a conditional maximum likelihood estimator, $\widehat{phi_{C}}$, of the log odds ratio $\phi = \log{\psi}$ \citep[see][]{Breslow:1981ta}.

Another estimator of the common odds ratio, $\psi$, is the Mantel-Haenszel estimator \citep{MANTEL:1959uf}, $\widehat{\psi_{MH}} = \frac{\sum_{k} (n_{11k} n_{22k} / n_{\cdot \cdot k})}{n_{12k} n_{21k} / n_{\cdot \cdot k}}$. \citet{Robins:1986vh} derived a simple robust estimator of the variance for $\widehat{\phi_{MH}} = \log{\widehat{\psi_{MH}}}$, $\hat{\sigma}^{2}(\widehat{\phi_{MH}})$. This variance estimator did not appear until some 25 years after the initial publication of the Mantel-Haenszel estimator[^mh_variance]. This delay was in part due to the two asymptotic forms of $2 \times 2 \times K$ contingency tables.

[^mh_estimator]: Other estimators had been proposed but were supplanted by the work of \citet{Robins:1986vh}.

The first asymptotic environment, $A1$, assumes that the number of $2 \times 2$ tables, $K$, remains small while the the sample sizes in each strata, $n_{\cdot \cdot k}$, grows large. The second asymptotic environment, $A2$, assumes that $K$ grows while $n_{\cdot \cdot k}$ remains small. The variance estimator proposed by \citet{Robins:1986vh} is consistent under both $A1$ and $A2$. For our study of within-fragment co-methylation, $A2$ is the more relevant asymptotic environment.

\citet{Breslow:1981ta} showed that under $A2$ neither the unconditional maximum likelihood estimator, $\widehat{\psi_{U}}$, nor the '0.5-adjusted' unconditional maximum likelihood estimator, $\widehat{\psi_{0.5}}$, converge to the true odds ratio. In contrast, both the Mantel-Haenszel estimator, $\widehat{\psi_{MH}}$, and the conditional maximum likelihood estimator, $\widehat{\psi_{C}}$, do convert to $\psi$, with $\widehat{\psi_{MH}}$ having the advantage of a simple closer form expression \citep{Breslow:1996uc}.

\subsection{Homogeneity of odds ratios assumption}

So far we have assumed a common odds ratio for the $K$ $2 \times 2$ tables.

__UP TO HERE__


- Move to $2 \times 2 \times K$ tables
  - Go through same comparison of estimators
- Move to sparseness
  - Asymptopia I and II
  - Go through same comparison of estimators
  - The homogeneity assumption

 ## Setting for studying within-fragment co-methylation

 - Asymptopia II with sparse tables
 - Homogeneity??? Don't/can't know.

 ## Simulation

-

## Summary

\section{Within-fragment co-methylation}\label{subsec:wf_cometh}

The key advantage of studying within-fragment co-methylation is that it measures the dependence of DNA methylation at the scale on which the biological process acts. I have developed a method based on odds-ratios to quantify the strength of co-methylation for a pair of methylation loci in a single sample.

\subsection{Methods}

Here I describe a simple method based on analysing methylation patterns at 2-tuples. For simplicity, I will use CpG 2-tuples, but the method is applicable to any methylation type. This method is implemented in the `cometh()` function that is part of the `MethylationTuples` software.

The aim of this analysis is to address the questions:

1. How dependent are methylation states at nearby methylation loci on the same DNA fragment?
2. What influences this dependence?

There are two steps in this analysis:

1. Define the 2-tuples.
2. For each 2-tuple, estimates the dependence between the methylation state at the first locus and the second locus. For this I use the logarithm of the odds ratio ($LOR$).

Step (2) may be further stratified by the genomic context of the pair of CpGs, such as whether they are in a CpG island, or other variables of interest. This is done in the same manner as for pairs of $\beta$-values (Section \ref{sec:pair_stratification}).

A similar analysis could be performed using a different measure of dependence, which I discuss in Section \ref{sec:choice_of_dependence_measure}.

\subsubsection{Constructing the 2-tuples}

The construction of 2-tuples is very similar to constructing the pairs of methylation levels, described in Section \ref{sec:beta_pair_construction}. However, we are now limited to analysing co-methylation of loci that can be captured within a single read. With current technology this means we are limited to studying within-fragment co-methylation for 2-tuples with $IPD \lessapprox 250$.

Again, we can employ either the $NIL = 0$ or the $NIL \geq 0$ strategy. Both strategies start with CpGs in the reference genome that, for each sample, are then filtered to remove reference-specific loci. Rather than actually remove these loci, I simply set the corresponding counts of the methylation patterns to `NA`.

As in Section \ref{subsec:beta_cor} I consider the $NIL = 0$ and $NIL \geq 0$ strategies for constructing pairs. The $NIL \geq 0$ strategy corresponds to using the `--all-combination` flag in `methtuple`. Again, the $NIL = 0$ strategy is simpler, both conceptually and computationally, and so I prefer it. For example, if we have a pair with $NIL = 3$ then we need to take care to ensure that this $LOR$ is due to a genuine dependence between these two CpGs above and beyond the pairwise dependencies $(CpG_{1}, CpG_{2}), \ldots, (CpG_{4}, CpG_{5})$.

\FloatBarrier
\subsubsection{Choice of dependence measure}\label{sec:choice_of_dependence_measure}

As discussed, I use the logarithm of the odds ratio to summarise within-fragment co-methylation. Here I explain my reasons for choosing this measure of association.

Recall that for each 2-tuple and sample we can summarise the methylation patterns by a $2 \times 2$ contingency table, like that shown in Table \ref{tab:2x2}. The within-fragment co-methylation of the pair of CpGs, $(i, i')$, in the $j^{th}$ sample is then defined as the logarithm of the odds ratio of this table, $LOR_{(i, i'), j} = \log_{2} \Big (\frac{mm_{i, i'} \times uu_{i, i'}}{mu_{i, i'} \times um_{i, i'}} \Big)$. I use base-2 logarithms due to their widespread use and familiarity within biology. In the `cometh` function I also add a small offset (default `offset = 1`) to avoid problems arising from zero counts, which frequently occur. For the example in Table \ref{tab:2x2} with `offset = 1` we have $LOR = \log_{2} \Big (\frac{(1 + 1) \times (4 + 1)}{(0 + 1) \times (4 + 1)} \Big) = 1.74$.

From a Bayesian perspective, the offset may be interpreted as a prior count from a table with no association, i.e., a table with $LOR = 0$. The offset then has the effect of shrinking the raw estimate of the $LOR$, with more shrinkage applied for larger values of `offset`. Analyses using different offsets should be compared with caution.

\begin{table}[h]
\centering
\begin{tabular}{lcccl}
                           & \multicolumn{1}{l}{} & \multicolumn{2}{l}{Second CpG} &                \\
                           &                      & \textbf{M}     & \textbf{U}    & \textbf{Total} \\ \cline{2-5}
\multirow{2}{*}{First CpG} & \textbf{M}           & 1              & 2             & 3              \\
                           & \textbf{U}           & 0              & 4             & 4              \\ \cline{2-5}
                           & \textbf{Total}       & 1              & 6             & 7
\end{tabular}
\caption{Example $2 \times 2$ contingency table summarising methylation patterns at a 2-tuple.}
\label{tab:2x2}
\end{table}

A positive $LOR$ is evidence that the methylation states of two CpGs on the same DNA fragment are likely to be the identical, a negative $LOR$ is evidence that the methylation states of two CpGs on the same DNA fragment are likely to be the different, and a value close to zero is evidence that the methylation states of two CpGs on the same DNA fragment are independent of one another. More precisely, we may wish to compute a confidence interval for the $LOR$ or perform a hypothesis test to determine whether the observed $LOR$ is evidence for co-methylation at this pair of CpGs.

There are several alternative measures of association for a $2 \times 2$ table of binary variables. An interesting review of these is given by \citet{Kraemer:2006te}. The Pearson correlation coefficient, also known as the phi coefficient when applied to binary $2 \times 2$ tables, was used by \citet{Landan:2012kp} to analyse within-fragment co-methylation. There are a few reasons that I prefer $LOR$ to Pearson's correlation.

Firstly, the ``possible range of values of the [log] OR is not influenced by the marginal distributions of the variables involved'' \citep{Kraemer:2004el}. Another way of putting this is that given an odds ratio and a set of marginal probabilities, there exists one and only one $2 \times 2$ table satisfying these constraints \citep{plackett1981analysis}. In contrast, Pearson's correlation of a $2 \times 2$ is restricted by the marginal probabilities \citep{kraemer1986measure}, which I find undesirable.

Secondly, the empirical distribution of $LOR$ for all 2-tuples in a sample  (Figure \ref{fig:EPISCOPE_lor_histogram}) is much nicer than that for Pearson's correlation coefficient (Figure \ref{fig:EPISCOPE_pearson_histogram}). Specifically, the distributions of $LOR$ for each sample are approximately unimodal and symmetric whereas the distributions of Pearson's correlation are multimodal, asymmetric, and clearly constrained by the underlying marginal probabilities. It is far easier to work with symmetric, unimodal distributions, such as when wanting to summarise these by measures of central tendency and spread, which is important in what follows.

\begin{figure}[!htb]
\centering
\includegraphics[width=\textwidth]{../figures/EPISCOPE_lor_histogram.pdf}
\caption{Histograms of $LOR$ for 2-tuples from the EPISCOPE data (binwidth $= 0.5$).}
\label{fig:EPISCOPE_lor_histogram}
\end{figure}

\begin{figure}[!htb]
\centering
\includegraphics[width=\textwidth]{../figures/EPISCOPE_pearson_histogram.pdf}
\caption{Histograms of Pearson correlation coefficient for 2-tuples from the EPISCOPE data (binwidth $= 0.05$).}
\label{fig:EPISCOPE_pearson_histogram}
\end{figure}

Finally, and related to the second point, there are $2 \times 2$ tables that are clear evidence of co-methylation that have a Pearson correlation close to zero. An example is shown in Table \ref{tab:2x2_big_lor_small_cor}. This table has a $LOR = 3.3$ and Pearson's correlation of $-0.025$, which is due to the table being dominated by one cell's count, $uu = 39$. This occurs for all tables where one cell dominates. The problem is only partially fixed by adding an offset. For example, adding $1$ to all cells gives a Pearson's correlation of $0.29$, which is still a low correlation.

\begin{table}[h]
\centering
\begin{tabular}{lcccl}
& \multicolumn{1}{l}{} & \multicolumn{2}{l}{Second CpG} &                \\
&                      & \textbf{M}     & \textbf{U}    & \textbf{Total} \\ \cline{2-5}
\multirow{2}{*}{First CpG} & \textbf{M}           & 0              & 1             & 1            \\
& \textbf{U}           & 1              & 39             & 40             \\ \cline{2-5}
& \textbf{Total}       & 1              & 40             & 41
\end{tabular}
\caption{A $2 \times 2$ contingency table with a large $LOR$ but small Pearson's correlation.}
\label{tab:2x2_big_lor_small_cor}
\end{table}

\FloatBarrier
\subsection{Results}

For all 40 samples in the EPISCOPE, Lister, Seisenberger and Ziller datasets I compute the $LOR$ of all 2-tuples using both the $NIL = 0$ and $NIL \geq 0$ strategies. I compute these genome-wide and after stratifying CpG 2-tuples by whether they are in a CpG island. These analyses are performed separately for each strand and also for strand-collapsed counts of methylation patterns.

For any given $IPD$ there are hundreds of thousands of CpG pairs, and so there is a _distribution_ of co-methylation for any given $IPD$. For a given $IPD$ we can visualise the distribution of $LOR$ using a standard plot such as a histogram, kernel density estimate or boxplot. However, these $IPD$-specific plots don't give a sense of how $IPD$ affects within-fragment co-methylation. For this reason I instead plot selected quantiles --- $10\%, 25\%, 50\%, 75\% and 90\%$ --- of each $IPD$-specific $LOR$ distribution as a function of $IPD$. I refer to these as _quantile plots_ of $LOR$ as a function of $IPD$[^quantile]. These quantile plots may also be stratified by additional annotations such as genomic features.

[^quantile]: Because the distributions of $LOR$ are approximately symmetric and unimodal, these quantiles give a fair summary of each of these distributions. The same would not be true of the highly asymmetric and multimodal distributions of Pearson's correlations, hence one reason for my preference of $LOR$ over Pearson's correlation.

\FloatBarrier
\subsubsection{Can we collapse by strand?}

Figures \ref{fig:EPISCOPE_lor_min_cov_5}, \ref{fig:Lister_lor_min_cov_5}, \ref{fig:Seisenberger_lor_min_cov_5} and \ref{fig:Ziller_lor_min_cov_5} show the $LOR$ quantile plots for $NIL = 0$ CpGs 2-tuples stratified by strand. Since the distributions of $LOR$ are very similar for the two strands, I decided to use the strand-collapsed counts of methylation patterns in what follows. This is not to discount the possibility of 2-tuples with strand-specific differences in co-methylation, however, it does mean that it is generally safe to aggregate counts across strands.

\begin{figure}[!htb]
\centering
\includegraphics[width=\textwidth]{../figures/EPISCOPE_lor_min_cov_5.pdf}
\caption{$LOR$ as a function of $IPD$ for strand-specific $NIL = 0$ CpG 2-tuples for samples from the EPISCOPE dataset. The dark line plots the median of these distributions, the inner (darker) band the $25\% - 75\%$ percentiles, and the outer (lighter) bands the $10\% - 25\%$ and $75\% - 90\%$ percentiles. Both the x-axis and y-axis have been truncated for clarity.}
\label{fig:EPISCOPE_lor_min_cov_5}
\end{figure}

\begin{figure}[!htb]
\centering
\includegraphics[width=\textwidth]{../figures/Lister_lor_min_cov_5.pdf}
\caption{$LOR$ as a function of $IPD$ for strand-specific $NIL = 0$ CpG 2-tuples for samples from the Lister dataset. The dark line plots the median of these distributions, the inner (darker) band the $25\% - 75\%$ percentiles, and the outer (lighter) bands the $10\% - 25\%$ and $75\% - 90\%$ percentiles. Both the x-axis and y-axis have been truncated for clarity.}
\label{fig:Lister_lor_min_cov_5}
\end{figure}

\begin{figure}[!htb]
\centering
\includegraphics[width=\textwidth]{../figures/Seisenberger_lor_min_cov_5.pdf}
\caption{$LOR$ as a function of $IPD$ for strand-specific $NIL = 0$ CpG 2-tuples for samples from the Seisenberger dataset. The dark line plots the median of these distributions, the inner (darker) band the $25\% - 75\%$ percentiles, and the outer (lighter) bands the $10\% - 25\%$ and $75\% - 90\%$ percentiles. Both the x-axis and y-axis have been truncated for clarity.}
\label{fig:Seisenberger_lor_min_cov_5}
\end{figure}

\begin{figure}[!htb]
\centering
\includegraphics[width=\textwidth]{../figures/Ziller_merged_lor_min_cov_5.pdf}
\caption{$LOR$ as a function of $IPD$ for strand-specific $NIL = 0$ CpG 2-tuples for samples from the Ziller dataset. The dark line plots the median of these distributions, the inner (darker) band the $25\% - 75\%$ percentiles, and the outer (lighter) bands the $10\% - 25\%$ and $75\% - 90\%$ percentiles. Both the x-axis and y-axis have been truncated for clarity.}
\label{fig:Ziller_lor_min_cov_5}
\end{figure}

\FloatBarrier
\subsubsection{$NIL = 0$}

Figures \ref{fig:EPISCOPE_lor_min_cov_10_strand_collapsed}, \ref{fig:Lister_lor_min_cov_10_strand_collapsed}, \ref{fig:Seisenberger_lor_min_cov_10_strand_collapsed} and \ref{fig:Ziller_lor_min_cov_10_strand_collapsed} show the $LOR$ quantile plots for strand-collapsed $NIL = 0$ CpGs 2-tuples. Like for the correlations of $\beta$-values, two things immediately stand out. Firstly, for all samples the trend is that the strength of the within-fragment co-methylation decays as a function of $IPD$ --- 2-tuples with a larger $IPD$ are on average less co-methylated than2-tuples with a smaller $IPD$. Secondly, the estimated $LOR$s violently jump around for larger $IPD$s. The latter is simply due to the aforementioned lack of $NIL = 0$ pairs at larger $IPD$s with which to estimate these $LOR$s. Therefore, it is the trend in these plots that should be analysed.

It is also obvious that longer paired-end reads are far more preferable to shorter single-end reads when studying within-fragment co-methylation. Even when using paired-end reads, getting to $IPD = 150$ with $NIL = 0$ pairs is challenging. This is unfortunate because it doesn't allow us to investigate the potential nucleosome-related $\approx 150-170$ bp periodicity of methylation levels using $NIL = 0$ pairs.

\begin{figure}[!htb]
\centering
\includegraphics[width=\textwidth]{../figures/EPISCOPE_lor_min_cov_10_strand_collapsed.pdf}
\caption{$LOR$ as a function of $IPD$ for strand-collapsed $NIL = 0$ CpG 2-tuples for samples from the EPISCOPE dataset. The dark line plots the median of these distributions, the inner (darker) band the $25\% - 75\%$ percentiles, and the outer (lighter) bands the $10\% - 25\%$ and $75\% - 90\%$ percentiles. Both the x-axis and y-axis have been truncated for clarity.}
\label{fig:EPISCOPE_lor_min_cov_10_strand_collapsed}
\end{figure}

\begin{figure}[!htb]
\centering
\includegraphics[width=\textwidth]{../figures/Lister_lor_min_cov_10_strand_collapsed.pdf}
\caption{$LOR$ as a function of $IPD$ for strand-collapsed $NIL = 0$ CpG 2-tuples for samples from the Lister dataset. The dark line plots the median of these distributions, the inner (darker) band the $25\% - 75\%$ percentiles, and the outer (lighter) bands the $10\% - 25\%$ and $75\% - 90\%$ percentiles. Both the x-axis and y-axis have been truncated for clarity.}
\label{fig:Lister_lor_min_cov_10_strand_collapsed}
\end{figure}

\begin{figure}[!htb]
\centering
\includegraphics[width=\textwidth]{../figures/Seisenberger_lor_min_cov_10_strand_collapsed.pdf}
\caption{$LOR$ as a function of $IPD$ for strand-collapsed $NIL = 0$ CpG 2-tuples for samples from the Seisenberger dataset. The dark line plots the median of these distributions, the inner (darker) band the $25\% - 75\%$ percentiles, and the outer (lighter) bands the $10\% - 25\%$ and $75\% - 90\%$ percentiles. Both the x-axis and y-axis have been truncated for clarity.}
\label{fig:Seisenberger_lor_min_cov_10_strand_collapsed}
\end{figure}

\begin{figure}[!htb]
\centering
\includegraphics[width=\textwidth]{../figures/Ziller_merged_lor_min_cov_10_strand_collapsed.pdf}
\caption{$LOR$ as a function of $IPD$ for strand-collapsed $NIL = 0$ CpG 2-tuples for samples from the Ziller dataset. The dark line plots the median of these distributions, the inner (darker) band the $25\% - 75\%$ percentiles, and the outer (lighter) bands the $10\% - 25\%$ and $75\% - 90\%$ percentiles. Both the x-axis and y-axis have been truncated for clarity.}
\label{fig:Ziller_lor_min_cov_10_strand_collapsed}
\end{figure}

The are some consistent patterns in comparing these quantile plots across samples. Those 2-tuples with a small $IPD$ have a median $LOR = 3-4$ meaning that the loci in these 2-tuples are $8-16\times$ more likely to have the same rather than opposite methylation states. However, we also see than for any given $IPD$ that there is considerable variation in $LOR$. For example, for all samples and $IPD$s, the $10\%$ and $90\%$ quantiles span $3-5$ (base 2) orders of magnitude.

For each sample the $LOR$ distributions are consistently centred away from zero, meaning that over the observable $IPD$s the vast majority of 2-tuples are co-methylated. As $IPD$ increases, the CpG $NIL = 0$ 2-tuples become less co-methylated on average. However, the width of each sample's $LOR$ distributions remains relatively constant, so the variation in $LOR$ doesn't greatly depend on $IPD$.

While the similarities between samples are obvious, the differences are somewhat harder to detect from these genome-wide summaries. The ADS and FF mini datasets provide us with an opportunity to discuss some potential biological differences in within-fragment co-methylation. In Figure \ref{fig:Lister_lor_min_cov_10_strand_collapsed} we see that the _ADS-iPSC_ sample has a consistently higher level of co-methylation than the _ADS_ and _ADS-adipose_ samples. Likewise, the _FF-iPSC\_6.9_, _FF-iPSC\_19.7_ and _FF-iPSC\_19.11_ have consistently higher levels of co-methylation than the _FF_  sample, although not the _FF-iPSC\_19.11+BMP4_ sample.

I can think of two possible interpretations of this result. Firstly, iPSCs may have a more regulated methylome than do somatic cells. The fact that these same iPSCs have very little intermediate methylation could be seen as support of this interpretation. However, this same lack of intermediate methylation may simply reflect that the iPSC samples are a more homogeneous population of cells. It would be possible to have a sample with no within-fragment co-methylation but where all $\beta$-values are 0 or 1, e.g., a sample where all cells are perfect clones of a cell whose CpG methylation states are independent of one another.

A second interpretation of this result is that it is due to 'reseting' of the methylome upon induction of pluripotency \citep{Lister:2011kg, Stricker:2013kl}. As a population of cells develops from a single cell, there are errors in copying the DNA methylation patterns from mother to daughter cell. These errors disrupt the co-methylation of 2-tuples. If the iPSC and iPSC-derived cell lines have had less (developmental) time between the laying down of the original DNA methylation pattern and the point at which they were sequenced, then there is a lower probability that the co-methylation has been disrupted by replication errors. However, high levels of within-fragment co-methylation does not on its own imply that the sample is homogeneous. It would be possible to have a sample with high levels of within-fragment co-methylation that also has high levels of intermediate methylation, e.g., a sample with multiple distinct clonal subpopulations where each subpopulation has high levels of co-methylation but  also has a distinct methylome.

Taking the co-methylation results and the distribution of $\beta$-values together, I conclude that for the ADS and FF mini datasets the iPSC and iPSC-derived samples truly have higher levels of co-methylation and are also more homogeneous than their precursor cells.

The above analyses are all based on the genome-wide distributions of within-fragment co-methylation. Again, we can stratify these by whether the CpG 2-tuples are inside or outside of a CpG island. Figures \ref{fig:EPISCOPE_lor_min_cov_10_strand_collapsed_CGI}, \ref{fig:Lister_lor_min_cov_10_strand_collapsed_CGI}, \ref{fig:Seisenberger_lor_min_cov_10_strand_collapsed_CGI} and \ref{fig:Ziller_lor_min_cov_10_strand_collapsed_CGI} are plots of $LOR$ distributions for $NIL = 0$ CpG 2-tuples stratified by whether the pair is inside or outside of a CpG island[^spanning]. Again, the substantial noise in these plots is due to a lack of $NIL = 0$ pairs with large $IPD$s, particularly inside CpG islands.

Again, the message from these plots is clear, $NIL = 0$ CpG 2-tuples inside CpG islands have higher levels of within-fragment co-methylation on average than those 2-tuples outside of islands[^hepg2_2]. For each sample, the difference between the CpG island and non-island distributions is a vertical translation. For many samples this vertical shift means that the median $LOR$ for islands is approximately equal to the $75\%$ percentile for non-islands. The width of these distributions is not greatly affected by the CpG island status.

[^hepg2_2]: Again, the minor exception is the _HepG2\_cell\_line_ where the CpG island and non-island distributions approximately overlap by $IPD = 50$, but we have already seen that this sample has an unusual methylome compared to the other samples.

\begin{figure}[!htb]
\centering
\includegraphics[width=\textwidth]{../figures/EPISCOPE_lor_min_cov_10_strand_collapsed_CGI.pdf}
\caption{$LOR$ as a function of $IPD$ for strand-collapsed CpG 2-tuples stratified by CpG island status with $NIL = 0$ for samples from the EPISCOPE dataset. The dark line plots the median of these distributions, the inner (darker) band the $25\% - 75\%$ percentiles, and the outer (lighter) bands the $10\% - 25\%$ and $75\% - 90\%$ percentiles. Both the x-axis and y-axis have been truncated for clarity.}
\label{fig:EPISCOPE_lor_min_cov_10_strand_collapsed_CGI}
\end{figure}

\begin{figure}[!htb]
\centering
\includegraphics[width=\textwidth]{../figures/Lister_lor_min_cov_10_strand_collapsed_CGI.pdf}
\caption{$LOR$ as a function of $IPD$ for strand-collapsed CpG 2-tuples stratified by CpG island status with $NIL = 0$ for samples from the Lister dataset. The dark line plots the median of these distributions, the inner (darker) band the $25\% - 75\%$ percentiles, and the outer (lighter) bands the $10\% - 25\%$ and $75\% - 90\%$ percentiles. Both the x-axis and y-axis have been truncated for clarity.}
\label{fig:Lister_lor_min_cov_10_strand_collapsed_CGI}
\end{figure}

\begin{figure}[!htb]
\centering
\includegraphics[width=\textwidth]{../figures/Seisenberger_lor_min_cov_10_strand_collapsed_CGI.pdf}
\caption{$LOR$ as a function of $IPD$ for strand-collapsed CpG 2-tuples stratified by CpG island status with $NIL = 0$ for samples from the Seisenberger dataset. The dark line plots the median of these distributions, the inner (darker) band the $25\% - 75\%$ percentiles, and the outer (lighter) bands the $10\% - 25\%$ and $75\% - 90\%$ percentiles. Both the x-axis and y-axis have been truncated for clarity.}
\label{fig:Seisenberger_lor_min_cov_10_strand_collapsed_CGI}
\end{figure}

\begin{figure}[!htb]
\centering
\includegraphics[width=\textwidth]{../figures/Ziller_merged_lor_min_cov_10_strand_collapsed_CGI.pdf}
\caption{$LOR$ as a function of $IPD$ for strand-collapsed CpG 2-tuples stratified by CpG island status with $NIL = 0$ for samples from the Ziller dataset. The dark line plots the median of these distributions, the inner (darker) band the $25\% - 75\%$ percentiles, and the outer (lighter) bands the $10\% - 25\%$ and $75\% - 90\%$ percentiles. Both the x-axis and y-axis have been truncated for clarity.}
\label{fig:Ziller_lor_min_cov_10_strand_collapsed_CGI}
\end{figure}

\FloatBarrier

\subsubsection{$NIL \geq 0$}\label{sec:cometh_nil_geq0}

Figures \ref{fig:EPISCOPE_lor_min_cov_10_strand_collapsed_ac}, \ref{fig:Lister_lor_min_cov_10_strand_collapsed_ac}, \ref{fig:Seisenberger_lor_min_cov_10_strand_collapsed_ac} and \ref{fig:Ziller_lor_min_cov_10_strand_collapsed_ac} show the $LOR$ quantile plots for strand-collapsed $NIL \geq 0$ CpGs 2-tuples. By using $NIL \geq 0$ 2-tuples we can explore the effect of within-fragment co-methylation at greater $IPD$s, potentially beyond the crucial $IPD = \approx 150-170$ bp distance.


\begin{figure}[!htb]
\centering
\includegraphics[width=\textwidth]{../figures/EPISCOPE_lor_min_cov_10_strand_collapsed_ac.pdf}
\caption{$LOR$ as a function of $IPD$ for strand-collapsed $NIL \geq 0$ CpG 2-tuples for samples from the EPISCOPE dataset. The dark line plots the median of these distributions, the inner (darker) band the $25\% - 75\%$ percentiles, and the outer (lighter) bands the $10\% - 25\%$ and $75\% - 90\%$ percentiles. Both the x-axis and y-axis have been truncated for clarity.}
\label{fig:EPISCOPE_lor_min_cov_10_strand_collapsed_ac}
\end{figure}

\begin{figure}[!htb]
\centering
\includegraphics[width=\textwidth]{../figures/Lister_lor_min_cov_10_strand_collapsed_ac.pdf}
\caption{$LOR$ as a function of $IPD$ for strand-collapsed $NIL \geq 0$ CpG 2-tuples for samples from the Lister dataset. The dark line plots the median of these distributions, the inner (darker) band the $25\% - 75\%$ percentiles, and the outer (lighter) bands the $10\% - 25\%$ and $75\% - 90\%$ percentiles. Both the x-axis and y-axis have been truncated for clarity.}
\label{fig:Lister_lor_min_cov_10_strand_collapsed_ac}
\end{figure}

\begin{figure}[!htb]
\centering
\includegraphics[width=\textwidth]{../figures/Seisenberger_lor_min_cov_10_strand_collapsed_ac.pdf}
\caption{$LOR$ as a function of $IPD$ for strand-collapsed $NIL \geq 0$ CpG 2-tuples for samples from the Seisenberger dataset. The dark line plots the median of these distributions, the inner (darker) band the $25\% - 75\%$ percentiles, and the outer (lighter) bands the $10\% - 25\%$ and $75\% - 90\%$ percentiles. Both the x-axis and y-axis have been truncated for clarity.}
\label{fig:Seisenberger_lor_min_cov_10_strand_collapsed_ac}
\end{figure}

\begin{figure}[!htb]
\centering
\includegraphics[width=\textwidth]{../figures/Ziller_merged_lor_min_cov_10_strand_collapsed_ac.pdf}
\caption{$LOR$ as a function of $IPD$ for strand-collapsed $NIL \geq 0$ CpG 2-tuples for samples from the Ziller dataset. The dark line plots the median of these distributions, the inner (darker) band the $25\% - 75\%$ percentiles, and the outer (lighter) bands the $10\% - 25\%$ and $75\% - 90\%$ percentiles. Both the x-axis and y-axis have been truncated for clarity.}
\label{fig:Ziller_lor_min_cov_10_strand_collapsed_ac}
\end{figure}

__TODO__: I need to re-write this section. I'm really not yet sure if this result is biology or an artefact. My instinct says artefact but I haven't yet been able to prove this either way to my satisfaction
These plots appear to show that there is a dramatic increase in within-fragment co-methylation at precisely these $IPD$s for paired-end samples from the Lister dataset. However, I am very skeptical as to the biological relevance of these results.

There are three reasons for my skepticism. Firstly, similar results are not observed for any of the EPISCOPE samples nor any of the Ziller samples[^mouse_vs_man]. Secondly, the Lister samples with single-end sequencing have a similar spike in the estimated within-fragment co-methylation at the edge of the observable $IPD$s. However, for single-end data this corresponds to an $IPD \approx 60-80$ bp and we see no evidence of an increase in within-fragment co-methylation at these $IPD$s in any of the paired-end samples. This leads to my final reason, and what I think is the source of this artefact. Namely, estimates of within-fragment co-methylation are highly susceptible bias due to the incomplete removal of M-bias. This is particularly true for larger $IPD$s where the loci in the 2-tuples come from the ends of the reads, which are where M-bias is most prevalent.

[^mouse_vs_man]: I exclude the Seisenberger data simply because these are of mice rather than humans. The Seisenberger data also display a dramatic increase in within-fragment co-methylattion but since this increase occurs at the edge of the observable $IPD$s, I am concerned it suffers from the same bias as the Lister data.

As discussed in Section \ref{sec:m-bias}, the incomplete removal of M-bias is a real problem for reads that have undergone pre-trimming because this destroys the relationship between read-position and sequencing cycle. FIGURE shows that while there is considerable variation in the read lengths of the _ADS_ sample, which is indicative of aggressive pre-trimming of these reads, most reads are full length or close to it (TODO: Histogram of read lengths).

To investigate my suspicion that these results are due to the incomplete removal of M-bias, I used an aggressive filtering strategy that attempts to only use read-positions that are unlikely to suffer from M-bias. I applied this to the _ADS_ sample. In this analysis I ignored read-positions $1-20$ and $55-80$ of both _read 1_ and _read 2_. Unfortunately, this is a situation where the M-bias filtering strategy of `methtuple` is a disadvantage. `methtuple` is designed to filter out specific read-positions and cannot 'exclude the first/last $n$ bases'. For example, using this aggressive strategy will have a detrimental effect for a read that has been pre-trimmed to $30$ bp, of which the last $10$ bp are affected by M-bias, because this filtering strategy will _retain_ read-positions $21-30$, which are exactly those affected by M-bias. Nevertheless, it is the best that can be done with the current version of the software[^methtuple_refactoring] and with such aggressively pre-trimmed data.

[methtuple_refactoring]: We have now seen that both the 'filter specific read-positions' and 'filter the first/last $n$ read-positions' can lead to undesirable behaviour and so switching strategies is no saviour. The real solution is not to pre-trim reads and to instead use an aligner such as `bwa-meth` that can softclip these positions. Unfortunately, `bwa-meth` was only available after my data were already aligned and after `methtuple` was written. To use `methtuple` with `bwa-meth` output requires significant changes to `methtuple` because the `bwa-meth` output does not include anything analogous to the XM-tag of `Bismark`, which `methtuple` uses to do methylation calling.

The results for the aggressively trimmed _ADS_ sample are shown in FIGURE.

- TODO: Comment on these results and how we should interpret the Lister co-methylation results (particularly NIL >= 0)
- TODO: Summarise the M-bias issue
__

Comparing CpG islands to non-islands we again, Figures \ref{fig:EPISCOPE_lor_min_cov_10_strand_collapsed_ac_CGI}, \ref{fig:Lister_lor_min_cov_10_strand_collapsed_ac_CGI}, \ref{fig:Seisenberger_lor_min_cov_10_strand_collapsed_ac_CGI} and \ref{fig:Ziller_lor_min_cov_10_strand_collapsed_ac_CGI}, see that the methylation levels of pairs of CpGs within islands are consistently more correlated than those outside of islands.

\begin{figure}[!htb]
\centering
\includegraphics[width=\textwidth]{../figures/EPISCOPE_lor_min_cov_10_strand_collapsed_ac_CGI.pdf}
\caption{$LOR$ as a function of $IPD$ for strand-collapsed $NIL \geq 0$ CpG 2-tuples stratified by CpG island status for samples from the EPISCOPE dataset. The dark line plots the median of these distributions, the inner (darker) band the $25\% - 75\%$ percentiles, and the outer (lighter) bands the $10\% - 25\%$ and $75\% - 90\%$ percentiles. Both the x-axis and y-axis have been truncated for clarity.}
\label{fig:EPISCOPE_lor_min_cov_10_strand_collapsed_ac_CGI}
\end{figure}

\begin{figure}[!htb]
\centering
\includegraphics[width=\textwidth]{../figures/Lister_lor_min_cov_10_strand_collapsed_ac_CGI.pdf}
\caption{$LOR$ as a function of $IPD$ for strand-collapsed $NIL \geq 0$ CpG 2-tuples stratified by CpG island status for samples from the Lister dataset. The dark line plots the median of these distributions, the inner (darker) band the $25\% - 75\%$ percentiles, and the outer (lighter) bands the $10\% - 25\%$ and $75\% - 90\%$ percentiles. Both the x-axis and y-axis have been truncated for clarity.}
\label{fig:Lister_lor_min_cov_10_strand_collapsed_ac_CGI}
\end{figure}

\begin{figure}[!htb]
\centering
\includegraphics[width=\textwidth]{../figures/Seisenberger_lor_min_cov_10_strand_collapsed_ac_CGI.pdf}
\caption{$LOR$ as a function of $IPD$ for strand-collapsed $NIL \geq 0$ CpG 2-tuples stratified by CpG island status for samples from the Seisenberger dataset. The dark line plots the median of these distributions, the inner (darker) band the $25\% - 75\%$ percentiles, and the outer (lighter) bands the $10\% - 25\%$ and $75\% - 90\%$ percentiles. Both the x-axis and y-axis have been truncated for clarity.}
\label{fig:Seisenberger_lor_min_cov_10_strand_collapsed_ac_CGI}
\end{figure}

\begin{figure}[!htb]
\centering
\includegraphics[width=\textwidth]{../figures/Ziller_merged_lor_min_cov_10_strand_collapsed_ac_CGI.pdf}
\caption{$LOR$ as a function of $IPD$ for strand-collapsed $NIL \geq 0$ CpG 2-tuples stratified by CpG island status for samples from the Ziller dataset. The dark line plots the median of these distributions, the inner (darker) band the $25\% - 75\%$ percentiles, and the outer (lighter) bands the $10\% - 25\%$ and $75\% - 90\%$ percentiles. Both the x-axis and y-axis have been truncated for clarity.}
\label{fig:Ziller_lor_min_cov_10_strand_collapsed_ac_CGI}
\end{figure}

\FloatBarrier
\section{Summary}

Correlations of $\beta$-values and estimates of within-fragment co-methylation are complementary methods to better understand the complex dependence structure of bisulfite-sequencing data. Which is more relevant depends on the question at hand. Within-fragment co-methylation seems closer to the biology because it measures the dependence of DNA methylation on the scale that the DNA methyltransferases act. However, the most common downstream analyses are based on $\beta$-values and so these correlations may be more relevant to analyses of DNA methylation data. Ultimately, the two measures are intertwined since the correlations of $\beta$-values are driven by co-methylation on individual DNA fragments. This idea is explored in greater detail in Chapter \ref{chap:methsim}.

A major challenge in interpreting within-fragment co-methylation is its susceptibility to M-bias or, rather, the difficulty of accounting for M-bias in pre-trimmed data. Since we can do nothing about data that has already been pre-trimmed, we are forced to be cautious in our interpretation of within-fragment co-methylation for 2-tuples with larger $IPD$s. This is particularly relevant to the increase in within-fragment co-methylation for $NIL \geq 0$ CpG 2-tuples with $IPD \approx 150-170$. This result is tantalisingly similar to the periodicity we see in the correlations of $\beta$-values and the approximate spacing of nucleosomes in human DNA (see Section \ref{sec:nucleosome_and_chromatin}). However, I do not feel comfortable concluding that this is truly what we are observing because of the potential for these systematic biases to create an artificial signal. Furthermore, we do not detect this signal any of the EPISCOPE samples, which are some of the best quality data I have and, being samples from human tissues rather than cell lines, are arguably more biologically relevant.

The presented analyses only measure pairwise dependencies. More specifically, the $NIL = 0$ results measure the 'first-order' or 'one-step' correlations of $\beta$-values whereas the $NIL \geq 0$ measure a mixture of different orders of dependences but all are done in a pairwise manner. I find the $NIL = 0$ results easier interpret but they clearly do not tell the whole story.

By comparing the $NIL = 0$ and $NIL \geq 0$ results we see that the strength of the dependency itself depends on not just the distance between the two CpGs but the number of intervening CpGs. This indicates that these dependencies are more than 'first-order'. Moreover, the differences in dependencies between CpG islands and non-islands indicates that the 'order' of these dependencies itself varies across the genome. All this means that the DNA methylation measurements for a single sample form a complicated and highly non-stationary process with variable order dependencies.

In theory, we could explore these higher-order dependencies of within-fragment co-methylation by analysing m-tuples with $\text{m} > 2$. However, the complexity of the challenge increases exponentially with m. For example, even when analysing 3-tuples there are four $LOR$s for each $2 \times 2 \times 2$ contingency table and the $IPD$ vector now has two dimensions, meaning that the results have to be visualised over a surface.

Another complexity that these analyses haved brushed over is the fact that the genome is not a 1D line but is rather a complex 3D structure whose shape also varies in time. Therefore, the relevant distance between two loci is not necessarily the number of basepair between them but may instead be the Euclidean distance in 3D space. Such complexities cannot be resolved with bisulfite-sequencing data alone but will require measures of the 3D structure of the genome, such as those provided by chromatin conformation capture technologies \citep{Dekker:2013hi}.

The results of this chapter clearly demonstrate that DNA methylation has a highly complex dependence structure. These dependencies exist on individual DNA fragments and also manifest as correlations of aggregate measures of methylation, such as $\beta$-values. Bisulfite-sequencing give the highest resolution picture of this dependence structure. Nonetheless, this is still a rather limited picture, particularly for within-fragment co-methylation. Data from longer reads will be immensely useful in developing a better model of within-fragment co-methylation. In particular, reads that can span multiple nucleosomes will allow for co-methylation to be better tied in with and expand upon existing results on the relationship between nucleosomes and DNA methylation. Until such data are available, we can try to learn more about plausible models of DNA co-methylation through computational methods such as the simulation method described in the next chapter.

# Problem with LOR estimates

The simple estimator of within-fragment co-methylation is biased when the marginal probability of methylation is close to the boundary (0 or 1) unless the 2-tuple is sequenced at very high depth. I demonstrate this by simulating methylation patterns at 2-tuples as independent $binomail(1, p)$ random variables and showing that the simple LOR estimator is positively biased.

__TODO: Move this code chunk to supplementary material and only include results in thesis.__

```{r}
# Requires the installation of methsim
library(mipfp)

# rtpois is a helper function to simulate n truncated Poisson(lambda) with
# truncation specified by 'trunc'.
# It is adapted from a method for simulating zero-truncated Poisson suggested
# by Peter Dalgaard https://stat.ethz.ch/pipermail/r-help/2005-May/070680.html
rtpois <- function(n, lambda, trunc) {
  qpois(runif(n, ppois(trunc, lambda), 1), lambda)
}

# Function to simulate 2x2xK tables.
# Each 2x2 table represents the counts of unmethylated (0) and methylated (1) states
# at each CpG for each read.
# The number of reads per table (i.e., 2x2 table sample size) is simulated
# from a truncated Poisson(lambda) distribution; truncation is to simulate
# minimum coverage cutoffs and is specified by the 'trunc' argument.
# The marginal probabilities of each table are p1 = Prob(methylated at first CpG)
# and p2 = Prob(methylated at second CpG).
# The odds ratio (or) is for each 2x2 table and should be of length K or 1 (in
# which case all K tables have the same odds ratio).
simTables <- function(K, lambda, trunc, p1, p2, or) {

  # d is the sample size of each 2x2 table (i.e, the sequencing depth for the 2-tuple)
  d <- rtpois(n = K, lambda = lambda, trunc = trunc)
  if (length(p1) == 1L) {
    p1 <- rep(p1, K)
  }
  if (length(p2) == 1L) {
    p2 <- rep(p2, K)
  }
  if (length(or) == 1L) {
    or <- rep(or, K)
  }
  .simTables <- function(k, d, p1, p2, or) {
    # Compute joint probabilities
    P <- mipfp::Ipfp(seed = matrix(c(or, 1, 1, 1), ncol = 2),
                     target.list = list(1, 2),
                     target.data = list(c(1 - p1, p1), c(1 - p2, p2)))$x.hat
    matrix(rmultinom(1, d, as.vector(P)), ncol = 2)
  }
  mapply(.simTables, k = K, d = d, p1 = p1, p2 = p2, or = or, SIMPLIFY = "array")
}

# Estimator of the (offset-adjusted) unconditional log odds ratios of K 2x2 tables
LOR <- function(x, offset = 0.5) {
  x <- x + offset
  MM <- x[2, 2, ]
  MU <- x[2, 1, ]
  UM <- x[1, 2, ]
  UU <- x[1, 1, ]
  log2(MM) + log2(UU) - log2(MU) - log2(UM)
}

# Mantel-Haenszel estimator of a common log odds ratio of a 2x2xK table
MH <- function(x, offset = 0) {
  x <- x + offset
  # TODO: return the standard error
  MM <- x[2, 2, ]
  MU <- x[2, 1, ]
  UM <- x[1, 2, ]
  UU <- x[1, 1, ]
  d <- apply(x, 3, sum)
  or <- sum((UU * MM) / d) / sum((MU * UM) / d)
  # One or more margins must be identically zero so add offset
  if (is.nan(or) | isTRUE(all.equal(or, 0)) | is.infinite(or)) {
    warning("One or more cell is (probably) identically zero in all K 2x2 tables. Adding offset = 0.5")
    MH(x, 0.5)
  } else {
    log2(or)
  }
}

# Estimator of the (offset-adjusted) unconditional log odds ratio of the
# 2x2xK collapsed over K.
LORCollapsed <- function(x, offset = 0.5) {
  z <- apply(x, c(1, 2), sum)
  # Artificially augment so it's a 2x2x1 table
  dim(z) <- c(dim(z), 1)
  lor <- LOR(z, 0)
  if (is.nan(lor) | is.infinite(lor)) {
    warning("One or more cell is (probably) identically zero in all K 2x2 tables. Adding offset = 0.5")
    LOR(z, 0.5)
  } else {
    lor
  }
}

# Simulate and plot data when p1 = p2.
# 100 CpG pairs, 30x average coverage truncated at 10
# odds ratio = 1, i.e., independence
set.seed(666)
par(mfrow = c(3, 3))
K <- 100
lambda <- 20
trunc <- 11
or <- 1
p1 <- c(0.01, 0.05, 0.1, 0.25, 0.5, 0.75, 0.9, 0.95, 0.99)
p2 <- p1
for (i in 1:9) {
  x <- simTables(K = K, lambda = lambda, trunc = trunc, p1 = p1[i], p2 = p2[i], or = or)
  hist(LOR(x), main = paste0("p1 = ", p1[i], ", p2 = ", p2[i], ", lor = ", log2(or)),
       xlim = c(-10, 10), breaks = 20,
       xlab = expression(paste(log[2], "(odds ratio)")),
       freq = FALSE)
  abline(v = MH(x), col = "dodgerBlue", lwd = 3)
  abline(v = LORCollapsed(x), col = "orange", lwd = 3)
  abline(v = log2(or), col = "black", lty = 2)
}
title(paste0("K = ", K, ", lambda = ", lambda, ", trunc = ", trunc), outer = TRUE, line = -1)

# Simulate and plot data when p1 = p2.
# 1000 CpG pairs, 30x average coverage truncated at 10
# odds ratio = 1, i.e., independence
set.seed(666)
par(mfrow = c(3, 3))
K <- 1000
lambda <- 20
trunc <- 10
or <- 1
p1 <- c(0.01, 0.05, 0.1, 0.25, 0.5, 0.75, 0.9, 0.95, 0.99)
p2 <- p1
for (i in 1:9) {
  x <- simTables(K = K, lambda = lambda, trunc = trunc, p1 = p1[i],
                 p2 = p2[i], or = or)
  hist(LOR(x),
       main = paste0("p1 = ", p1[i], ", p2 = ", p2[i], ", lor = ", log2(or)),
       xlim = c(-10, 10), breaks = 20,
       xlab = expression(paste(log[2], "(odds ratio)")),
       freq = FALSE)
  abline(v = MH(x), col = "dodgerBlue", lwd = 3)
  abline(v = LORCollapsed(x), col = "orange", lwd = 3)
  abline(v = log2(or), col = "black", lty = 2)
}
title(paste0("K = ", K, ", lambda = ", lambda, ", trunc = ", trunc), outer = TRUE, line = -1)

# Simulate and plot data when p1, p2 ~ Uniform(0, 1)
# 100 CpG pairs, 30x average coverage truncated at 10
# log odds ratio = -4, -3, ..., 3, 4.
set.seed(666)
par(mfrow = c(3, 3))
K <- 100
lambda <- 20
trunc <- 11
or <- 2^seq(-4, 4, 1)
p1 <- runif(K)
p2 <- runif(K)
for (i in 1:9) {
  x <- simTables(K = K, lambda = lambda, trunc = trunc, p1 = p1, p2 = p2,
                 or = or[i])
  hist(LOR(x),
       main = paste0("p1 = U(0, 1), p2 = U(0, 1)", ", lor = ", log2(or[i])),
       xlim = c(-10, 10), breaks = 20,
       xlab = expression(paste(log[2], "(odds ratio)")),
       freq = FALSE)
  abline(v = MH(x), col = "dodgerBlue", lwd = 3)
  abline(v = LORCollapsed(x), col = "orange", lwd = 3)
  abline(v = log2(or[i]), col = "black", lty = 2)
}
title(paste0("K = ", K, ", lambda = ", lambda, ", trunc = ", trunc),
      outer = TRUE, line = -1)

# Simulate and plot data when p1 ~ Uniform(0, 1), p2 = p1
# 100 CpG pairs, 30x average coverage truncated at 10
# log odds ratio = -4, -3, ..., 3, 4.
set.seed(666)
par(mfrow = c(3, 3))
K <- 100
lambda <- 20
trunc <- 11
or <- 2^seq(-4, 4, 1)
p1 <- runif(K)
p2 <- p1
for (i in 1:9) {
  x <- simTables(K = K, lambda = lambda, trunc = trunc, p1 = p1, p2 = p2, or = or[i])
  hist(LOR(x), main = paste0("p1 = U(0, 1), p2 = p1", ", lor = ", log2(or[i])),
       xlim = c(-10, 10), breaks = 20,
       xlab = expression(paste(log[2], "(odds ratio)")),
       frequeny = FALSE)
  abline(v = MH(x), col = "dodgerBlue", lwd = 3)
  abline(v = LORCollapsed(x), col = "orange", lwd = 3)
  abline(v = log2(or[i]), col = "black", lty = 2)
}
title(paste0("K = ", K, ", lambda = ", lambda, ", trunc = ", trunc),
      outer = TRUE, line = -1)

# Simulate and plot data when p1 ~ Beta(0.3, 0.2), p2 = p1.
# 100 CpG pairs, 30x average coverage truncated at 10
# log odds ratio = -4, -3, ..., 3, 4.
# Beta(0.3, 0.2) is roughly like the genome-wide distribution of methylation
# levels (B).
set.seed(666)
par(mfrow = c(3, 3))
K <- 100
lambda <- 20
trunc <- 11
or <- 2^seq(-4, 4, 1)
p1 <- rbeta(K, 0.3, 0.2)
p2 <- p1
for (i in 1:9) {
  x <- simTables(K = K, lambda = lambda, trunc = trunc, p1 = p1, p2 = p2, or = or[i])
  hist(LOR(x), main = paste0("p1 = Beta(0.3, 0.2), p2 = p1", ", lor = ", log2(or[i])),
       xlim = c(-10, 10), breaks = 20,
       xlab = expression(paste(log[2], "(odds ratio)")),
       freq = FALSE)
  abline(v = MH(x), col = "dodgerBlue", lwd = 3)
  abline(v = LORCollapsed(x), col = "orange", lwd = 3)
  abline(v = log2(or[i]), col = "black", lty = 2)
}
title(paste0("K = ", K, ", lambda = ", lambda, ", trunc = ", trunc), outer = TRUE, line = -1)

# Simulate and plot data when p1 ~ Beta(0.3, 0.2), p2 = p1.
# 100 CpG pairs, 30x average coverage truncated at 10
# log2(or) ~ Gaussian(log2(or_mean), 1)
# Beta(0.3, 0.2) is roughly like the genome-wide distribution of methylation
# levels (B).
set.seed(666)
par(mfrow = c(3, 3))
K <- 100
lambda <- 20
trunc <- 11
or_mean <- 2 ^ seq(-4, 4, 1)
p1 <- rbeta(K, 0.3, 0.2)
p2 <- p1
for (i in 1:9) {
  or <- 2 ^ rnorm(K, log2(or_mean[i]), 1)
  x <- simTables(K = K, lambda = lambda, trunc = trunc, p1 = p1, p2 = p2, or = or)
  hist(LOR(x), main = paste0("p1 = Beta(0.3, 0.2), p2 = p1", ", lor = Norm(", log2(or_mean[i]), ", 1)"),
       xlim = c(-10, 10), breaks = 20,
       xlab = expression(paste(log[2], "(odds ratio)")),
       freq = FALSE)
  abline(v = MH(x), col = "dodgerBlue", lwd = 3)
  abline(v = LORCollapsed(x), col = "orange", lwd = 3)
  lines(density(log2(or)), col = "black", lty = 2)
}
title(paste0("K = ", K, ", lambda = ", lambda, ", trunc = ", trunc), outer = TRUE, line = -1)

# Simulate and plot data when p1 ~ Beta(0.3, 0.2), p2 = p1.
# 1000 CpG pairs, 30x average coverage truncated at 10
# log2(or) ~ Gaussian(log2(or_mean), 1)
# Beta(0.3, 0.2) is roughly like the genome-wide distribution of methylation
# levels (B).
set.seed(666)
par(mfrow = c(3, 3))
K <- 1000
lambda <- 20
trunc <- 11
or_mean <- 2 ^ seq(-4, 4, 1)
p1 <- rbeta(K, 0.3, 0.2)
p2 <- p1
for (i in 1:9) {
  or <- 2 ^ rnorm(K, log2(or_mean[i]), 1)
  x <- simTables(K = K, lambda = lambda, trunc = trunc, p1 = p1, p2 = p2, or = or)
  hist(LOR(x), main = paste0("p1 = Beta(0.3, 0.2), p2 = p1", ", lor = Norm(", log2(or_mean[i]), ", 1)"),
       xlim = c(-10, 10), breaks = 20,
       xlab = expression(paste(log[2], "(odds ratio)")),
       freq = FALSE)
  abline(v = MH(x), col = "dodgerBlue", lwd = 3)
  abline(v = LORCollapsed(x), col = "orange", lwd = 3)
  lines(density(log2(or)), col = "black", lty = 2)
}
title(paste0("K = ", K, ", lambda = ", lambda, ", trunc = ", trunc), outer = TRUE, line = -1)

# Simulate and plot data when p1 ~ Beta(0.3, 0.2), p2 = p1.
# 1000 CpG pairs, 30x average coverage truncated at 10
# log2(or) ~ Skewed-Gaussian(log2(or_mean), 1, -10)
# Beta(0.3, 0.2) is roughly like the genome-wide distribution of methylation
# levels (B).
library(sn)
set.seed(666)
par(mfrow = c(3, 3))
K <- 1000
lambda <- 20
trunc <- 11
or_mean <- 2 ^ seq(-4, 4, 1)
p1 <- rbeta(K, 0.3, 0.2)
p2 <- p1
for (i in 1:9) {
  or <- 2 ^ rsn(K, log2(or_mean[i]), 1, -10)
  x <- simTables(K = K, lambda = lambda, trunc = trunc, p1 = p1, p2 = p2, or = or)
  hist(LOR(x), main = paste0("p1 = Beta(0.3, 0.2), p2 = p1", ", lor = SN(", log2(or_mean[i]), ", 1, -10)"),
       xlim = c(-10, 10), breaks = 20,
       xlab = expression(paste(log[2], "(odds ratio)")),
       freq = FALSE)
  abline(v = MH(x), col = "dodgerBlue", lwd = 3)
  abline(v = LORCollapsed(x), col = "orange", lwd = 3)
  lines(density(log2(or)), col = "black", lty = 2)
}
title(paste0("K = ", K, ", lambda = ", lambda, ", trunc = ", trunc), outer = TRUE, line = -1)

# Simulate and plot data when p1 ~ Beta(0.3, 0.2), p2 = p1.
# 1000 CpG pairs, 30x average coverage truncated at 10
# log2(or) ~ 0.5 * Gaussian(log2(or_mean), 1) + 0.5 * Gaussian(-log2(or_mean))
# Beta(0.3, 0.2) is roughly like the genome-wide distribution of methylation
# levels (B).
set.seed(666)
par(mfrow = c(3, 3))
K <- 1000
lambda <- 20
trunc <- 11
or_mean <- 2 ^ seq(-4, 4, 1)
w <- sample(c(-1, 1), K, replace = TRUE)
p1 <- rbeta(K, 0.3, 0.2)
p2 <- p1
for (i in 1:9) {
  or <- 2 ^ rnorm(K, w * log2(or_mean[i]), 1)
  x <- simTables(K = K, lambda = lambda, trunc = trunc, p1 = p1, p2 = p2, or = or)
  hist(LOR(x), main = paste0("p1 = Beta(0.3, 0.2), p2 = p1", ", lor = NM(", log2(or_mean[i]), ", 1)"),
       xlim = c(-10, 10), breaks = 20,
       xlab = expression(paste(log[2], "(odds ratio)")),
       freq = FALSE)
  abline(v = MH(x), col = "dodgerBlue", lwd = 3)
  abline(v = LORCollapsed(x), col = "orange", lwd = 3)
  lines(density(log2(or)), col = "black", lty = 2)
}
title(paste0("K = ", K, ", lambda = ", lambda, ", trunc = ", trunc), outer = TRUE, line = -1)

# TODO: Increase d for each table
```

## Following chat with Terry

These are my notes following a Skype conversion with Terry to discuss the issue.

- [ ] Try a higher order Taylor expansion of the log odds-ratio to get e.g., $n^{1}$ or $n^{3/2}$ convergence. The $+\frac{1}{2}$ fudge-factor is motivated by such an expansion and perhaps it is possible to do better by a higher order expansion.
  - [ ] First try showing the $c = \frac{1}{2}$ is optimal.
- [ ] Read texts about this issue
- [ ] Look for research papers on the general issue of dealing with bias in log odds-ratio estimates
- [ ] What if I just increase the fudge-factor to increase the shrinkage/smoothing of the estimates? It'll shrink to zero alright, but what is the theoretical justification?
- [ ] Check out the behaviour of the $\phi$ coefficient, i.e., Pearson's $r$ of the $2 \times 2$ table.
  - Landan et al.: "Note that correlation values are subject to considerable noise when coverage is low or when m1 or m2 are close to 0 or 1."
    - They don't define whether 'noise' means bias, variance, skewness, etc.

## Miscellaneous notes

Agresti in Categorical Data Analysis:

> In terms of bias and mean square error, Gart and Zweiful (1967) and Haldane (1955) showed the amended [i.e, add 0.5 to each count] estimators of the [odds-ratio log odds-ratio] behave well (see also Problem 12.4)

The Mantel-Haenszel estimator might be my best bet. Breslow (1981) says:

> The principal conclusion to be drawn from these results is that one should avoid using two of the standard techniques of odds ratio estimation with sparse data.

> The bias of the empirical logit estimator [i.e., + 0.5 to each cell], on the other hand, is sufficient erratic that the possibility of correction seems remote.

> Breslow (1981, Biometrika 68, 73-84) has shown that the Mantel-Haenszel odds ratio is a consistent estimator of a common odds ratio in sparse stratifications (Greenland, S. & Robins, J. M. Biometrics 41, 5568 (1985).)

The Mantel-Haenszel estimator is an estimator of the common odds ratio for $2 \times 2 \times K$ tables. Therefore, using the MH estimator requires the assumption that all $2 \times 2$ tables have the same underlying odds ratio. The strata for co-methylation would be 'sample', 'IPD' and possibly 'region', where region is something like 'UMRs, LMRs, PMRs and other' or 'CpG island and non-island'.

Some good notes on MH estimator [http://isites.harvard.edu/fs/docs/icb.topic79671.files/stratcch1.pdf](http://isites.harvard.edu/fs/docs/icb.topic79671.files/stratcch1.pdf).

An offset of 0.5 eliminates the first-order bias of the maximum likelihood estimator of the (log) odds ratio (Walter, S. D. Biostatistics (1987).)


Regarding the effect of the homogeneity on the Mantel-Haenszel estimator, Greenland, S. Stat Med 1, 217227 (1982):

> Under the homogeneity assumption, OR^MH is an almost fully efficient estimator of the common value, and remains a consistent estimator of this value even with sparse data, i.e. when the fineness of stratification results in many small cell counts. __As shown later, its convergence properties under heterogeneity are not as desirable, although even in this case it remains insensitive to small cells.__ [emphasis added]

The _Miettenen_ estimator is like a generalised form of the Mantel-Haenszel estimator, but assumes each cell count is greater than approximately 4 (Greenland, S. Stat Med 1, 217227 (1982)).

## Plan of attack

Following discussion with Terry (2015-04-21), this is my plan of attack to finish this chapter and the `methsim` chapter.

- [x] Finish reading Greenland (1982).
  - What happens with the MH estimator when the common odds ratio assumption is invalid. Is there a better estimator in this setting?

  > As a consequence, when there is heterogeneity, the expectation of MH varies with the overall and stratum-specific case-control ratios. The usual common odds ratio estimators share this disconcerting property \citep{Greenland:1982uk}.

  In contrast to the MH estimator, the Miettenen estimator consistently estimates meaningful epimiological parameters. However, it "[requires] consideration in the selection of the weights".


- [ ] Demonstrate via simulation that the unconditional LOR estimator applied to $2 \times 2$ tables is biased when $d$ is small and the marginal probabilities are near the boundaries.
  - Corrollary: Don't use the LOR estimates of within-fragment co-methylation from individual $2\times 2$ tables. I know they're biased, so there's no reason to show these results.
- [ ] Demonstrate that the MH estimator applied to $2 \times 2 \times K$ tables is unbiased when $d$ is small (per-table) and the marginal probabilities are near the boundaries.
  - Need to investigate the setting where the marginal probabilities are heterogeneous across the $K$ tables but with the same odds-ratio.
  - Need to investigate the setting where the odds ratio __is not__ common to the $K$ tables; what does the MH estimator end up estimating?

- Run Woolf's test or Breslow-Day test of homogeneity of odds ratios (noting that these are not powerful tests).

__In fact, Breslow himself advocates for the Tarone statistic in place of the Breslow-Day statistic (Breslow, N. E. JASA 91, 1428 (1996)). Also, Woolf's test, Breslow-Day and Tarone's statistic are only appropriate in Asymptopia I (a small number of tables with large frequencies) and not in Asymptopia II (a large number of tables with small frequencies); see Liang and Self (1985) for Asymptopia II (the scenario I am in)__

Liang, K.-Y. & Self, S. G. Biometrika (1985) advocate for a conditional test of homogeneity of odds ratios when the data are sparse (Asymptopia II) but note that:

> The problem with the conditional approach is its complexity. It is clear that when the sample size of each table increases, the computation of T2, T4 or T5 [ the conditional tests] becomes a burden which, eventually, will be prohibitive.
This raises the question of how large the sample sizes in each table must be before the unconditional score test T1 is reasonable to use. Further simulation results suggest that even when the number of cases and controls are as small as 10 with as many as 10 tables, the large-sample chi-squared approximation is adequate in that the observed sizes are near to the nominal 0 05 level.

I have $K \\g 10$ ($K$ = number of $2 \times 2$ tables) and $d \geq 10$ ($d$ = sequencing coverage), so the unconditional test may well be fine.

__TODO: Look up Liang-Self in Agresti's CDA (2013)__

- Algorithm: If the null of homogeneity of odds ratio is rejected => stratify further(?), otherwise compute MH or Miettenen estimates of the log odds ratio (and standard errors) for different stratification strategies (see [https://stat.ethz.ch/pipermail/r-help/2007-February/126254.html](https://stat.ethz.ch/pipermail/r-help/2007-February/126254.html))
  - Stratify by IPD
  - Stratify by IPD + chromosome
  - Stratify by IPD + chromosome + chromosome arm
  - Stratify by IPD + chromosome + chromosome arm + bin (Mb)
  - Check whether the more-stratified estimates fall within a reasonable distance from the less-stratified estimates (i.e., with respect to standard errors of the less-stratified estimates).
- Settle on the most appropriate level of stratification and apply to all samples.
  - Discuss results.
- `methsim`: Sample LORs from a Normal(mu, sigma^2) distribution, where mu (resp. sigma) is the MH estimator (resp. standard error? empirical variance of stratified MH estimators?) of within-fragment co-methylation for that IPD and region-type.
