---
author: "Peter Hickey"
output:
  pdf_document:
    fig_caption: yes
    includes:
      in_header: ../latex/header.tex
  html_document:
    keep_md: true
bibliography: ../latex/phd_thesis.bib
---

\chapter{Co-methylation}\label{chap:co-methylation}

## Chapter overview

In this chapter I describe two measures of within-sample co-methylation, _within-fragment co-methylation_ and _correlation of $\beta$-values_. Using both these measures I explore how co-methylation is affected by the distance between methylation loci, the density of methylation loci and the genomic context of the methylation loci. I apply these methods to 40 whole-genome bisulfite-sequencing samples to study the between-sample variability of co-methylation.

Throughout this chapter I focus on co-methylation of CpGs, although the methods described are also applicable to non-CpG methylation loci. The reason I do not explore co-methylation at non-CpG methylation loci is because such loci are rarely methylated in the samples that I am studying.

\section{Correlations of $\beta$-values}\label{subsec:beta_cor}

Chapter \ref{chap:co-methylation_review} documented several previous analyses based on 'correlations' of aggregate methylation levels, such as $\beta$-values. What was lacking from these were clear descriptions of the proposed methods and software implementations.

\subsection{Methods}

Here I describe two simple methods based on correlations of $\beta$-values. Both of these are implemented in the `methLevelCor` function that is part of the `MethylationTuples` software.

The aim of this analysis is to address the question of 'how are aggregate methylation levels correlated as a function of the distance between loci?'. There are two steps in this analysis:

1. Define the pairs of methylation loci.
2. For each $IPD$, compute the correlation of $\beta$-values for all pairs of methylation loci separated by that $IPD$.

Step (2) may be further stratified by the genomic context of the pair of CpGs, such as whether they are in a CpG island, or other variables of interest.

The same analysis could be performed using $\mathcal{M}$-values rather than $\beta$-values, and this is available as an option in the `methLevelCor` function.

\subsubsection{Constructing the pairs}\label{sec:beta_pair_construction}

I consider two different strategies for creating pairs of CpGs. Both strategies are based on CpGs in the reference genome after having filtered out those reference-specific CpGs. That is, I begin with $\mathcal{I}^{ref}$ and, based on the `Bis-SNP` output, I filter out loci that are not CpGs in the sample's genome. Rather than actually remove these sites, I simply set the corresponding $\beta$-value to `NA`. I also set the $\beta$-values of 'missing' CpGs, those without sufficient sequencing coverage to call a $\beta$-value, to `NA` so that these are still included in the set of CpGs.

The first strategy creates all pairs of adjacent CpGs on the same chromosome and strand, which are then stratified by $IPD$ and any secondary variables, such as genomic context. For a chromosome with $N$ CpGs there are $(N - 1)$ pairs. I call this the 'adjacent pairs' or $NIL = 0$ strategy, since all pairs have zero intervening loci^[nil0].

[^nil]: Recall that this is with respect to the reference genome and not the sample genome.

The second strategy uses all pairs of CpGs on the same chromosome and strand, which are then stratified by $IPD$ and any secondary variables, such as genomic context. In practice, the second strategy only uses pairs for a set of $IPD$s, e.g., $IPD = 2 - 2000$, otherwise the number of pairs becomes unwieldy. This is more like a traditional autocorrelation analysis of $\bm{\beta} = (\beta_{1}, \ldots, \beta_{N_{loci}})$, albeit of a vector of unequally spaced observations and with care taken to only use pairs on the same chromosome and strand. I refer to this as the $NIL \geq 0$ strategy. Using the $NIL \geq 0$ strategy, two pairs of CpGs with an identical $IPD$ may have a very different number of intervening CpGs.

The interpretation of correlations computed using the $NIL \geq 0$ is complicated because the set of pairs are more heterogeneous than those under the $NIL = 0$ strategy. For this reason I prefer the $NIL = 0$ strategy, however, it may be useful to consider and contrast the two strategies, which is what I have done in my thesis.

\subsubsection{Stratifying by a genomic feature}\label{sec:pair_stratification}

Even when using the $NIL = 0$ strategy, two pairs of CpGs with an identical $IPD$ may come from two distinct regions of the genome that have very different methylation dynamics. In other words, the vector of $\beta$-values for a given sample is highly non-stationary --- the covariance and correlation between two loci depends on more than just the $IPD$. An obvious thing to investigate is how strong an influence the genomic context has on these correlations of $\beta$-values. To do this, we can stratify our analysis by a genomic feature. While this does not eliminate this non-stationarity, it may help identify factors that drive some of the heterogeneity.

For any genomic feature, a pair of loci may be inside, outside or spanning the boundary of the feature. Figure \ref{fig:feature_status} illustrates the 'feature status' of some pairs of loci. Note that a pair where each loci is in a different CpG island is declared to be inside the feature, i.e., elements need not be in the same feature but rather in the same type of feature. The genomic feature should partition the genome so that each loci is either inside or outside of the feature. Since the number of 'spanning pairs' is substantially fewer than those inside or outside of the feature, and the boundaries of the feature are oftentimes fuzzy, I have excluded these 'spanning pairs' in the results shown below.

\begin{figure}[h!]
\centering
\includegraphics[width=0.8\textwidth]{../figures/feature_status.pdf}
\caption{Schematic illustrating the 'feature status' of four pairs of methylation loci. The circles represent loci and the grey bars a genomic feature, such as CpG islands. Each pair's 'feature status' is determined by whether both loci are outside of the feature (outside), one locus is inside and one outside of the feature (spanning), or both inside the feature (inside). Note that the rightmost pair is 'inside' even though the two loci that comprise the pair are in different elements of the genomic feature.}
\label{fig:feature_status}
\end{figure}

\subsubsection{Choice of correlation coefficient}

The three commonly used correlation coefficients are Pearson's $r$ (__CITE__), Spearman's $\rho$ (__CITE__) and Kendall's $\tau$ (__CITE__). Briefly, Pearson's product-moment correlation is designed to detect linear relationships between two variables. However, it is not robust to outliers, which motivates the use of Spearman's and Kendall's rank-based correlation coefficients. Spearman's $\rho$ is simply the Pearson product-moment correlation of the ranks of the data while Kendall's $\tau$ is based on the relative frequency of concordantly ranked pairs.

Owing to the size of whole-genome bisulfite-sequencing data, I have chosen to only compute Pearson's $r$ and Spearman's $\rho$, which are much faster to compute than Kendall's $\tau$. The `methLevelCor` function in the `MethylationTuples` will compute a confidence interval, $95\%$ by default, for the Pearson correlation. This feature is not yet available when using Spearman's or Kendall's correlation coefficient.

\subsubsection{Interpretation of correlation}\label{sec:interpretation_of_correlation}

Before computing correlations and trying to interpret these, it is a good idea to look at the data from which these correlations are computed. Figure \ref{fig:ADS_beta_scatterplots} shows kernel density smoothed scatterplots for pairs of $\beta$-values separated by $IPD = 2, 20, 200, 2000$ with $NIL = 0$ or $NIL \geq 0$ from the _ADS_ sample. It highlights two major problems.

Firstly, the bimodality of the distribution of $\beta$-values means that the points concentrate in the corners of the plot --- $(0, 0), (0, 1), (1, 0)$ and $(1, 1)$. A correlation coefficient is not well suited to summarising these distributions of points. Secondly, when using the $NIL = 0$ strategy, as the $IPD$ increases, the number of points in each scatterplot decreases and so the correlations are very unstable.

In summary, the correlations of pairs of $\beta$-values should be cautiously interpreted, bearing in mind that these are rather crude summaries of the underyling scatterplots that hide many details.

\begin{figure}[h!]
\centering
\includegraphics[width=\textwidth]{../figures/ADS_beta_scatterplots.pdf}
\caption{Kernel density smoothed scatterplots for pairs of CpG $\beta$-values, $(\beta_{1}, \beta_{2})$, from the \em{ADS} sample. $\beta$-values are computed using strand-collapsed counts. $n$ is the number of pairs in each scatterplot.  The Spearman correlation of each scatterplot is also reported. Individual plots created using the \texttt{smoothScatter} function in R.}
\label{fig:ADS_beta_scatterplots}
\end{figure}

\subsection{Results}

For all 40 samples in the EPISCOPE, Lister, Seisenberger and Ziller datasets, I compute the Pearson and Spearman correlations using both the $NIL = 0$ and $NIL \geq 0$ strategies. I compute these genome-wide and after stratifying CpG pairs by whether they are in a CpG island. These analyses are performed separately for each strand and also for strand-collapsed $\beta$-values.

\subsubsection{Can we collapse by strand?}

Figures \ref{fig:EPISCOPE_spearman_beta_cors_min_cov_5}, \ref{fig:Lister_spearman_beta_cors_min_cov_5}, \ref{fig:Seisenberger_spearman_beta_cors_min_cov_5} and \ref{fig:Ziller_merged_spearman_beta_cors_min_cov_5} show that the Spearman correlations of $\beta$-values for $NIL = 0$ pairs of CpGs stratified by strand. Since the correlations are very similar for the two strands, I decided to use the strand-collapsed $\beta$-values in all that follows.

\begin{figure}[h!]
\centering
\includegraphics[width=\textwidth]{../figures/EPISCOPE_spearman_beta_cors_min_cov_5.pdf}
\caption{Correlations of $\beta$-values for strand-specific pairs of CpGs with $NIL = 0$ for samples from the EPISCOPE dataset.}
\label{fig:EPISCOPE_pearson_beta_cors_min_cov_5}
\end{figure}

\begin{figure}[h!]
\centering
\includegraphics[width=\textwidth]{../figures/Lister_spearman_beta_cors_min_cov_5.pdf}
\caption{Correlations of $\beta$-values for strand-specific pairs of CpGs with $NIL = 0$ for samples from the Lister dataset.}
\label{fig:Lister_pearson_beta_cors_min_cov_5}
\end{figure}

\begin{figure}[h!]
\centering
\includegraphics[width=\textwidth]{../figures/Seisenberger_spearman_beta_cors_min_cov_5.pdf}
\caption{Correlations of $\beta$-values for strand-specific pairs of CpGs with $NIL = 0$ for samples from the Seisenberger dataset.}
\label{fig:Seisenberger_pearson_beta_cors_min_cov_5}
\end{figure}

\begin{figure}[h!]
\centering
\includegraphics[width=\textwidth]{../figures/Ziller_merged_spearman_beta_cors_min_cov_5.pdf}
\caption{Correlations of $\beta$-values for strand-specific pairs of CpGs with $NIL = 0$ for samples from the Ziller dataset.}
\label{fig:Ziller_merged_pearson_beta_cors_min_cov_5}
\end{figure}

\subsubsection{Choice of correlation coefficient}

Figures \ref{fig:EPISCOPE_pearson_vs_spearman_beta_cors_min_cov_10_strand_collapsed}, \ref{fig:Lister_pearson_vs_spearman_beta_cors_min_cov_10_strand_collapsed}, \ref{fig:Seisenberger_pearson_vs_spearman_beta_cors_min_cov_10_strand_collapsed} and \ref{fig:Ziller_merged_pearson_vs_spearman_beta_cors_min_cov_10_strand_collapsed} compares the Pearson correlation coefficient to the Spearman correlation coefficient. The two correlation coefficient give qualitatively similar results, however, the Spearman correlation is consistently smaller in magnitude than the Pearson correlation.

I have elected to use Spearman's correlation coefficient in what follows because it is more robust to outliers than Pearson's correlation \citep{Kraemer:2006te}. However, this does not remove the general limitations of using correlation coefficients (discussed in Section \ref{sec:interpretation_of_correlation}).

\begin{figure}[h!]
\centering
\includegraphics[width=\textwidth]{../figures/EPISCOPE_pearson_vs_spearman_beta_cors_min_cov_10_strand_collapsed.pdf}
\caption{Pearson correlation versus Spearman correlations of $\beta$-values for strand-collapsed pairs of CpGs with $NIL = 0$ for samples from the EPISCOPE dataset.}
\label{fig:EPISCOPE_pearson_vs_spearman_beta_cors_min_cov_10_strand_collapsed}
\end{figure}

\begin{figure}[h!]
\centering
\includegraphics[width=\textwidth]{../figures/Lister_pearson_vs_spearman_beta_cors_min_cov_10_strand_collapsed.pdf}
\caption{Pearson correlation versus Spearman correlations of $\beta$-values for strand-collapsed pairs of CpGs with $NIL = 0$ for samples from the Lister dataset.}
\label{fig:Lister_pearson_vs_spearman_beta_cors_min_cov_10_strand_collapsed}
\end{figure}

\begin{figure}[h!]
\centering
\includegraphics[width=\textwidth]{../figures/Seisenberger_pearson_vs_spearman_beta_cors_min_cov_10_strand_collapsed.pdf}
\caption{Pearson correlation versus Spearman correlations of $\beta$-values for strand-collapsed pairs of CpGs with $NIL = 0$ for samples from the Seisenberger dataset.}
\label{fig:Seisenberger_pearson_vs_spearman_beta_cors_min_cov_10_strand_collapsed}
\end{figure}

\begin{figure}[h!]
\centering
\includegraphics[width=\textwidth]{../figures/Ziller_merged_pearson_vs_spearman_beta_cors_min_cov_10_strand_collapsed.pdf}
\caption{Pearson correlation versus Spearman correlations of $\beta$-values for strand-collapsed pairs of CpGs with $NIL = 0$ for samples from the Ziller dataset.}
\label{fig:Ziller_merged_pearson_vs_spearman_beta_cors_min_cov_10_strand_collapsed}
\end{figure}

\subsubsection{$NIL = 0$ versus $NIL \geq 0$}

__UP TO HERE__

__Key points__


- NIL = 0 vs. NIL >= 0
- CGI vs. non-CGI
- periodicities

__TODO: Results__

\section{Within-fragment co-methylation}\label{subsec:wf_cometh}

The key advantage of studying within-fragment co-methylation is that measures the dependence of DNA methylation at the scale on which the biological process acts. I have developed a method based on odds-ratios to quantify the strength of co-methylation for a pair of methylation loci in a single sample.

\subsection{Methods}

Here I describe a simple method based on analysing methylation patterns at 2-tuples[^higher_order_m-tuples]. For simplicity, I will use CpG 2-tuples, but the method is applicable to any methylation type. This method is implemented in the `cometh` function that is part of the `MethylationTuples` software.


[^higher_order_m-tuples]: Extensions to m-tuples with $\text{m} > 2$ are discussed in __SECTION__.

The aim of this analysis is to address the question of 'how are methylation states on individual DNA fragments correlated as a function of the distance between loci?'. There are two steps in this analysis:

1. Define the 2-tuples.
2. For each 2-tuple, estimates the dependence between the methylation state at the first locus and the second locus. For this I use the logarithm of the odds ratio ($LOR$).

Step (2) may be further stratified by the genomic context of the pair of CpGs, such as whether they are in a CpG island, or other variables of interest. This is done in the same manner as for pairs of $\beta$-values (Section \ref{sec:pair_stratification}).

Finally, a similar analysis could be performed using a different measure of dependence, which I discuss in Section \ref{sec:choice_of_dependence_measure}.

\subsubsection{Constructing the 2-tuples}

The construction of 2-tuples is very similar to constructing the pairs of methylation levels, described in Section \ref{sec:beta_pair_construction}. However, we are now limited to analysing co-methylation of loci that can be captured within a single read. In practice, this means we are limited to studying within-fragment co-methylation for 2-tuples with $IPD \lessapprox 200$.

Again, we can employ either the $NIL = 0$ or the $NIL \geq 0$ strategy. Both strategies start with CpGs in the reference genome that, for each sample, are then filtered to remove reference-specific loci. Rather than actually remove these loci, I simply set the corresponding counts of the methylation patterns to `NA`.

As in Section \ref{subsec:beta_cor} I consider the $NIL = 0$ and $NIL \geq 0$ strategies for constructing pairs. The $NIL \geq 0$ strategy corresponds to using the `--all-combination` flag in `methtuple`. Again, the $NIL = 0$ strategy is simpler, both conceptually and computationally, and so I prefer it. For example, if we have a pair with $NIL = 3$ then we need to take care to ensure that this $LOR$ is due to a genuine dependence between these two CpGs above and beyond the pairwise dependencies $(CpG_{1}, CpG_{2}), \ldots, (CpG_{4}, CpG_{5})$.

\subsubsection{Choice of dependence measure}\label{sec:choice_of_dependence_measure}

Recall that for each 2-tuple and sample we can summarise the methylation patterns by a $2 \times 2$ contingency table (__FIGURE__). The within-fragment co-methylation of the pair of CpGs, $(i, i')$, in the $j^{th}$ sample is then defined as the logarithm of the odds ratio of this table, $LOR_{(i, i'), j} = log_2 \Big (\frac{MM_{i, i'} \times UU_{i, i'}}{MU_{i, i'} \times UM_{i, i'}} \Big)$. I use base-2 logarithms due to their widespread use and familiarity within biology. In the `cometh` function I also add a small offset (default `offset = 1`) to avoid problems arising from zero counts, which frequently occur.

A positive $LOR$ is evidence that the methylation states are likely to be the same at both CpGs in the pair on the same DNA fragment, a negative $LOR$ is evidence that the methylation states are likely to be different at each CpG in the pair on the same DNA fragment, and a value close to zero is evidence that the methylation state at each CpG in the pair on the same DNA fragment are independent of one another. More precisely, we may wish to compute a confidence interval for $LOR$ or perform a hypothesis test to determine whether the observed $LOR$ is evidence for co-methylation at this pair of CpGs.

__TODO: Comparison of LORs and Pearson correlations of Landan__

__TODO: Discussion of Kraemer's work, in particular, the fact that non-zero LOR are not informative for the strength of the correlation__

__TODO: \cite{Bonett:2007bd} is a reference that Pearson correlation of 2x2 is phi coefficient__

\subsection{Results}

$LOR$s are only computed for 2-tuples with sufficient sequencing coverage (here, $5 \times$ for stranded 2-tuples and $10 \times$ for unstranded 2-tuples). For any given $IPD$ there are hundreds of thousands of CpG pairs, and so there is a distribution of co-methylation for any given $IPD$. We could visualise the distribution for a given $IPD$ using classical visualisation techniques such as histograms, kernel density estimates and boxplots. These summaries may be stratified by one or more annotations. However, what these $IPD$-specific plots don't give the reader is a sense of how $IPD$ affects within-fragment co-methylation.

Instead, for each sample I plot selected quantiles --- $10\%, 25\%, 50\%, 75\% and 90\%$ --- of each $IPD$-specific $LOR$ distribution as a function of $IPD$. I refer to these as _quantile plots_ of $LOR$ as a function of $IPD$. These quantile plots may also be stratified by additional annotations.


Key points:

- LOR decreases with IPD
- For a given IPD, a wide variation in LOR
- LOR does not go to zero over observable distance
- Variation is fairly constant across IPD
- Noise at the end is due to limited data
- CGIs are more co-methylated than non-CGI
- SE data can't cut the mustard
- Intriguing mode/plateua around 150-200 bp (NIL >= 0)
  - SE data suggests some is technical, e.g., incomplete removal of M-bias
  - Also, IMR90 PE data doesn't have it
- LOR vs. Pearson
- NIL = 0 vs. NIL >= 0
- iPSC vs. somatic
- Plot histogram of LOR and histogram of r
- Look for tables with LOR > 0 and r = 0


An example quantile plot for the __WHICH SAMPLE__ is shown in __FIGURE__. The vast majority of $LOR$s are positive, meaning that most 2-tuples are co-methylated. The strength of co-methylation decays roughly as $IPD$ increases, which can be seen by the downward shift in the quantiles as a function of $IPD$. I

__FIGURES__ show that the two DNA strands have near identical $LOR$ distributions. This is not to discount the possibility of 2-tuples with strand-specific differences in co-methylation, however, it does mean that it is generally safe to aggregate counts across strands.

__FIGURES__


__TODO__


__OLD WRITING BELOW THIS LINE__

### Visualisations

For a typical whole-genome bisulfite-sequencing experiment of human DNA, tens of millions of neighbouring CpG pairs are sequenced (and hundreds of millions of non-neighbouring CpG pairs). For every pair of CpGs, $(i, i')$, in the $j^{th}$ sample we know the $IPD$ and have computed a log-odds ratio, $LOR$_{(i, i'), j}$. Each such pair may have further annotations, such as the genomic context of the pair (e.g. within a CpG island, within a gene body, within a repetitive region, etc.). We need a simple way to summarise these data so that we might explore the effect of each of these variables on within-fragment co-methylation.

For any given $IPD$ there are hundreds of thousands of CpG pairs and so there is a _distribution_ of co-methylation for any given $IPD$. We can visualise the distribution for a given $IPD$ using classical visualisation techniques such as histograms, kernel density estimates and boxplots (__TODO: Example figures of each of these__). These summaries may be stratified by one or more annotations (__TODO: Example figures with stratification__).

What these plots don't give the reader is a sense of how $IPD$ affects within-fragment co-methylation. __FIGURE__ and __FIGURE__ (__TODO:__ two plots of co-methylation as boxplots, one with outliers plotted and one without) show initial attempts to create a visualisation that shows the relationship between $LOR$ and $IPD$. Unfortunately, these plots are rather "busy".

Instead, I plot selected quantiles of each $IPD$-specific $LOR$ distribution as a function of $IPD$ (__FIGURE__). These convey the same information as the series of boxplots in a simpler and cleaner fashion. I refer to these as _quantile plots_ of $LOR$ as a function of $IPD$. These quantile plots may also be stratified by additional annotations (__FIGURE__).

The simplicity and interpretability of the log-odds ratio make the above described method an attractive one for studying within-fragment co-methylation. However, by using 2-tuples we are limited to exploring "one-step" co-methylation, i.e., one-step dependencies of the stochastic process $Z_{h} = (Z_{h, 1}, Z_{h, 2}, \ldots, Z_{h, N_{loci}})$. In the next section I describe the difficulties of using higher order $m-tuples$ ($m > 2$) to explore within-fragment co-methylation.

## Using higher-order $m$-tuples of methylation loci

In general, to explore $m$-step dependencies we need to use $(m + 1)$-tuples. We can generalise the use of a two-way contingency table for studying co-methylation of a 2-tuple to the use of an $m$-way contingency table for co-methylation of an m-tuple (see Table ~\ref{reads_containing_three_CpGs_counts}).

__TODO: Add real numbers to table__

\begin{table}[h]
\centering
\begin{tabular}{@{}cccc@{}}
\toprule
           &            & \multicolumn{2}{c}{Third CpG} \\ \cmidrule(l){3-4}
First CpG  & Second CpG & \textbf{M}    & \textbf{U}    \\ \midrule
\textbf{M} & \textbf{M} &      4        &      0        \\
\textbf{}  & \textbf{U} &      2        &      0        \\
\textbf{U} & \textbf{M} &      0        &      1        \\
\textbf{}  & \textbf{U} &      1        &      1        \\ \bottomrule
\end{tabular}
\caption{Example three-way table summarising read counts at a 3-tuple.}
\label{reads_containing_three_CpGs_counts}
\end{table}

     chr strand pos1 pos2 pos3 MMM MMU MUM MUU UMM UMU UUM UUU cov
 4: chr1      +  489  493  497   4   0   2   0   0   1   1   0   8


However, as we increase $m$ we have more parameters to estimate - an $m$-way table has $\binom{m}{2}$ two-way dependencies, $\binom{k}{3}$ three-way dependencies, $\ldots$, and $\binom{m}{m} = 1$ $m$-way dependency.

Furthermore, as we increase $m$ we have fewer data points with which to estimate these parameters, since we only use reads that contain all $m$ methylation loci in the $m$-tuple (__TODO: Bar chart of how many CpGs per read__). This makes the estimates quite noisy unless some form of regularisation is applied, such as a parametric model of the dependencies or a non-parametric, data-driven smoothing of the estimates. I explore the use of the Expectation-Maximisation algorithm in __SECTION__ to incorporate all reads that overlap at least one methylation loci in the $m$-tuple.

### An EM algorithm to estimate co-methylation

__TODO:__ Discuss with Terry the idea of fitting a log-linear model to the counts of methylation states using all reads that overlap at least one methylation loci in the tuple. Effectively, this is a log-linear model with missing data.

### Visualisations

Aside from estimating this increased number of parameters, we must also interpret these with respect to $IPD$ and other annotations. The $IPD$ vector for $m$-tuples has $(m - 1)$ elements meaning we now have to explore the effect of $IPD$ on $LOR$ over an $(m - 1)$-dimensional _grid_ rather than along a 1-dimensional line. Again, the sparsity of the data make this difficult - approximately $20,000 - 30,000$ unique $IPD$s are observed for 3-tuples - but even more difficult is creating a clear and coherent visualisation of this data.

As is clear from __FIGURE__, these is very difficult for 3-tuples, let alone for higher-order $m$-tuples. We have an $(m - 1)$ grid of $IPD$s and a _distribution_ of $LOR$ for each $IPD$. In the case of 2-tuples I used lines to summarise the quantiles of the $LOR$ distributions, but this is not effective when $m > 2$. Instead, we must create separate figures for each quantile, which makes it difficult to interpret the effects of $IPD$ on $LOR$.


## Results of co-methylation analyses


## A Markov model of within-fragment co-methylation

__Possibly better off in the chapter on simulation__

A $k$-order Markov model of within-fragment co-methylation assumes that the dependence of $Z_{h, i}$ on the "history" of the process, $(Z_{h, 1}, \ldots, Z_{h, i - 1})$ is through the $k$ most recent observations, $Z_{h, i - k}, \ldots, Z_{h, i - 1}$. This could be generalised via a Markov random field model of within-fragment co-methylation to assume that the dependence of $Z_{h, i}$ on the "rest" of the process $Z_{h, 1}, \ldots, Z_{h, i - 1}, Z_{h, i + 1}, Z_{N_{loci}}$ is through the k-nearest observations, $Z_{h, i - k_{1}}, \ldots, Z_{h, i - 1}, Z_{h, i + 1}, \ldots, Z_{h, i + k_{2}}$, where $k_{2} - k_{1} = k$.

Regardless of whether we use a Markov chain or a Markov random field, a complication is that $Z_{h}$ is unevenly spaced along the genome. I use the intra-pair distance ($IPD$) to measure the distance between observations along $Z_{h}$. This distance measure introduces additional dimension(s) to our model.
