---
title: "Co-methylation"
author: "Peter Hickey"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  pdf_document:
    fig_caption: yes
    includes:
      in_header: ../latex/header.tex
  html_document:
    keep_md: true
bibliography: ../latex/phd_thesis.bib
---

## Chapter overview

Co-methylation is the dependence of DNA methylation at different methylation loci within a sample. As such, it is a kind of spatial dependence where the "space" is the genome of the sample. In this chapter I describe two measures of co-methylation, namely, _within-fragment co-methylation_ and _correlation of $\beta$-values_. Using both these measures I explore how co-methylation is affected by the distance between methylation loci, the density of methylation loci and the genomic context of the methylation loci.

I begin by explaining various methods for estimating co-methylation and their associated challenges. I then use these methods to analyse more than fifty whole-genome bisulfite-sequencing libraries to study the variation of co-methylation within and between different cell types.

Throughout this chapter I focus on co-methylation of CpGs, although the methods described are also applicable to non-CpG methylation loci. The reason I do not explore co-methylation at non-CpG methylation loci is because such loci are rarely methylated in the samples that I am studying.

\section{Previous work}\label{subsec:previous work}

I define co-methylation as the dependence of DNA methylation at different methylation loci within a sample. In other words, it is the dependence or correlation structure of DNA methylation data within a sample. This is slightly different to \cite{Schatz:2004cha} who defined it as, "the presence of methylation over a stretch of neighbouring CpG positions"; it is more in line with \cite{Eckhardt:2006gh} who defined it as, "the relationship between the degree of methylation over distance".

There have been several attempts to explore the relationship between methylation events at different loci within an individual. However, frequently the exact question being asked and the methodology have not been well-defined. Moreover, to the best of my knowledge, there is no software available that implements any of the proposed methods.

From a biological perspective, by studying co-methylation we hope to better understand the regulation of DNA methylation. From a statistical perspective, more powerful and efficient analysis methods can be developed if this correlation structure is better understood, accounted for and exploited.

An increased understanding of co-methylation could also lead to cheaper and more efficient assays of DNA methylation. For example, we could identify methylation loci whose methylation status is highly predictive of the methylation level of a surrounding region and design an assay to interrogate those highly predictive loci. This idea is analogous to the use of tag-SNPs on whole-genome genotyping microarrays (__CITE__).

### Correlations of aggregate measurements of methylation

To date, most analyses of co-methylation have been based on aggregate measurements of methylation, such as $\beta$-values \cite{Eckhardt:2006gh, Cokus:2008fc, Li:2010fb, Lyko:2010dr, Hebestreit:2013ko, Lacey:2013iy, Liu:dy}. This is true of studies using both microarray data, where only these type of measurements are available, and sequencing data, where base-resolution measurements are also available. Very often the exact analysis is not well-described but it can generally be thought of as studying the autocorrelation of the $\beta$-values along the genome.

#### \cite{Eckhardt:2006gh}

\cite{Eckhardt:2006gh} is an important paper in the field and has been cited as evidence that "methylation levels are strongly spatially correlated" \cite{Hebestreit:2013ko} and that "proximal CpGs [have] similar methylation levels" \cite{Hansen:2011gu}. \cite{Eckhardt:2006gh} performed  PCR-amplicon bisulfite-sequencing of 2,524 amplicons on human chromosomes 6, 20 and 22.  PCR-amplicon bisulfite-sequencing is a labour-intensive assay and so it was a considerable effort to produce these data. Nonetheless, it is important to bear in mind that these data are nowhere near as comprehensive as data generated using modern-day whole-genome bisulfite-sequencing assays. Samples were provided by male and female donors across a wide variety of tissue types and ages, and were were generated to begin a catalogue of CpG methylation in various tissue types.

As part of their analysis, the authors analyse co-methylation, which they define as "the relationship between the degree of methylation over distance". They state that "we were able to establish a significant correlation for comethylation over short distances ($\leq 1,000$ bp), it deteriorated rapidly for distances $ >2,000$ bp" and reference Figure 3a in support of this claim. However, what is plotted in Figure 3a is not "correlation" in the usual sense of the word.

In Figure 3a, the authors plot the "percentage identical methylation" against "distance in bp", which ranges from $0$ to $20,000$ bp (__TODO: Check permission on re-priting this figure__). The average amplicon length is reported as $411$ bp with a standard deviation of $71$ bp. Therefore, it is not possible that the "percentage identical methylation" was obtained from methylation events co-occuring on the same amplicon, at least for those pairs separated by more than approximately $600$ bp. Nowhere in the original paper is it clearly described what the "percentage identical methylation" is nor how it is computed. An email discussion with the senior author, Stephan Beck, was unable to clarify the definition.

My best guess as to the definition of "percentage identical methylation" is that it is the proportion of pairs of methylation calls at different CpGs from (generally) different amplicons where the methylation states are identical, that is, $Pr(z_{h, i} = z_{h', i'})$. I will refer to my best guess of "percentage identical methylation" as $PIM^{*}$, and now turn to some of its properties.

Let $0 \leq p_{h} \leq 1$ be the genome-wide average CpG methylation level for the sample from which $z_{i, h}$ is sampled, that is, $Pr(z_{h, i} = 1) = p_{h}$. For two distinct CpGs sequenced on two (potentially) distinct amplicons, $z_{h, i}$ and  $z_{h', i'}$, we want to compute the probability that the methylation states are identical, that is, $Pr(z_{h, i} = z_{h', i'})$. Not only may the two CpGs be from distinct amplicons, but the amplicons themselves may be from two distinct samples with different genome-wide average methylation levels. This most general setting is described by $z_{h, i} \eqd \text{Bernoulli}(p_{h})$ and $z_{h, i'} \eqd \text{Bernoulli}(p_{h'})$, with $cor(z_{h, i}, z_{h', i'}) = c$. In this setting we can compute the theoretical $PIM^{*}$ as follows:

\begin{align*}
PIM^{*} &= Pr(z_{h, i} = z_{h', i'}) \\
        &= Pr(z_{h, i} = 0, z_{h', i'} = 0) + Pr(z_{h, i} = 1, z_{h', i'} = 1) \\
        &= Pr(z_{h', i'} = 0 | z_{h, i} = 0) Pr(z_{h, i} = 0) +
           Pr(z_{h', i'} = 1 | z_{h, i} = 1) Pr(z_{h, i} = 1) \\
        &= \big [ (1 - p_{h'}) + c (1 - p_{h}) \big ] (1 - p_{h}) +
           \big [ (p_{h'} + c (1 - p_{h})) \big ] p_{h} \\
        &= (1 - p_{h}) (1 - p_{h'}) + p_{h} p_{h'} + 2 c p_{h} (1 - p_{h})
\end{align*}

This result follows by substituting $z_{h, i}$ and $p_{h}$ (resp. $z_{h', i'}$ and $p_{h'}$) for $X_{1}$ and $p_{1}$ (resp. $X_{2}$ and $p_{2}$) into the result derived in __TODO: Cite and link to appendix__.

__TODO: What is the general restriction on $c$ when $p_{h, i} \neq p_{h', i'}$?__

In general, $PIM^{*} \in [0, 1]$. However, there are a few important special cases that are worth discussing:


1. $PIM^{*}_{inid} = (1 - p_{h}) (1 - p_{h'}) + p_{h} p_{h'} \in [0, 1]$: Independent and non-identically distributed (_inid_), i.e., $z_{h, i} \eqd Bernoulli(p_{h})$ and $z_{h, i'} \eqd Bernoulli(p_{h'})$ and with $cor(z_{h, i}, z_{h', i'}) = c = 0$.
2. $PIM^{*}_{did} = (1 - p)^{2} + p^{2} + 2 c p ( 1 - p)$: Dependent and identically distributed (_did_), i.e., $z_{h, i}, z_{h', i'} \eqd Bernoulli(p)$ and $cor(z_{h, i}, z_{h', i'}) = c \neq 0$.
3. $PIM^{*}_{iid} = (1 - p)^{2} + p^{2}$: Independent and identically distributed (_iid_), i.e., $z_{h, i}, z_{h', i'} \eqd Bernoulli(p)$ and $cor(z_{h, i}, z_{h', i'}) = c = 0$.

As for $PIM^{*}$, $PIM^{*}_{inid} \in [0, 1]$. In contrast, it is easy to show by differentiation that $PIM^{*}_{iid} \geq 0.5$ and that $PIM^{*}_{did} \geq 0.5 (c + 1)$, which is $> 0.5$ (resp. $< 0.5$) for $c > 0$ (resp. $c < 0$). __TODO: Illustrate these special cases with a plot.__

To return to the plot shown in Figure 3a; the "percentage identical methylation" asymptotes at $40 \%$, or $0.4$, which is less than the theoretical lower-bound of $PIM^{*}_{iid}$. If we believed that all samples had the same average CpG methylation levels, $p_{h}$, then we would conclude that the methylation states of two CpGs separated by a large distance are in fact __anti-correlated__ ($c < 0$) for CpGs separated by approximately $1,000$ bp. However, since we know that there is sample-to-sample variability in the average level of CpG methylation, the more likely explanation for this result is that the $p_{h}$ vary between samples and therefore $PIM^{*} < 0.5$ are possible even supposing $c > 0$.

Of course, the above discussion pre-supposes that what is labelled "percentage identical methylation" in Figure 3a is mathematically equivalent to $PIM^{*}$. I believe this to be true but due to the aforementioned deficiencies of the method description it is not possible to verify this derivation.

Quite apart from what exactly is plotted in Figure 3a, the analysis is based on data from only $2,524$ amplicons, which is nowhere near a genome-wide analysis. Furthermore, methylation levels have been aggregated across multiple samples and so the profile in Figure 3a does not represent any one sample but a pool of samples.

#### \cite{Cokus:2008fc}

\cite{Cokus:2008fc} published the BS-seq protocol for performing whole-genome bisulfite-sequencing. They demonstrated this technique by generating a genome-wide map of DNA methylation in wildtype  _Arabidopsis Thaliana_. They performed an autocorrelation analysis of the $\beta$-values for CG, CHG and CHH methylation[^arabidopsis]. The autocorrelation computes, for a set of distances ($0-5,000$ nt), the (Pearson) correlation coefficient of methylation levels for all pairs of methylation loci separated by each distance. To emphasise, I believe this analysis uses all pairs of methylation loci, that is, regardless of how many intervening methylation loci there are for each pair.

[^arabidopsis]: Many plant species, such as _A. thaliana_, have much higher levels of non-CG methylation than do mammals.

\cite{Cokus:2008fc} found that methylation levels are highly correlated, particularly for CG methylation levels, and that the correlation decays as a function of genomic distance between methylation loci. From this they concluded that the "significant correlation between methylated cytosines for distances up to 5,000 nucleotides or more [were] probably a reflection of regional foci of methylation throughout the genome and of large blocks of pericentromeric heterochromatin". They also identified correlations between different methylation contexts, e.g. CG vs. CHG, which "suggest[s] complex interactions between the different types of methylation".

The autocorrelation plots showed two periodicities, one of $\approx 167$ nucleotides and the other of $10$ nucleotides (CHH only). Both periodicities are visible in plots of the raw autocorrelations and are confirmed by a Fourier transform analysis of the autocorrelation.

[^CHH_only]: The $10$ nucleotide periodicity was later observed for CG and CHG methylation as well, at least for nucleosome bound DNA \cite{Chodavarapu:2010iqa}

The $167$ nucleotide periodicity is "is similar to, but slightly shorter than, estimates of the average spacing of nucleosomes in plant chromatin" and Cokus et al. hypothesise that this periodicity is due to histones dictating access to DNA by the DNA methyltransferases.

The $10$ nucleotide periodicity is equal to the length of one helical DNA turn. From this result Cokus et al. hypothesise that the DOMAINS REARRANGED METHYLASE 2 (_DRM2_), "the main enzyme controlling asymmetric methylation in _Arabidopsis_", contains two active sites that methylate sites 8-10 apart[^comethylation_def2], like its mammalian homologue, DNA methyltransferase 3 (_DNMT3_) (__TODO: If I want a picture explaining this then Figure 3b from \cite{Jurkowska:2011dt} might form the basis for my own illustration). However, in a later paper by the same group, this $10$ nucleotide periodicity was also observed for CG and CHG methylation, at least for nucleosome bound DNA \cite{Chodavarapu:2010iqa}.

[^comethylation_def2]: This is referred to as "co-methylation", here meaning the methylation of multiple loci in a single event, by \cite{Jurkowska:2011dt}. This article reviews a plausible biochemical model for how _DNMT3a_ could co-methylate two CpGs separated by $8-10$ nucleotides; this model includes both the scenario where both CpGs are on the same DNA strand and the scenario where the CpGs are on opposite strands.

The $10$ nucleotide CHH-only periodicity in the autocorrelation analysis of the $\beta$-values is also observed in a _within-fragment analysis_ (discussed in \ref{sec:previous_within-fragment_analyses})

#### \cite{Chodavarapu:2010iqa}

\cite{Chodavarapu:2010iqa} use MNase-seq to study nucleosome positioning and the relationship to DNA methylation, including co-methylation, in _Arabidopsis_ and an embryonic human stem cell line (HSF1). The authors plot a "weighted average methylation", which I presume to be a weighted version of the $\beta$-values, against the distance from predicted nucleosome start sites, which are obtained via MNase-seq. The nucleosome start sites anchor the methylation observations to a common grid and is used to show that nucleosome bound DNA has a greater level of methylation than DNA that is not bound to nucleosomes. A $10$ nucleotide periodicity is readily observed in these plots, consistent with the discovery of $10$ nucleotide periodicities in the autocorrelation of $\beta$-values published by \cite{Cokus:2008fc}. Again, the observed periodicity is confirmed by analysing a Fourier transform of these plots.

In addition to the genome-wide analyses in both _Arabidopsis_ and human embryonic stem cell line (HSF1), the same method is used to examine promoters, genes, repeats, euchromatic regions and centromeric regions of _Arabidopsis_ and promoters, genes, repeats and CpG islands of the HSF1 sample, which confirmed the $10$ nucleotide periodicity. This led \cite{Chodavarapu:2010iqa} to propose an extended version of their earlier hypothesis \cite{Cokus:2008fc}: namely, "that nucleosomes are to some extent dictating access to the DNA and therefore setting the register of methylation __for all DNA methyltransferases__" (emphasis mine) and not just for the DRM2 methyltransferase.

#### \cite{Li:2010fb}

\cite{Li:2010fb} compute "the correlation of methylation level of any two nearby CpGs and the relationship between spatial distance (from one CpG to another) and strength of this correlation". The authors conclude that "co-methylation deteriorates over distance and becomes nearly undetectable at distances $> 1,000$ bp", citing their results as being consistent with those in \cite{Eckhardt:2006gh}. In addition to this genome-wide analysis, \cite{Li:2010fb} perform the same analysis for CpGs within 19 different genomic contexts such as upstream of a gene, untranslated region, coding DNA sequences and various repetitive elements. Somewhat surprisingly, they do not specifically compare CpG islands to non CpG islands, although some of these differences will be captured by the analysis of regions upstream of a gene, which include many CpG island promoters.

Unfortunately, just as in \cite{Eckhardt:2006gh}, \cite{Li:2010fb} never define exactly what is being computed in their analysis of co-methylation and no software is available. All the co-methylation results are presented as supplementary figures, which lack figure legends and are supported by only very brief captions. So, once again, I have been forced to make my best guess as to what these figures show and how it was computed.

Supplementary Figure 6a plots the "correlation" against "distance of CpG" ($0 - 1000$), which I believe are the genome-wide co-methylation estimates over the range $0 - 1000$ bp. Based on the "$n = 18,936,995$" quoted in the header of this figure, which I interpret as the number of pairs used in this plot of "correlation" against "istance of CpG" ($0 - 1000$), I believe that what is shown are the correlations of pairs of CpG $\beta$-values where the pairs are separated by a common distance and have no intervening CpGs. If all pairs were used this $n$ would be much higher since there are more than twenty million CpGs on one strand of the haploid human genome. To emphasise, the correlation shown by \cite{Li:2010fb} is not the same as the autocorrelation computed in \cite{Cokus:2008fc} since the autocorrelation uses all pairs of CpGs, regardless of the number of intervening CpGs.

The genome-wide analysis correlation plot contains a periodicity of approximately $170$ nucleotides, very similar to the $167$ nucleotide period identified in _Arabidopsis_ by \cite{Cokus:2008fc} (highlighted in Supplementary Figure 7). This is further evidence that nucleosome spacing effects the strength of co-methylation. However, \cite{Li:2010fb} found no evidence of higher frequency periods, such as the $10$ nucleotide period reported by \cite{Cokus:2008fc} and \cite{Chodavarapu:2010iqa}.

#### \cite{Lyko:2010dr}

In their study of differential methylation in queen and worker honey bees (_Apis mellifera_), \cite{Lyko:2010dr} perform an autocorrelation analysis of CpG methylation (Supplementary Figure 7). They report that the "correlation of methylation status of neighboring CpGs increases sharply between 1 bp and 20 bp, then drops rapidly between 40 bp and 100 bp, and then slowly fades away". This is the first report of an organism in which the correlation of methylation levels first increases before decreasing (__TODO: Is this the "first and only" such report? Check my data any such increases__). __TODO: Is this the fastest decay of correlation in any species?__ However, I am concerned that the maximum autocorrelation is less $0.015$, which doesn't seem to make sense and is very close to zero. This may simply be due to a mislabelled or "normalised" value being plotted as the "correlation". (__TODO: I have emailed the senior author, Ryszard Maleszka, to try to resolve this concern__).

Additionally, \cite{Lyko:2010dr} identify a $3$ nucleotide periodicity in the autocorrelation of CpG methylation and provide evidence that this "reflect[s] a preferential methylation of CpGs occupying the first and second position of the arginine codons", which is coded for by the DNA triplets CGT, CGC, CGA, CGG, AGA, and AGG (__CITE__). They do not detect a $10$ nucleotide periodicity, which the authors claim is due to "a much lower density of modified CpGs in this species [_Apis mellifera_]" than in human or _Arabidopsis_ genomes.

Since the authors specifically describe their analysis as an "autocorrelation", I believe it uses all pairs of CpGs regardless of the number of intervening CpGs.

#### Other co-methylation analyses using aggregate measurements

* \cite{Hebestreit:2013ko, Lacey:2013iy, Liu:dy}



\paragraph{Previous within-fragment co-methylation analyses}
\label{sec:previous_within-fragment_analyses}

* More recently, within-fragment analyses. Describe concept (move from next section to this section).
* Cokus

#### Other definitions

* Lister's mCG correlations
* co-methylation across samples
* "vertical" correlations of \cite{Capra:2014uh}

The method I developed, which is described in Section \ref{subsec:wf_cometh}, is superior because of __TODO__.


\subsection{Within-fragment co-methylation}\label{subsec:wf_cometh}

Within-fragment co-methylation is the dependence of DNA methylation at methylation loci that occur on the same DNA fragment. It can be thought of as the dependence structure of the binary stochastic process $Z_{h} = (Z_{h, 1}, Z_{h, 2}, \ldots, Z_{h, N_{loci}})$. For example, we may wish to understand how $Z_{h, i}$ affects $Z_{h, i'}$ and to develop a simple, tractable model of this dependence structure. We can study what variables effect this dependence, such as the distance between methylation loci and the genomic context of the methylation loci.

As every sequencing read is produced from a single DNA fragment[^chimera], we can study within-fragment co-methylation by analysing the methylation patterns along individual sequencing reads, that is, by studying the methylation patterns at m-tuples. Furthermore, since the DNA methyltransferases act on individual DNA fragments (__TODO: check and cite__), these estimates of within-fragment co-methylation are on the same scale as the physical process that lays down and maintains DNA methylation.

[^chimera]: Here I ignore the possibility of chimeric reads that may be artificially produced by DNA fragments ligating to one another during the library preparation or sequencing process (__TODO: reference__).

In studying within-fragment co-methylation, I assume that all copies of each chromosome have the same dependence structure (__TODO: Although not the same mean structure? Make this mathematically precise__). The data are sampled from a population containing multiple copies of each chromosome and if each copy has a different dependence structure then we cannot hope to say anything about these structures since these copies are not identifiable.

If we could sequence entire chromosomes by single reads then we could analyse co-methylation between any set of methylation loci. However, current technologies for bisulfite-sequencing mean that we are limited to studying reads that, at most, span about $200-250$ bp and so are limited to studying co-methylation of methylation loci that are very close to one another along the chromosome. Obviously, we cannot study within-fragment co-methylation from microarray data since these produce measurements of methylation that are aggregated across many hundreds or thousands of cells.

\subsubsection{Using pairs of methylation loci}\label{subsec:pairs}

The simplest summary of within-fragment co-methylation uses pairs (2-tuples) of methylation loci. Consider a single pair of CpGs in the reference genome and all reads that contain both CpGs of the pair (see Figure \ref{reads_containing_pair_of_CpGs}).

\begin{figure}[h!]
  \centering
    \includegraphics[width=0.5\textwidth]{../figures/reads_containing_pair_of_CpGs.pdf}
      \caption{All reads containing a pair of CpGs. The CpGs are highlighted by the short blue bars along the bottom of the figure. Each long grey (resp. green) bar is a read mapped to the forward (resp. reverse) strand. Those reads highlighted in yellow contain both CpGs of the pair. Data are from the IMR90-iPSC sample.}
      \label{reads_containing_pair_of_CpGs}
\end{figure}

We tabulate the methylation states of each of these reads as shown in Table \ref{reads_containing_pair_of_CpGs_counts}. The `methtuple` software can perform this tabulation for all pairs of methylation loci in a sample.

The within-fragment co-methylation of the pair of CpGs, $(i, i')$, in the $j^{th}$ sample is then defined as the log-odds ratio of this table, $LOR_{(i, i'), j} = log_2 \Big (\frac{MM_{i, i'} \times UU_{i, i'}}{MU_{i, i'} \times UM_{i, i'}} \Big)$. For the data shown in Figure \ref{reads_containing_pair_of_CpGs} the $LOR =$ `r round(log2((1.5 * 4.5) / (0.5 * 2.5)), 1)`.

\begin{table}[h]
  \centering

  \begin{tabular}{lcccl}
                           & \multicolumn{1}{l}{} & \multicolumn{2}{l}{Second CpG} &                \\
                           &                      & \textbf{M}     & \textbf{U}    & \textbf{Total} \\ \cline{2-5}
\multirow{2}{*}{First CpG} & \textbf{M}           & 1              & 2             & 3              \\
                           & \textbf{U}           & 0              & 4             & 4              \\ \cline{2-5}
                           & \textbf{Total}       & 1              & 6             & 7
\end{tabular}
\caption{Counts of each methylation pattern for reads shown in Figure \ref{reads_containing_pair_of_CpGs}}
\label{reads_containing_pair_of_CpGs_counts}
\end{table}

A positive $LOR$ is evidence that the methylation states are likely to be the same at both CpGs in the pair on the same DNA fragment (_co-methylation_), a negative $LOR$ is evidence that the methylation states are likely to be different at each CpG in the pair on the same DNA fragment and a value close to zero is evidence that the methylation state at each CpG in the pair on the same DNA fragment are independent of one another. More precisely, we may wish to compute a confidence interval for $LOR$ or perform a hypothesis test to determine whether the observed $LOR$ is evidence for co-methylation at this pair of CpGs.

In theory we can compute $LOR$ for all pairs of methylation loci. In practice, using current technology, we can only compute it for pairs of methylation loci that are at most separated by 200-250 bp.

#### Comparison of LOR to correlations

__TODO: Compare use of log-odds ratio to use of correlation coefficient and justify using LOR__

#### Constructing pairs of methylation loci

There are additional considerations in constructing pairs of methylation loci on top of those constraints imposed by the sequencing technology.

For a sequence containing $k$ methylation loci there are $\binom{k}{2}$ pairs, which is $\mathcal{O}(k^{2})$ as $\lim_{k \to \infty}$. In contrast, there are $k - 1$ pairs of _neighbouring_ methylation loci (i.e., pairs where the _number of intervening methylation loci_ equal to zero, $NIML = 0$), which is $\mathcal{O}(k)$ as $\lim_{k \to \infty}$.

Apart from this difference in the number of pairs, the interpretation of co-methylation at neighbouring loci is far simpler than for non-neighbouring loci and so it may be preferable to focus on neighbouring loci. For example, if we observe $LOR >> 0$ for a pair of CpGs, $(CpG_{1}, CpG_{5})$, with $NIML = 3 > 0$ then we need to take care to ensure that this is due to a genuine dependence between these two CpGs above and beyond the pairwise dependencies $(CpG_{1}, CpG_{2}), \ldots, (CpG_{4}, CpG_{5})$.

The `methtuple` software by default tries to only construct pairs of neighbouring methylation loci[^all-combinations]. However, because `methtuple` processes each read independently, filters methylation loci on user-specified criteria and is unaware of the reference genome, it will occasionally create pairs that are not truly neighbouring. Therefore, if strictly neighbouring pairs are required, non-neighbouring pairs should be post-hoc filtered (__TODO: e.g. using the `whatFunction` in `myRMethTuplePackage`__).

[^all-combinations]: `methtuple` also provides the option `--all-combinations` to specifically create all $\binom{k}{m}$ $m$-tuples. However the computational requirements are generally prohibitive for $m > 2$.

Furthermore, while CpG methylation is typically strand-symmetric, it may be preferable to compute co-methylation separately for each strand. This will double the number of CpG 2-tuples. By default, `methtuple` does not collapse CpGs across strands, but this option is available via the `--strand-collapse` option or as a post-hoc step in __TODO: e.g. using the `whatFunction` in `myRMethTuplePackage`__.

#### Visualisations

For a typical whole-genome bisulfite-sequencing experiment of human DNA, tens of millions of neighbouring CpG pairs are sequenced (and hundreds of millions of non-neighbouring CpG pairs). For every pair of CpGs, $(i, i')$, in the $j^{th}$ sample we know the $IPD$ and have computed a log-odds ratio, $LOR$_{(i, i'), j}$. Each such pair may have further annotations, such as the genomic context of the pair (e.g. within a CpG island, within a gene body, within a repetitive region, etc.). We need a simple way to summarise these data so that we might explore the effect of each of these variables on within-fragment co-methylation.

For any given $IPD$ there are hundreds of thousands of CpG pairs and so there is a _distribution_ of co-methylation for any given $IPD$. We can visualise the distribution for a given $IPD$ using classical visualisation techniques such as histograms, kernel density estimates and boxplots (__TODO: Example figures of each of these__). These summaries may be stratified by one or more annotations (__TODO: Example figures with stratification__).

What these plots don't give the reader is a sense of how $IPD$ affects within-fragment co-methylation. __FIGURE__ and __FIGURE__ (__TODO:__ two plots of co-methylation as boxplots, one with outliers plotted and one without) show initial attempts to create a visualisation that shows the relationship between $LOR$ and $IPD$. Unfortunately, these plots are rather "busy".

Instead, I plot selected quantiles of each $IPD$-specific $LOR$ distribution as a function of $IPD$ (__FIGURE__). These convey the same information as the series of boxplots in a simpler and cleaner fashion. I refer to these as _quantile plots_ of $LOR$ as a function of $IPD$. These quantile plots may also be stratified by additional annotations (__FIGURE__).

The simplicity and interpretability of the log-odds ratio make the above described method an attractive one for studying within-fragment co-methylation. However, by using 2-tuples we are limited to exploring "one-step" co-methylation, i.e., one-step dependencies of the stochastic process $Z_{h} = (Z_{h, 1}, Z_{h, 2}, \ldots, Z_{h, N_{loci}})$. In the next section I describe the difficulties of using higher order $m-tuples$ ($m > 2$) to explore within-fragment co-methylation.

### Using higher-order $m$-tuples of methylation loci

In general, to explore $m$-step dependencies we need to use $(m + 1)$-tuples. We can generalise the use of a two-way contingency table for studying co-methylation of a 2-tuple to the use of an $m$-way contingency table for co-methylation of an m-tuple (see Table ~\ref{reads_containing_three_CpGs_counts}).

__TODO: Add real numbers to table__

\begin{table}[h]
\centering
\begin{tabular}{@{}cccc@{}}
\toprule
           &            & \multicolumn{2}{c}{Third CpG} \\ \cmidrule(l){3-4}
First CpG  & Second CpG & \textbf{M}    & \textbf{U}    \\ \midrule
\textbf{M} & \textbf{M} &      4        &      0        \\
\textbf{}  & \textbf{U} &      2        &      0        \\
\textbf{U} & \textbf{M} &      0        &      1        \\
\textbf{}  & \textbf{U} &      1        &      1        \\ \bottomrule
\end{tabular}
\caption{Example three-way table summarising read counts at a 3-tuple.}
\label{reads_containing_three_CpGs_counts}
\end{table}

     chr strand pos1 pos2 pos3 MMM MMU MUM MUU UMM UMU UUM UUU cov
 4: chr1      +  489  493  497   4   0   2   0   0   1   1   0   8


However, as we increase $m$ we have more parameters to estimate - an $m$-way table has $\binom{m}{2}$ two-way dependencies, $\binom{k}{3}$ three-way dependencies, $\ldots$, and $\binom{m}{m} = 1$ $m$-way dependency.

Furthermore, as we increase $m$ we have fewer data points with which to estimate these parameters, since we only use reads that contain all $m$ methylation loci in the $m$-tuple (__TODO: Bar chart of how many CpGs per read__). This makes the estimates quite noisy unless some form of regularisation is applied, such as a parametric model of the dependencies or a non-parametric, data-driven smoothing of the estimates. I explore the use of the Expectation-Maximisation algorithm in __SECTION__ to incorporate all reads that overlap at least one methylation loci in the $m$-tuple.

#### An EM algorithm to estimate co-methylation

__TODO:__ Discuss with Terry the idea of fitting a log-linear model to the counts of methylation states using all reads that overlap at least one methylation loci in the tuple. Effectively, this is a log-linear model with missing data.

#### Visualisations

Aside from estimating this increased number of parameters, we must also interpret these with respect to $IPD$ and other annotations. The $IPD$ vector for $m$-tuples has $(m - 1)$ elements meaning we now have to explore the effect of $IPD$ on $LOR$ over an $(m - 1)$-dimensional _grid_ rather than along a 1-dimensional line. Again, the sparsity of the data make this difficult - approximately $20,000 - 30,000$ unique $IPD$s are observed for 3-tuples - but even more difficult is creating a clear and coherent visualisation of this data.

As is clear from __FIGURE__, these is very difficult for 3-tuples, let alone for higher-order $m$-tuples. We have an $(m - 1)$ grid of $IPD$s and a _distribution_ of $LOR$ for each $IPD$. In the case of 2-tuples I used lines to summarise the quantiles of the $LOR$ distributions, but this is not effective when $m > 2$. Instead, we must create separate figures for each quantile, which makes it difficult to interpret the effects of $IPD$ on $LOR$.

## Correlations of $\beta$-values





## Results of co-methylation analyses

__TODO__

## A Markov model of within-fragment co-methylation

__Possibly better off in the chapter on simulation__

A $k$-order Markov model of within-fragment co-methylation assumes that the dependence of $Z_{h, i}$ on the "history" of the process, $(Z_{h, 1}, \ldots, Z_{h, i - 1})$ is through the $k$ most recent observations, $Z_{h, i - k}, \ldots, Z_{h, i - 1}$. This could be generalised via a Markov random field model of within-fragment co-methylation to assume that the dependence of $Z_{h, i}$ on the "rest" of the process $Z_{h, 1}, \ldots, Z_{h, i - 1}, Z_{h, i + 1}, Z_{N_{loci}}$ is through the k-nearest observations, $Z_{h, i - k_{1}}, \ldots, Z_{h, i - 1}, Z_{h, i + 1}, \ldots, Z_{h, i + k_{2}}$, where $k_{2} - k_{1} = k$.

Regardless of whether we use a Markov chain or a Markov random field, a complication is that $Z_{h}$ is unevenly spaced along the genome. I use the intra-pair distance ($IPD$) to measure the distance between observations along $Z_{h}$. This distance measure introduces additional dimension(s) to our model.
