---
title: "Concluding remarks"
output:
  pdf_document:
    fig_caption: yes
    includes:
      in_header: ../latex/header.tex
  html_document:
    keep_md: true
bibliography: ../latex/phd_thesis.bib
---

\chapter{Concluding remarks}\label{chap:concluding_remarks}

This thesis has examined the statistical and computational challenges raised by high-throughput assays of DNA methylation. The first chapter introduces DNA methylation and then extensively reviews different high-throughput assays for measuring it, focusing on the current gold standard assay of whole-genome bisulfite-sequencing.

The second chapter details the bioinformatics analysis of whole-genome bisulfite-sequencing data. Beginning with the `FASTQ` files created by the DNA sequencers, we review the quality control procedures, read mapping and post-processing essential to creating a `BAM` file creating high-quality mapped reads. We then describe methods for calling methylation events from bisulfite-sequencing data. The chapter concludes by introducing m-tuples and developing the `methtuple` software that can, uniquely, call methylation patterns at m-tuples from whole-genome bisulfite-sequencing data.

The third chapter describes the 40 whole-genome bisulfite-sequencing samples that are analysed throughout the remainder of the thesis.

Chapter \ref{chap:wgbs_statistical_framework} lays out a statistical framework for analysing bisulfite-sequencing data. This process illustrates that the commonly used summary of methylation levels, the $\beta$-value, aggregates over many potential sources of variation. In light of this, we discuss the interpretation of $\beta$-values and examine the distributions of $\beta$-values in our 40 samples.

Chapter \ref{chap:wgbs_downstream_analyses} examines methods for analysing bisulfite-sequencing data to address key biological questions. Most downstream analyses are based on $\beta$-values, such as the identification of differential methylation, and we review these. There is, however, a growing interest in analyses based on methylation patterns at m-tuples since these can address novel biological questions. A limitation to the development of these methods has been a lack of software. To complement to the `methtuple` software, the chapter concludes with the development of the `MethylationTuples` R package, which is designed to facilitate analyses based on methylation patterns at m-tuples.

The sixth and seventh chapters are a comprehensive review and analysis of _co-methylation_, the spatial dependence of DNA methylation. Chapter \ref{chap:co-methylation} re-examines previous analyses of co-methylation and finds that these have been limited by inadequate data and deficiencies in the statistical methods applied. We address some of these limitations with an extensive analysis of co-methylation in Chapter \ref{chap:co-methylation}. We propose a novel analysis of co-methylation using methylation patterns from individual DNA fragments, which we call _within-fragment co-methylation_. A simulation study demonstrates that, in general, we cannot estimate this within-fragment co-methylation for individual pairs of CpGs. By aggregating pairs of CpG, however, the simulation study shows that we can estimate a summary of this spatial dependence of DNA methylation along individual DNA fragments. We apply this method to 40 whole-genome bisulfite-sequencing samples to identify genomic features that influence co-methylation and study how it varies between different tissues.

To conclude, Chapter \ref{chap:methsim} details our efforts to develop the `methsim` software to simulate whole-genome bisulfite-sequencing data. Unlike existing software, `methsim` seeks to model co-methylation so as to create more realistic simulated data. Such simulated data will be useful in developing and benchmarking methods for the analysis of DNA methylation data. While aspects of `methsim` are promising, the overly-strong correlations of $\beta$-values is an obvious inadequacy in the current model. We continue to work to improve this and to develop the software more generally.

DNA methylation data have a complex dependence structure, as demonstrated by the analyses presented in this thesis. Any analysis must bear in mind the multiple sources of variation in the data and the level of aggregation required in order to extract meaningful results.

The large size of the data generated by these experiments can make it challenging to develop analysis methods. This is particularly true when coupled with a lack of software for extracting, manipulating and summarising the relevant features of the data. Much of the effort in this thesis has gone into developing software to extract and manage this information (`methtuple` and `MethylationTuples`) so that new tools can be built on a higher foundation (`methsim`). It is my hope that, with continued development, these tools may facilitate other analyses to unravel the complexity of DNA methylation and its many roles in biology.
