---
title: "Datasets used in thesis"
author: "Peter Hickey"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  pdf_document:
    fig_caption: yes
    includes:
      in_header: ../latex/header.tex
  html_document:
    keep_md: true
bibliography: ../latex/phd_thesis.bib
---

\chapter{Datasets used in thesis}\label{chap:datasets}

\section{Overview of data processing}

This chapter briefly documents the 40 whole-genome bisulfite-sequencing samples that I use in my thesis. The `BAM` files containing the aligned reads for each sample underwent the same basic processing:

1. Genetic variants were called using the `bissnp_easy_usage.pl` script included with `Bis-SNP (v0.82.2)`.
2. M-bias was estimated using `bismark_methylation_extractor` with the `--mbias_only` flag set. These output files were then analysed using the  `MethylationTuples` R package (see section \ref{sec:MethylationTuples}) and all read-positions with an CpG normalised read-position methylation level ($nrpml$) more than $0.03$ from the median, i.e. $|nrpml_{CpG} - median(nrpml_{CpG})| > 0.03$, were excluded from methylation calling (_read-1_ and _read-2_ analysed separately where applicable).
3. CpG methylation calling was performed using `methtuple (v1.4.0)` for m-tuples $\text{m} = 1, \ldots, 8$. CpG 2-tuples were called both with and without the `--all-combinations` flag; all other tuple sizes were called without the `--all-combinations` flag. The following `methtuple` flags were also used: `--methylation-type CG --ignore-duplicates --min-mapq 0 --overlap-filter XM_ol --ignore-duplicates`.
4. Sample-level m-tuples were combined at the dataset-level using the `MethylationTuples` R package (see Section \ref{sec:MethylationTuples}). Specifically, a `MethPat` object was created for each of the _EPISCOPE_, _Lister_, _Seisenberger_ and _Ziller_ datasets for 1-tuples, 2-tuples and 2ac-tuples (2-tuples with the  `--all-combinations` flag in `methtuple` set), 3-tuples and 4-tuples. I did not create `MethPat` objects for m-tuples with $\text{m} > 4$ because the data are too sparse at this larger sizes to be generally useful.

The raw data for the _Lister_, _Seisenberger_ and _EPISCOPE_ datasets are all publicly available. The _EPISCOPE_ data are not yet published and I do not have permission to make these publicly available. The scripts used to prepare the results for each chapter are available from \url{https://github.com/PeteHaitch/phd_thesis_analyses}. Further details of software used are available in Appendix \ref{app:software_details}.

\section{\emph{Lister} dataset}\label{sec:Lister_dataset}

The _Lister_ dataset refers to whole-genome bisulfite-sequencing libraries used in \cite{Lister:2009hy} and \cite{Lister:2011kg}. The _Lister_ data were the largest publicly available human whole-genome bisulfite-sequencing datasets until quite recently.

\subsection{Sample descriptions}

The methylC-seq libraries from the \cite{Lister:2009hy} paper were the first published whole-genome bisulfite-sequencing libraries of mammalian DNA. A focus of this paper was comparing DNA methylation levels in a somatic tissue, fetal lung fibroblasts (_IMR90_), with those from a pluripotent tissue, embryonic stem cells (_H1_). Each tissue was run in duplicate. While \cite{Lister:2009hy} refer to these "biological" replicates I believe that these are better described as technical replicates since each replicate is from the same cell line; what distinguishes the replicates are the number of cell passages and the subsequent library preparations and sequencing. In any case, the published analyses pool these duplicates, which ignores all between-replicate variability. These samples are detailed in Table \ref{tab:Lister_dataset_pt1}.

\begin{table}[!htb]
\centering
\caption[Description of \emph{Lister} dataset part 1]{Samples in \emph{Lister} dataset from \cite{Lister:2009hy}. The reported read lengths are prior to any trimming of the reads. Abbreviations: \emph{ESC} = embryonic stem cell; \emph{SE} = single-end sequencing.}
\label{tab:Lister_dataset_pt1}
\begin{tabulary}{\textwidth}{LLLLL}
\toprule
Sample              & Tissue type             & Sequencing & Read length & Ave. coverage \\ \midrule
IMR90\_r1           & Lung fibroblasts        & SE         & $87$ bp    & $14\times$ \\
IMR90\_r2           & Lung fibroblasts        & SE         & $87$ bp    & $15\times$ \\
H1\_r1              & ESC                     & SE         & $85$ bp    & $15\times$ \\
H1\_r2              & ESC                     & SE         & $85$ bp    & $14\times$ \\
\bottomrule
\end{tabulary}
\end{table}

The methylC-seq libraries from the \cite{Lister:2011kg} include some created by the authors and some published by other groups. These samples include cell lines from differentiated cell lines, embryonic stem cell lines, pluripotent stem cell lines and _in vitro_ differentiated from pluripotent stem cells. There are no replicates for any of the \cite{Lister:2011kg} samples. These samples are detailed in Table \ref{tab:Lister_dataset_pt2}.

\begin{table}[!htb]
\centering
\caption[Description of \emph{Lister} dataset part 2]{Samples in \emph{Lister} dataset from \cite{Lister:2011kg}. The reported read lengths are prior to any trimming of the reads. Abbreviations: \emph{iPSC} = induced pluripotent stem cell; \emph{ESC} = embryonic stem cell; \emph{IVD} = \emph{in vitro} differentiated from pluripotent cell line; \emph{SE} = single-end sequencing; \emph{PE} = paired-end sequencing.}
\label{tab:Lister_dataset_pt2}
\begin{tabulary}{\textwidth}{LLLLL}
\toprule
Sample              & Tissue type             & Sequencing & Read length & Ave. coverage \\ \midrule
ADS                 & Adipose                 & PE         & $75$ bp    & $23\times$ \\
ADS-adipose         & Adipocytes from ADS     & PE         & $75$ bp    & $24\times$ \\
ADS-iPSC            & iPSC from ADS           & PE         & $75$ bp    & $26\times$ \\
FF                  & Foreskin fibroblasts    & SE         & $85$ bp    & $16\times$ \\
FF-iPSC\_6.9        & iPSC from FF            & SE         & $85$ bp    & $10\times$ \\
FF-iPSC\_19.7       & iPSC from FF            & SE         & $85$ bp    & $9\times$  \\
FF-iPSC\_19.11      & iPSC from FF            & SE         & $85$ bp    & $8\times$  \\
FF-iPSC\_19.11+BMP4 & IVD from FF-iPSC\_19.11 & SE         & $85$ bp    & $17\times$ \\
IMR90-iPSC          & iPSC from IMR90         & SE         & $85$ bp    & $9\times$ \\
H1+BMP4             & IVD from H1             & SE         & $85$ bp    & $33\times$ \\
H9                  & ESC                     & SE         & $85$ bp    & $9\times$ \\
H9\_Laurent         & ESC                     & PE         & $75$ bp    & $8\times$ \\
HSF1                & ESC                     & SE         & $47$ bp    & $5\times$ \\
\bottomrule
\end{tabulary}
\end{table}

\FloatBarrier

There are four 'mini datasets' within the Lister data that I make some use of in my thesis. The first I refer to as the _Lister-ADS_ data and includes samples _ADS_, _ADS-adipose_ and _ADS-iPSC_, all from the 2011 paper. The _ADS_ sample, a human adipose tissue cell line, is the 'founder' of this mini dataset. The _ADS-adipose_ and _ADS-iPSC_ are both derived from the _ADS_ cell line. The _ADS-adipose_ sample are "adipocytes derived from the _ADS_ cells through adipogenic differentiation conditions". The _ADS-iPSC_ cell line is an induced pluripotent stem cell line derived from _ADS_.

The second mini dataset I refer to as the _Lister-FF_ data and includes samples _FF_, _FF-iPSC\_6.9_, _FF-iPSC\_19.7_, _FF-iPSC\_19.11_ and _FF-iPSC\_11.11+BMP4_, all from the 2011 paper. The _FF_ sample, a foreskin fibroblast cell line, is the 'founder' of this this mini dataset. The _FF-iPSC\_6.9_, _FF-iPSC\_19.7_ and _FF-iPSC\_19.11_ are all induced pluripotent stem cell lines derived from _FF_. In fact, _FF-iPSC\_19.7_ and _FF-iPSC\_19.11_ are subclones derived from _FF-iPSC\_19_, whose methylome was not sequenced. I believe _FF-iPSC\_6.9_ is an independently derived iPSC cell line from _FF_, although this isn't made clear in the original publication. The _FF-iPSC\_19.11+BMP4_ sample is a trophoblast cell line derived by _in vitro_ differentiating the _FF-iPSC\_19.11_ by growing a clone of it in bone morphogenic protein 4 (_BMP4_).

The third mini dataset I refer to as the _Lister-IMR90_ data and includes samples _IMR90\_r1_ (2009), _IMR90\_r2_ (2009) and _IMR90-iPSC_ (2011). The _IMR90-iPSC_ sample is an induced pluripotent stem cell line derived from the _IMR90_ cell line.

The final mini dataset I refer to as the _Lister-H1_ data and includes samples _H1\_r1_ (2009), _H1\_r2_ (2009) and _H1+BMP4_ (2011). The _H1+BMP4_ sample is a trophoblast cell line derived by _in vitro_ differentiating the _H1_ by growing a clone of it in bone morphogenic protein 4 (BMP4).

\subsection{Creation of \texttt{BAM} files}

The aligned reads for the \cite{Lister:2009hy} data were downloaded from [http://neomorph.salk.edu/human_methylome/](http://neomorph.salk.edu/human_methylome/). The aligned reads for the \cite{Lister:2011kg} data were downloaded from [http://neomorph.salk.edu/ips_methylomes/](http://neomorph.salk.edu/ips_methylomes/). These samples had been aligned against the hg18 build of the human reference genome.

As the aligned reads were in a custom file format, I wrote Python scripts to convert these files to the canonical `SAM` format. These scripts are available from [https://github.com/PeteHaitch/Lister2BAM](https://github.com/PeteHaitch/Lister2BAM). These `SAM` files were then converted to `BAM` files with `SAMtools` \citep{Li:2009ka} and duplicate reads were marked using Picard's `MarkDuplicates` routine ([http://broadinstitute.github.io/picard/](http://broadinstitute.github.io/picard/)).

\section{\emph{EPISCOPE} dataset}\label{sec:EPISCOPE_dataset}

The _EPISCOPE_ data were kindly provided to me by Professor Susan Clark (Garvan Institute of Medical Research, Sydney) and Dr Peter Molloy (CSIRO Animal, Food and Health Sciences). This dataset is not yet published.

\subsection{Sample descriptions}

The data are from three human donors across four different tissues, for a total of 12 whole-genome bisulfite-sequencing libraries. The four tissues are:

- _BUF_: Buffy coat layer, which are leukocytes and platelets derived by centrifugation of a whole blood sample.
- _SA_: Subcutaneous adidose tissue, which is fat found just below the skin. Unlike visceral adipose tissue, subcutaneous adipose tissue is thought to be protective against obesity-related metabolic dysfunction \citep{Chau:2014do}.
- _VA_: Visceral adipocytes, which are derived from _VAT_.
- _VAT_: Visceral adipose tissue, which is located inside the abdominal cavity, packed between the organs and is associated with metabolic dysfunction \citep{Chau:2014do}.

The data are summarised in Table \ref{tab:EPISCOPE_dataset}.

\begin{table}[!htb]
\centering
\caption[Description of \emph{EPISCOPE} dataset]{Samples in \emph{EPISCOPE} dataset. The reported read lengths are prior to any trimming of the reads. Abbreviations: \emph{PE} = paired-end sequencing; \emph{BUF} = buffy coat; \emph{SA} = subcutaneous adipose; \emph{VA} = visceral adipocytes; \emph{VAT} = visceral adipose tissue.}
\label{tab:EPISCOPE_dataset}
\begin{tabulary}{\textwidth}{LLLLL}
\toprule
Sample & Tissue & Sequencing & Read length & Ave. coverage \\ \midrule
E13BUF & BUF & PE         & $101$ bp    & $8\times$     \\
E13SA  & SA  & PE         & $101$ bp    & $28\times$    \\
E13VA  & VA  & PE         & $150$ bp    & $27\times$    \\
E13VAT & VAT & PE         & $101$ bp    & $25\times$    \\
E18BUF & BUF & PE         & $101$ bp    & $21\times$    \\
E18SA  & SA  & PE         & $101$ bp    & $25\times$    \\
E18VA  & VA  & PE         & $150$ bp    & $36\times$    \\
E18VAT & VAT & PE         & $101$ bp    & $26\times$    \\
E23BUF & BUF & PE         & $101$ bp    & $12\times$    \\
E23SA  & SA  & PE         & $101$ bp    & $29\times$    \\
E23VA  & VA  & PE         & $101$ bp    & $32\times$    \\
E23VAT & VAT & PE         & $101$ bp    & $31\times$    \\ \bottomrule
\end{tabulary}
\end{table}

\subsection{Creation of \texttt{BAM} files}

The sequencing data for these 12 samples were processed and aligned by Aaron Statham (Garvan Institute of Medical Research, Sydney). Each sample was aligned to the human reference genome (hg19) using `Bismark` (`v0.8.3`) with the `Bowtie2` backend. The default alignment options were used, except that the maximum insert size for valid paired-end alignments was set to 1000 instead of 500 (`-X 1000`). Duplicate reads had already been removed from the `BAM` files that I received.

\section{\emph{Seisenberger} dataset}\label{sec:Seisenberger_dataset}

The _Seisenberger_ data are from a study of the dynamics of DNA methylation reprogramming in mouse primordial germ cells \citep{Seisenberger:2012ko}. These were a convenience sample provided to me by a colleague, Felix Krueger (Babraham Institute). I thank Felix who sent me the `BAM` files containing processed and aligned reads.

\subsection{Sample descriptions}

I have the data for only three samples from the original publication, detailed in Table \ref{tab:Seisenberger_dataset}. The _J1\_1_ sample is from an embryonic stem cell line while both the _E6.5\_epiblast\_1_ and _E16.5\_male\_1_ samples are derived from pools of $10$ to $30$ embryos. Developmentally, the samples are ordered _J1\_1_ (embryonic stem cell), _E6.5\_epiblast\_1_ (embryonic day 6.5 epiblast) and _E16.5\_male\_1_ (embryonic day 16.5 male progenitor germ cells).

I believe that the samples I received labelled _J1\_1_ and _E16.5\_male\_1_  in fact correspond to _J1\_2_ and _E16.5\_male\_2_, respectively, i.e. the second  replicate rather than the first. The data I received are all $100$ bp paired-end sequences, which matches replicate 2 rather than replicate 1 for both of these samples \citet[Supplementary Table 1]{Seisenberger:2012ko}.

\begin{table}[!htb]
\centering
\caption[Description of \emph{Seisenberger} dataset]{Samples in \emph{Seisenberger} dataset. The reported read lengths are prior to any trimming of the reads. All samples were first published in \citet{Seisenberger:2012ko}. Abbreviations: \emph{PE} = paired-end sequencing}
\label{tab:Seisenberger_dataset}
\begin{tabulary}{\textwidth}{LLLLL}
\toprule
Sample            & Tissue                     & Sequencing & Read length & Ave. coverage             \\ \midrule
J1\_1             & ESC                        & PE         & $100$ bp    & $12\times$     \\
E6.5\_epiblast\_1 & Epiblast                   & PE         & $100$ bp    & $13\times$     \\
E16.5\_male\_1    & Male progenitor germ cells & PE         & $100$ bp    & $12\times$     \\ \bottomrule
\end{tabulary}
\end{table}

\subsection{Creation of \texttt{BAM} files}

The sequencing data for these 3 samples were processed and aligned by Felix Krueger (Babraham Institute). Each sample was aligned to the mouse reference genome (GRCm38/mm10) using `Bismark` (`v0.7.12`) with the `Bowtie1` backend. The default alignment options were used.

\section{\emph{Ziller} dataset}\label{sec:Ziller_dataset}

The _Ziller_ data are a subset of the data used in \cite{Ziller:2013fs}. Specifically, I use a convenience sample of 8 whole-genome bisulfite-sequencing libraries. These were made available to me by a collaborator, Aaron Statham (Garvan Institute of Medical Research, Sydney). I thank Aaron who sent me the `BAM` files containing processed and aligned reads.

\subsection{Sample descriptions}

The eight biological samples are as follows: frontal cortex from two 'normal' women donors (_Frontal\_cortex\_normal\_1_ and _Frontal\_cortex\_normal\_2_) and from two women who had Alzheimer's disease (_Frontal\_cortex\_AD\_1_ and _Frontal\_cortex\_AD\_2_); a sample from a human liver carcinoma cell line (_HepG2\_cell\_line_); a new sample from the _IMR90_ lung fibroblast cell line (_IMR90\_cell\_line_); and samples from a colon cancer matched tumour-normal pair (_Colon\_Tumor\_Primary_ and _Colon\_Primary\_Normal_).

Table \ref{tab:Ziller_dataset_sequencing_runs} summarises the data for the 19 individual sequencing runs[^Ziller_merged_ave_cov].

[^Ziller_merged_ave_cov]: The average sequencing coverage of the _post-hoc_ merged samples are approximately the sums of the average sequencing coverage for the corresponding individual sequencing runs.

\begin{table}[!htb]
\centering
\caption[Description of \emph{Ziller} dataset]{Sequencing runs in \emph{Ziller} dataset. The reported read lengths are prior to any trimming of the reads. All samples were first published in \citet{Ziller:2013fs}. Abbreviations: \emph{PE} = paired-end sequencing}
\label{tab:Ziller_dataset_sequencing_runs}
\begin{tabulary}{\textwidth}{LLLLL}
\toprule
Sample    & Tissue                     & Sequencing & Read length & Ave. coverage        \\ \midrule
SRR949193 & Frontal\_cortex\_normal\_1 & PE         & 101 bp         & $10\times$ \\
SRR949194 & Frontal\_cortex\_normal\_1 & PE         & 101 bp         & $10\times$    \\
SRR949195 & Frontal\_cortex\_normal\_1 & PE         & 101 bp         & $10\times$    \\
SRR949196 & Frontal\_cortex\_normal\_2 & PE         & 101 bp         & $9\times$     \\
SRR949197 & Frontal\_cortex\_normal\_2 & PE         & 101 bp         & $9\times$     \\
SRR949198 & Frontal\_cortex\_normal\_2 & PE         & 101 bp         & $9\times$     \\
SRR949199 & Frontal\_cortex\_AD\_1     & PE         & 101 bp         & $9\times$     \\
SRR949201 & Frontal\_cortex\_AD\_1     & PE         & 101 bp         & $9\times$     \\
SRR949202 & Frontal\_cortex\_AD\_2     & PE         & 101 bp         & $10\times$    \\
SRR949203 & Frontal\_cortex\_AD\_2     & PE         & 101 bp         & $10\times$    \\
SRR949206 & HepG2\_cell\_line          & PE         & 101 bp         & $2\times$     \\
SRR949207 & HepG2\_cell\_line          & PE         & 101 bp         & $1\times$     \\
SRR949208 & IMR90\_cell\_line          & PE         & 101 bp         & $1\times$     \\
SRR949209 & IMR90\_cell\_line          & PE         & 101 bp         & $3\times$     \\
SRR949210 & Colon\_Tumor\_Primary      & PE         & 101 bp         & $8\times$     \\
SRR949211 & Colon\_Tumor\_Primary      & PE         & 101 bp         & $8\times$     \\
SRR949212 & Colon\_Tumor\_Primary      & PE         & 101 bp         & $9\times$     \\
SRR949213 & Colon\_Tumor\_Primary      & PE         & 101 bp        & $8\times$     \\
SRR949215 & Colon\_Primary\_Normal     & PE         & 101 bp         & $8\times$     \\ \bottomrule
\end{tabulary}
\end{table}

\subsection{Creation of \texttt{BAM} files}

I received 19 `BAM` files from Aaron, which represent 19 sequencing runs of the eight biological samples. Each of the 19 `BAM` files was processed separately and I then _post hoc_ merged the processed data from sequencing runs. This unfortunately reduces the power to detect genetic variants since the coverage of individual sequencing runs is lower than merged data, but does not adversely affect methylation calling since the number of methylated and unmethylated reads can be summed across sequencing runs.

\section{CpG islands}

I have used the CpG island definition from the hidden Markov model proposed by \citet{Wu:2010do} and implemented in the `makeCGI` R package (`v1.2`, \url{http://rafalab.jhsph.edu/CGI/}). The predicted CpG islands for the human reference genome (both hg18 and hg19) were downloaded from [http://rafalab.jhsph.edu/CGI/](http://rafalab.jhsph.edu/CGI/) on 29/10/2014. I used the `makeCGI` R package to create the predicted CpG islands for the mouse reference genome (mm10) since these were not available for download.
