---
title: "Datasets used in thesis"
author: "Peter Hickey"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  pdf_document:
    fig_caption: yes
    includes:
      in_header: ../latex/header.tex
  html_document:
    keep_md: true
bibliography: ../latex/phd_thesis.bib
---

\chapter{Datasets used in thesis}\label{chap:datasets}

\section{Chapter overview}

I have analysed more than 50 whole-genome bisulfite-sequencing datasets during the course of my PhD studies. This short chapter documents those that I use in my thesis.

__TODO: Describe processing of data__

## Lister data

The _Lister data_ refer to whole-genome bisulfite-sequencing libraries used in \cite{Lister:2009hy} and \cite{Lister:2011kg}. The Lister data were the largest publicly available human WGBS datasets until quite recently.

The WGBS libraries from the \cite{Lister:2009hy} paper were the first published WGBS libraries from mammalian DNA. A focus of this paper was comparing DNA methylation levels in a somatic tissue, fetal lung fibroblasts (`IMR90`), with those from a pluripotent tissue, embryonic stem cells (`H1`). Each tissue was run in duplicate. While \cite{Lister:2009hy} refer to these "biological" replicates I believe that these are better described technical replicates since each replicate is from the same cell line; what distinguishes the replicates are the number of cell passages and the subsequent library preparations and sequencing. In any case, the published analyses pool these duplicates, which ignores all between-replicate variability.

The WGBS libraries from the \cite{Lister:2011kg} include some created and published by other groups (see __TABLE__). These samples include cell lines from differentiated cell lines, embryonic stem cell lines, pluripotent stem cell lines and _in vitro_ differentiated from pluripotent stem cells. There are no replicates for any of these samples.

__TODO: Table of datasets__

There are a four "mini datasets" within the Lister data that I make some use of in my thesis. These are described below and illustrated in __FIGURE__.

The first I refer to as the _Lister-ADS_ data and includes samples `ADS`, `ADS-adipose` and `ADS-iPSC`, all from the 2011 paper. The `ADS` sample, a human adipose tissue cell line, is the "founder" of this mini dataset. The `ADS-adipose` and `ADS-iPSC` are both derived from the `ADS` cell line. The `ADS-adipose` sample are "adipocytes derived from the `ADS` cells through adipogenic differentiation conditions". The `ADS-iPSC` cell line is an induced pluripotent stem cell line derived from `ADS`.

The second mini dataset I refer to as the _Lister-FF_ data and includes samples `FF`, `FF-iPSC\_6.9`, `FF-iPSC\_19.7`, `FF-iPSC\_19.11` and `FF-iPSC\_11.11+BMP4`, all from the 2011 paper. The `FF` sample, a foreskin fibroblast cell line, is the "founder" of this this mini dataset. The `FF-iPSC\_6.9`, `FF-iPSC\_19.7` and `FF-iPSC\_19.11` are all induced pluripotent stem cell lines derived from `FF`. In fact, `FF-iPSC\_19.7` and `FF-iPSC\_19.11` are subclones derived from `FF-iPSC\_19`, whose methylome was not sequenced. I believe `FF-iPSC\_6.9` is an independently derived iPSC cell line from `FF`, although this isn't made clear in the original publication. The `FF-iPSC\_19.11+BMP4` sample is a trophoblast cell line derived by _in vitro_ differentiating the `FF-iPSC\_19.11` by growing a clone of it in bone morphogenic protein 4 (_BMP4_).

The third mini dataset I refer to as the _Lister-IMR90_ data and includes samples `IMR90\_r1` (2009), `IMR90\_r2` (2009) and `IMR90-iPSC` (2011). The `IMR90-iPSC` sample is an induced pluripotent stem cell line derived from the `IMR90` cell line.

The final mini dataset I refer to as the _Lister-H1_ data and includes samples `H1\_r1` (2009), `H1\_r2` (2009) and `H1+BMP4` (2011). The `H1+BMP4` sample is a trophoblast cell line derived by _in vitro_ differentiating the `H1` by growing a clone of it in bone morphogenic protein 4 (BMP4).

The aligned reads for the \cite{Lister:2009hy} data were downloaded from [http://neomorph.salk.edu/human_methylome/](http://neomorph.salk.edu/human_methylome/). The aligned reads for the \cite{Lister:2011kg} data were downloaded from [http://neomorph.salk.edu/ips_methylomes/](http://neomorph.salk.edu/ips_methylomes/). I wrote Python scripts to convert these files to `SAM` files, these scripts are available from [https://github.com/PeteHaitch/Lister2BAM](https://github.com/PeteHaitch/Lister2BAM). These `SAM` files were then converted to `BAM` files and duplicate reads were marked using Picard's `MarkDuplictes` routine [http://broadinstitute.github.io/picard/](http://broadinstitute.github.io/picard/).

## EPISCOPE data

The EPISCOPE data were kindly provided to me by Professor Susan Clark (Garvan Institute of Medical Research, Sydney) and Dr Peter Molloy (CSIRO Animal, Food and Health Sciences). The data are from three human donors across four different tissues, for a total of 12 whole-genome bisulfite-sequencing libraries.

The four tissues are:

* BUF: Buffy coat layer, which are leukocytes and platelets derived by centrifugation of a whole blood sample.
* SA: Subcutaneous adidose tissue, which is fat found just below the skin. Unlike visceral fat, "subcutaneous fat is not related to many of the classic obesity-related pathologies, such as heart disease, cancer, and stroke" (__TODO: Get better reference than wikipedia__).
* VA: Visceral adipose tissue, which is located inside the abdominal cavity, packed between the organs (stomach, liver, intestines, kidneys, etc.) (__TODO: Get better reference than wikipedia__).
* VAT: Visceral adipocytes, which are derived from the VA tissue.

The data are summarised in __TABLE__

__TODO: Table of datasets__
__TODO: Get age of donors__

The sequencing data for these 12 samples were processed and aligned by Aaron Statham (Garvan Institute of Medical Research). Each sample was aligned to the human reference genome (hg19) using Bismark (v0.8.3) with the Bowtie2 backend. The default alignment options were used, except that the maximum insert size for valid paired-end alignments was set to 1000 instead of 500 (`-X 1000`). An example command is shown below:

```
bismark -p 4 --bowtie2 -X 1000 --unmapped --ambiguous --gzip --bam -o SAMPLE_NAME  REF_GENOME -1 READ_1.fq -2 READ_2.fq
```

Duplicate reads had been removed from the `BAM` files that I received.

## Whitelaw data

I describe this dataset in detail in Chapter \ref{chap:avy}.

## Ziller data

The Ziller are a subset of the data used in \cite{Ziller:2013fs}. Specifically, I use a convenience sample of 21 (__TODO: Check final numbers__) WGBS libraries. These were made available to me by a collaborator, Aaron Statham (Garvan Institute of Medical Research), who already had aligned and processed BAM files for this subset of samples.

There are samples from the frontal cortex from two "normal" women (`Frontal_cortex_normal_1` and `Frontal_cortex_normal_2`) and from two women who had Alzheimer's disease (`Frontal_cortex_AD_1` and `Frontal_cortex_AD_2`); a sample from a human liver carcinoma cell line (`HepG2_cell_line`); a new sample from the IMR90 lung fibroblast cell line (`IMR90_cell_line`); and samples from a matched tumour-normal pair for colon cancer (`Colon_Tumor_Primary` and `Colon_Tumor_Primary`). Full details of the Ziller data used in my thesis are described in __TABLE__. Note that I consider the samples described as `replicates` to be technical replicates, since each replicate comes from the same biological sample.

__TODO: Table of datasets. Only include those samples with properly processed files (i.e., exclude truncated files)__

## Seisenberger data

The Seisenberger data are from a study of the dynamics of DNA methylation reprogramming in mouse primordial germ cells \cite{Seisenberger:2012ko}. Like the Ziller data, these were a convenience sample provided to me by a colleague. I thank Felix Krueger (Babraham Institute) who processed and aligned the reads, and sent me the `BAM` files.

I have the data for three samples, detailed in __TABLE__. Developmentally, the samples are ordered `J1_1`, `E6.5_epiblast_1` and `E16.5_male_1`. Both the `E6.5_epiblast_1` and `E16.5_male_1` samples are from a pool of $10-30$ embryos.

__TODO: Table of datasets__

## CpG islands

I have used CpG island definitions from the hidden Markov model proposed by \cite{@Wu:2010do}. These predicted CpG islands were downloaded from [http://rafalab.jhsph.edu/CGI/](http://rafalab.jhsph.edu/CGI/) on 29/10/2014.

## Misc. notes

- `Bis-SNP` run without `--trim5` and `--trim5` flags. This is a bit unfortunate but it was because the `bissnp_easy_usage.pl` script, which I used to run `Bis-SNP`, didn't support these. To re-genotype all samples would have taken too long and it shouldn't greatly affect my analyses.
- Percentage of CpGs that are in Rafa's CGIs
