---
title: "A simulation model of DNA methylation data"
author: "Peter Hickey"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  pdf_document:
    fig_caption: yes
    includes:
      in_header: ../latex/header.tex
  html_document:
    keep_md: true
bibliography: ../latex/phd_thesis.bib
---

\chapter{A simulation model of DNA methylation data}\label{chap:methsim}

\section{Chapter overview}\label{sec:methsim_overview}

In this chapter I argue for the utility, even necessity, of methodology and software for simulating DNA methylation data. I present the `methsim` software to simulate DNA methylation data. `methsim` incorporates a model of within-fragment co-methylation and simulates individual sequencing reads, which sets it apart from existing software. I describe the statistical model behind `methsim` and its software implementation. Finally, I provide examples of how `methsim` can be used to learn about DNA methylation and how it might be used to develop simulation studies to compare competing analysis methods.

1. Introduction
  - Why simulation?
2. Literature review
  - Simulating reads vs. $\beta$-values
  -
3. Methods
  - Implementation
4. Case study
5. Conclusions

\section{Introduction}\label{sec:methsim_introduction}

- [x] DONE

It can be difficult to design an experiment that can be used for validating or benchmarking different analysis strategies. In fact, the huge variety of experimental factors, and their possible values, can make such a task infeasible. But perhaps the bigger hurdle is that there are not attractive experiments to perform; why spend your time and money on an experiment where you "know the answer" when you could be spending that same time and money on investigating something new? In these scenarios and more, simulation studies play a vital role in applied statistics, where they can be used in the development, validation and benchmarking of different analysis methodologies.

The key advantage of simulated data is that we know the truth _a priori_. Moreover, we can manipulate the truth via parameters in the simulation model and can examine how a method performs under a variety of scenarios. And this manipulation is cheap, a mere matter of changing parameters and re-running a piece of software, which means that we can investigate a broad range of plausible scenarios.

Simulation studies can provide many insights into the performance of a method. At the most basic level, if a method fails to work or performs poorly when applied to simulated data, then it is very unlikely to work well when applied to real data. We can also identify which scenarios are "easy", ones where most methods are able to identify the truth, and scenarios under which certain methods perform better than others. While simulated data can never fully capture the richness of "real" data, we can learn a lot about a method by studying how it performs when applied to simulated data.

Simulation methods may also be used to learn about the plausibility of hypothesised models of a phenomenon, be they mechanistic or stochastic. By simulating data from the proposed model, and comparing it to the real data, we can identify hypotheses that are incompatible with reality. This can also help identify shortcomings in the model so that it may be refined in an iterative manner.

There are a few key criteria when designing a simulation method:

1. Realism: The simulated data must be "similar" to the real data. While this is obvious, it is also often hard to pin down or agree upon what constitutes "similar enough".
2. Cost: It should be fast and cheap to simulate data. The most common use of simulation models in applied statistics is to repeatedly generate datasets under a range of parameter settings. This requires that each simulation is fast and computationally cheap, otherwise it will be prohibitive to analyse the full space of scenarios. An exception to this rule may occur when the simulation model is used to test a proposed mechanistic or stochastic model of a phenomenon, such as in studies of molecular dynamics. Even then, however, the cost of a simulation should be less than the cost of performing the equivalent experiments, otherwise the simulation is generally not worth the effort.
3. Usability: There __must__ be a software implementation. Simulation models exist to be simulated from; a simulation model without an implementation is next to useless. The implementation should give the user easy access to the key  parameters and have sensible default settings. The output of the software should be in a standard format or readily convertible to a simple, manipulable format.

There is a lack of gold-standard datasets in the field of DNA methylation for the benchmarking and validation of analysis methods. Therefore, the development of realistic, cheap and user-friendly software to simulate DNA methylation data will be of benefit.

\section{Literature review}\label{sec:methsim_lit_review}

- [x] DONE

The simulation methods described here can be thought to lie on a spectrum with model-based methods at one end and re-sampling based methods at the other end[^model-based]. It is often simpler to simulate from a model-based simulation, particularly if the model is a well-studied parametric distribution. However, this simplicity often comes at the cost of increased assumptions, whose validity may be questionable. Simulations based on sampling of real data may reduce the number of assumptions required. However, care must be taken in selecting the units to be sampled so that the sampling process is efficient and so that the sampled data don't grossly distort within- and between-sample dependencies.

Any procedure for simulating DNA methylation data should obviously be tailored to its purpose. For example, if the study is comparing alignment software for bisulfite-sequencing data then the simulation software should produce `FASTQ` files with realistic base quality score profiles and sequencing errors. On the other hand, it may be sufficient to simulate some aggregate data, such as $\beta$-values, when the simulated data are to be used for comparing methods to identify differential methylation.

[^model-based]: Of course, the parameters in any model-based simulation should be based on real data, although this may not use a formal estimation procedure such as maximum likelihood.

\subsection{Methods for simulating bisulfite-sequencing reads}

- [x] DONE

All of the currently available methods for simulating bisulfite-sequencing reads are designed for the comparison of alignment strategies and are model-based. These are not suitable for comparisons of downstream analysis methods.

`Sherman` ([http://www.bioinformatics.babraham.ac.uk/projects/sherman/](http://www.bioinformatics.babraham.ac.uk/projects/sherman/)) is software to simulate bisulfite-sequencing reads, including various 'contaminants', such as SNPs, basecall errors and sequence artefacts as `FASTQ` files. The simulated reads are designed for comparing the performance of different alignment strategies. `Sherman` has many parameters, of which the ones relevant to our discussion of simulating realistic DNA methylation data are `-CG` and `-CH`, the bisulfite conversion rates for CG and CH methylation loci, respectively. These are set by the user with values between $0$ and $100$ ($\%$). Reads are simulated by sampling from the user-specified reference genome. When a read contains a CG (resp. CH) locus, it is randomly assigned as being converted (unmethylated) with probability `-CG` / 100 (resp. `-CH` / 100).  

While suitable for comparing alignment strategies, `Sherman` produces data that is not suitable for use in comparing downstream analysis methods. All CG (resp. CH) loci have an average $\beta$-value of `-CG` (resp. `-CG`) regardless of the genomic context, which we know to be incompatible with real data. Furthermore, the methylation state of each methylation locus is independent, which is clearly inconsistent with the strong co-methylation observed in real data.

`DNemulator` ([http://www.cbrc.jp/dnemulator/README.html](http://www.cbrc.jp/dnemulator/README.html)) uses a slightly more sophisticated simulation strategy to simulate `FASTQ` files for use in comparing alignment strategies for bisulfite-sequencing data. `DNemulator` does this with three separate routines, `fasta-methly-sim`, `fasta-polymorph` and `fasta-bisulf-sim`:

1. `fasta-methyl-sim` converts cytosines in the reference genome (`FASTA` file) to a character indicating the methylation level of that locus: `C` represents $0\%$ methylated, `c` represents $10\%$ methylated, `d` represents $20\%$ methylated, `v` represents $50\%$ methylated and `t` represents $100\%$ methylated. Each of these conversions has a different probability in the CG and CH contexts.
2. `fasta-polymorph` simulates a polymorphic, diploid genome based on the modified reference sequence created by `fasta-methyl-sim`.
3. `fasta-bisulf-sim` simulates reads by sampling from the simulated genome created by `fasta-polymorph`. Read are simulated with bisulfite-conversion error and sequencing error.

The reads simulated by `DNemulator` will result in $\beta$-values that have more context-dependence than those resulting from reads generated by `Sherman`. However, methylation events are still generated independently of one another, which means there is no co-methylation in the simulated data. Therefore, reads simulated by `DNemulator`, while suitable for comparing alignment strategies, are not suitable for comparing downstream analysis methods.

Other software for simulating individual bisulfite-sequencing reads are `FastqToBS` ([http://users.dimi.uniud.it/~nicola.prezza/projects.html](http://users.dimi.uniud.it/~nicola.prezza/projects.html)), which uses a similar strategy as `Sherman`, and `BSsim` ([http://122.228.158.106/BSSim/](http://122.228.158.106/BSSim/), and used in \citet{Xie:2014ez}), which has a similar strategy to `DNemulator`.

\subsection{Methods for simulating aggregate methylation levels}

- [x] DONE

The most widely studied downstream analysis problem is that of identifying differential methylation, which is done by comparing summary measures of methylation, such as $\beta$-values, between two or more groups. It is therefore generally not necessary to simulate individual reads, but is sufficient to directly simulate these aggregated measures, when simulating data for these studies.

Most papers that propose a new method for downstream analysis of bisulfite-sequencing data include a simulation section. Generally, such a simulation method exists to support claims about the performance of the proposed method and is not a major feature of the paper. Consequently, the simulation model is often only briefly described and a software implementation is rarely made available. In fact, until recently, of the methods reviewed in this section, only that of \citet{Lacey:2013iy} had a software implementation available[^WGBSSuite]. As I was writing this chapter, the `WGBSSuite` software was published \citep{Rackham:2015bv}. `WGBSSuite` is the only software specifically published for the purpose of simulating whole-genome bisulfite-sequencing data when comparing methods for identifying differential methylation. `WGBSSuite` is available for download as a collection of R scripts[^WGBSSuite].

[^WGBSSuite]: Unfortunately, `WGBSSuite` is not available as an R package, which is the "fundamental unit of reproducible R code" ([http://r-pkgs.had.co.nz/](http://r-pkgs.had.co.nz/)) and would greatly simplify the installation and use of the software. Furthermore, no licence file is included in the download, meaning that it is unclear how the user is permitted to use `WGBSSuite` and whether they may modify or redistribute the code. (__TODO: Update if I received a response. I emailed the developer, Owen Rackham, asking him to add a licence. He is considering my request.__)

Fortunately, these simulation methods follow a common framework, even if the details differ:

0. [Optional] Simulate the locations of methylation loci.
1. Simulate the unobserved group-specific true methylation levels.
2. Simulate the observed sample-specific sequencing depths.
3. Simulate the observed sample-specific methylation levels, e.g., the $\beta$-values.

A popular choice of parametric model in this framework is the beta-binomial model \citep[e.g.,]{Feng:2014iq, Lacey:2013iy, Xu:2013eg, Chen:2014jb, Dolzhenko:2014bo}. This model, and others, are now discussed.

\subsubsection{Simulating methylation loci}

- [x] DONE

\citet{Lacey:2013iy} and \citet{Rackham:2015bv} are notable in that they choose to simulate the locations of CpGs rather than simply using their locations in a reference genome. Both papers use hidden Markov models to simulate genomes with regions of high and low CpG density.

I do not think this a useful or necessary step. In fact, I think it is counterproductive. While in truth the set of methylation loci do vary between samples due to genetic variation, it is a reasonable approximation to consider the positions of these loci as fixed. If sequence variation is required then it is easily accommodations by sampling from the set of methylation loci in the reference genome. And it is counterproductive because the aim of the simulation model is to realistically simulate methylation _levels_, not the _locations_ of these loci. Furthermore, almost all downstream analyses are _reference-based_ (see Section \ref{sec:methylated_or_unmethylated}), so it is desirable to know how these methods perform with respect to the relevant reference genome, not some simulated genome in which the location of methylation loci vary from simulation to simulation.

\subsubsection{Simulating $B_{i, j}$}

- [x] DONE

Recall that $B_{i, j}$ is the underlying _true_ methylation level at locus $i$ in sample $j$. In the context of simulating data for comparing differential methylation calling methods, we want these to be group-specific. That is, we want to specify $B_{i, j^{k}}$, where $j^{k}$ indicates that sample $j$ comes from group $k$. For non-differentially methylated loci all $k$ groups have identical $B_{i, j^{k}} = B_{i, j^{0}}$; differentially methylated loci are simulated by setting $B_{i, j^{k}} \neq B_{i, j^{k'}}$ for some $k \neq k'$.

Under the beta-binomial model, the true methylation level, $B^{i, j}$, is assumed to follow a $Beta(\mu_{i, j^{k}}, \phi_{i, j^{k}})$ distribution, where $\mu_{i, j^{k}}$ is the mean and $\phi_{i, j^{k}}$ is the dispersion of the beta distribution[^beta_parameterisation]. Both the means, $\mu_{i, j^{k}}$, and the dispersions, $\phi_{i, j^{k}}$, are group-specific and are allowed to vary across methylation loci ($i$). The dispersion parameter models the within-group variability of the $B_{i, j}$, i.e., the within-group _biological variability_ of DNA methylation.

[^beta_parameterisation]: Note that this is different to the standard parameterisation of the Beta distribution, which is described by two shape parameters, $\alpha$ and $\beta$. The relationship between the two parameterisations is $\mu = \frac{\alpha}{\alpha + \beta}$ and $\phi = \frac{1}{\alpha + \beta + 1}$ \citep{Feng:2014iq}.

As noted by \citet{Feng:2014iq}, the beta distribution is a very flexible distribution with support on $[0, 1]$ and has "long been a natural choice to model binomial proportions", particularly as a conjugate prior, as it is used in the empirical Bayes model of \citet{Feng:2014iq}. The beta-binomial model can also be viewed from a non-Bayesian perspective as a compound distribution or overdispersed binomial distributions

Other distributions may be used instead of the beta distribution for modelling $B_{i, j}$. For example, \citet{Xie:2014ez} consider both a single Gaussian distribution and a mixture of Gaussian distributions, while \citet{Xu:2013eg} consider both a truncated Gaussian and a mixture of truncated Gaussian distributions.

An alternative to specifying a parametric distribution for the $B_{i, j}$ is to sample these from real data, e.g., by sampling some observed $\beta_{i, j}$ for a particular dataset and treating them as if they were observations on $B_{i, j}$. `WGBSSuite`\citep{Rackham:2015bv} uses a modified form of this approach. In `WGBSSuite`, a hidden Markov model is used to classify every CpG as having an underlying state ("de-methylated", "1st transition", "2nd transition" or "methylated"). Each of these four states has a region-specific average methylation level that is based on the distribution of $\beta$-values for a chosen dataset. For example, the average methylation level for all "de-methylated" regions is defined as $B_{'de-methylated'} = median(\{ \beta_{i, j}: \beta_{i, j} \in [0, 0.5) \}])$. Then, each $B_{i, j}$ is a perturbed version of this region-level average methylation, $B_{region-type}$, obtained by adding on a zero-mean Gaussian random variable, i.e., $B_{i, j} = B_{region-type} + \epsilon_{i, j}$, where $\epsilon_{i, j} \eqd Normal(0, s^{2})$. Care needs to be taken that $0 \leq B_{i, j} \leq 1$.

\subsubsection{Simulating sequence depth}

- [x] DONE

The sequencing depth at each methylation loci, $m_{i, j} + u_{i, j}$, may be sampled from real data \citep[e.g.]{Feng:2014iq, Chen:2014jb, Dolzhenko:2014bo}, or simulated from a parametric distribution such as the Poisson \citep{Rackham:2015bv}, a rounded Gaussian distribution \citep{Xu:2013eg}, or a rounded mixture of Gamma distributions \citep{Lacey:2013iy}.

The most sophisticated approach to simulation of sequencing depth in bisulfite-sequencing experiments is given by \citet{Lacey:2013iy}. In addition to using a mixture of distributions to capture both the low-coverage and high-coverage modes observed in RRBS sequencing coverage, \citet{Lacey:2013iy} model the correlation of sequencing depth across samples for a given region. They do this by using a Normal copula to make the set of sequencing depths a jointly dependent set of random variables. While this is undoubtably sophisticated, the effect of correlated versus uncorrelated sequencing depths in a simulation model is not explored in the paper and so the benefits are unclear. Moreover, it is simpler, and likely more computationally efficient, to include such correlations by a sensible sampling of sequencing depths from real data.

\subsubsection{Simulating the observed methylation levels}

- [x] DONE

The final step is to simulate the read counts, $M_{i, j}$ and $U_{i, j}$. These are based on the true underlying methylation level, $B_{i, j}$, and the sequencing depth, $m_{i, j} + u_{i, j}$. In the beta-binomial model this is done by binomial sampling where $M_{i, j} \eqd binomial(m_{i, j} + u_{i, j}, B_{i, j})$. In addition to binomial read sampling, the `WGBSSuite` software also implements (truncated) negative binomial read sampling \citep{Rackham:2015bv}, which essentially introduces overdispersion in the read counts.

\subsubsection{Simulating differentially methylated regions}

- [x] DONE

As we have seen, the simulation of a differentially methylated locus is straightforward; for a given locus ($i$) simply vary $B_{i, j_{k}}$ across the $k$ groups. The simulation of a differentially methylated region is more complex.

In principal we can simulate a differentially methylated region by simply simulating runs of differentially methylated loci. However, it must be noted that this requires careful choice of parameters. Such parameters include the length of the DMR, the minimal number of loci it must contain, the maximal intra-pair distances of loci within the DMR and how many of the loci in the DMR must themselves be differentially methylated, e.g., should Figure \ref{fig:dmr_break} be considered one DMR or two DMRs? Such decisions ultimately have to be made by the user of the simulation software based on the types of events they are interested in.

\begin{figure}[!htb]
\centering
\includegraphics[width=\textwidth]{../figures/dmr_break.pdf}
\caption{A hypothetical region in a study of differential methylation in a two-group experiment. Plotted is the difference in $\beta$-value between the two groups ($\delta \beta$) with associated standard error against the position along the genome (\emph{Position (bp)}). The first three methylation loci are DMLs, as are the last five methylation loci. However, the fourth locus is not a DML. Should this region be considered as two distinct DMRs or as a single DMR?}
\label{fig:dmr_break}
\end{figure}

\subsection{Simulating co-methylation}

- [x] DONE

A key feature ignored by the majority of simulation methods, with the notable exceptions of \citet{Lacey:2013iy} and \citet{Rackham:2015bv}, is co-methylation. The methylation states of neighbouring loci are highly dependent and, consequently, the $B_{i, j}$ of nearby loci are highly dependent. To be clear, the majority of studies using simulated DNA methylation data have failed to model a very important feature of DNA methylation data, co-methylation.

As demonstrated in Chapter \ref{chap:co-methylation}, the physical process of co-methylation is at the level of individual DNA fragments, not at the level of aggregate methylation levels. The observed correlations of $\beta$-values are driven by this within-fragment co-methylation. However, a simulation model of DNA methylation data need not induce correlations at the level of individual DNA fragments. Indeed, if the model does not simulate the individual fragments then it cannot simulate within-fragment co-methylation, but it still may be able to capture correlations of methylation levels by inducing dependence in the $B_{i, j}$. For example, under the Beta-Binomial model these correlations could be induced by forcing the $\mu_{i, j}$ to be spatially dependent. This idea takes its inspiration from \citet{Jaffe:2012gx} who simulate spatially correlated DNA methylation microarray data by imposing an autocorrelation structure via a lag-1 autoregressive model of the simulated $\beta$-values.

\citet{Lacey:2013iy} take a different approach, but one that still results in correlated $B_{i, j_{k}}$ across $i$, within group $k$. To begin, they compute $\beta$-values from chromosome 11 for a single normal myotube cell line that was sequenced with RRBS. They then fit a Gaussian variogram to these $\beta$-values, which shows "a strong correlation for sites in close proximity, decaying to near independence at distances beyong 3000 bp". To simulate spatially correlated $B_{i, j}$ they use an iterative process:

1. Simulate $B_{i, j_{k}}$ from a Beta distribution with parameters estimated from the chromosome 11 MTCTL2 data. These are estimated under an assumption of independence.
2. Induce correlation amongst the $B_{i, j_{k}}$ (across $i$) by a transformation of the $B_{i, j_{k}}$.

The second step uses a method published by \citet{Zaykin:2002ko}. The transformed values, $B_{i, j_{k}}^{*}$, are created by the transformation $\mathbf{B_{j_{k}}^{*}} = 1 - \Phi{C \Phi - 1(1 - B_{j_{k}})}$, where $\mathbf{B_{j_{k}}}$ is the vector of $B_{i, j_{k}}$, $C$ is a factor of the correlation matrix $\Sigma = C C^{'}$, where $\Sigma$ is estimated from the fitted Gaussian variogram, and where $\Phi(\dot)$ denotes the standard normal distribution function.

`WGBSSuite` induces spatial correlation amongst the $B_{i, j_{k}}$ in a less direct manner. Recall that each methylation locus is assigned one of the four underlying states ("de-methylated", "1st transition", "2nd transition", "methylated") via a hidden Markov model. The transition matrix of this hidden Markov model ensures that neighbouring loci, $i, (i+1)$, are more likely to be assigned the same state. Furthermore, since all loci within each of the four states are assigned the same (perturbed) underlying methylation level, $B_{i, j_{k}} = B_{region-type} + \epsilon_{i, j_{k}}$, neighbouring loci have similar methylation levels. Note that the $IPD$ only plays a direct role in the initial segmentation of the genome, not in the assigning of the $B_{i, j_{k}}$.

\subsubsection{Other model-based simulations}

- [x] DONE

A separate class of model-based simulation methods are those that simulate the $\beta$-values directly, i.e., without simulating sequencing depth \citet[e.g.,]{Jaffe:2012gx, Chen:2013eh, Chen:2014jb, }. These models are designed for simulating microarray data and not sequencing data, since they do not include the variability due to variation in sequencing depth. Of these methods, only \citet{Jaffe:2012gx} simulate correlation amongst the $B_{i, j_{k}}$.

\subsection{Methods based on re-sampling real data}

- [x] DONE

A simulation may also be based entirely on re-sampling of real data[^synthetic]. This type of simulated data is attractive because through careful sampling it can capture behaviour that is otherwise very difficult, if not impossible, to capture in a parametric model. At the same time, however, if the sampling units are poorly chosen or the sampling strategy is incorrect, then it may ignore these same features or, worse still, introduce artefacts into the simulated data.

[^synthetic]: This may also be referred to as creating a synthetic dataset.

Re-sampling methods are most easily implemented at the level of $\beta$-values. Re-sampling reads is reads is more difficult, except in the special case of down-sampling whereby the positions of reads are held constant but only a sub-sample of them are used in downstream analysis. Down-sampling is only really of interest to examine the effects of sequencing coverage on downstream analyses.

\citet{Sofer:2013bk} use a re-sampling based simulation method in the development of their `Aclust` software for identifying differential methylation from microarray data. The idea is adapted from \citet{Gaile:2007bq}, which is a simulation method for array comparative genomic hybridization experiments. \citet{Sofer:2013bk} sample "blocks" of CpGs, a region of the genome where all CpGs are within 10 kb of the next, to "generate (spatial) correlation-preserved methylation data". By sampling blocks rather than individual CpGs, this sampling scheme preserves the correlation structure between CpGs occurring in the same block. Two blocks are unlikely to be correlated since there is little evidence that CpGs separated by more than 10 kb have spatially correlated methylation levels.

\citet{Sofer:2013bk} sample from a dataset of 539 Illumina 450k microarrays and select a small number of "target" CpGs, which are CpGs whose methylation level is highly variable across the 539 samples. If a block does not contain a "target" then it is sampled uniformly at random from the 539 samples. However, if a block does contain a "target" then the sampling is weighted so that the "cases" are preferentially sampled from blocks with a high level of methylation at the "target" CpG and the "controls" are preferentially sampled from blocks with a low level of methylation at the "target" CpG[^cases_and_controls]. This is essentially weighted sampling of the real data to induce differential methylation.

[^cases_and_controls] The use of "cases" and "controls" is arbitrary, as is the choice of highly methylated for "cases" and lowly methylated for "controls" at "target" CpGs.

Due to the correlation structure of the $\beta$-values, the co-methylation, it is likely, although not guaranteed, that other CpGs in the blocks containing targets also display differential methylation.

\subsection{Summary}

- [x] DONE

My initial interest in simulation methods was to explore models of co-methylation, particularly at the level of individual DNA fragments. This requires two key capabilities:

1. Simulation of individual reads.
2. Simulation of co-methylation.

None of the published simulation methods satisfied both these requirements. This motivated the development of `methsim`.

`methsim` is specifically designed to model the co-methylation structure of bisulfite-sequencing data by simulating individual DNA fragments rather than directly simulating summary methylation measurements. In order to model the co-methylation structure, I decided that `methsim` should simulate at the level of individual DNA fragments. An added bonus of this approach is that `methsim` can (in theory) generate data at multiple resolutions: $\beta$-values, methylation patterns at m-tuples or entire sequencing reads. While my initial motivation in developing `methsim` was to explore models of co-methylation, I realised that it could also be used in the development and comparison of methods for the downstream analysis of DNA methylation data.

\section{Methods}

- [x] DONE

Simulating data with `methsim` involves 3 steps:

1. Simulate the true methylome of each sample.
2. Simulate reads, including sequencing error and bisulfite-conversion error, by sampling from the true methylome.
3. Construct the output, be it reads, methylation patterns at m-tuples or $\beta$-values.

`methsim` requires an input dataset from which to estimate key parameters. For the input dataset, `methsim` requires the methylation patterns at various sized m-tuples, which can be produced by the `methtuple` software.

\subsection{Simulating a single methylome}

- [x] DONE

As we have seen, DNA methylation is highly heterogeneous along the genome. Nonetheless, there are clear regions of "similarity", regions where we might hope to model it by a simple parametric model. `methsim` is based on the idea of segmenting the genome into "regions of similarity", fitting a simple model to each region and then "stitching" the results together. The idea of segmenting a globally heterogenous stochastic process into a series of locally homogeneous processes is not new. A hidden Markov model is an example of such a process; while the entire process may highly heterogeneous, conditional on the hidden states the process may be homogeneous. Hidden Markov models, and other models assuming local similarity in spite of global heterogeneity, have been used with great success in bioinformatics.

`methsim` takes the following approach to simulating the true underlying methylome for each sample:

1. Segment the methylome into regions of similarity.
2. Within each region simulate the binary methylation states from a simple parametric model.
3. "Stitch" the resulting region-level simulations together[^stitch].

[^stitch]: In practice, the stitching together of segments is incorporated into step 2.

Suppose our sample contains $n_{h}$ haplotypes and, therefore, that our sample can be represented by[^strand_symmetric] $\mathbf{Z} = \{ \mathbf{Z_{h}} \}_{h = 1}^{h = n_{h}}$. We might try to simulate this sample by generating $n_{h}$ independent realisations of $\mathbf{Z_{h}}$. However, for a number of reasons, this is a little naive and ambitious. For example, under this naive model there is no control across haplotypes of the methylation patterns; although there exists between-haplotype heterogeneity, the $\mathbf{Z_{h}}$ are definitely not independent of one another.

[^strand_symmetric]: Currently, `methsim` can only simulate CpG methylation and assumes that the methylation states of all CpGs are strand symmetric.

Rather than try to explicitly model dependencies between the $\mathbf{Z_{h}}$, `methsim` uses the following heuristic procedure:

1. Simulate $n_{\tilde{h}} \ll n_{h}$ independent realisations of $\mathbf{{Z_{\tilde{h}}}}$, where $\mathbf{Z_{\tilde{h}}}$ has the same stochastic properties as $\mathbf{Z_{h}}$. `methsim` uses a default value of $n_{\tilde{h}} = 8$.
2. Within each region, $r_{i}$, assign each of $\mathbf{Z_{\tilde{h}}}$ a weight, $W_{r_{i}, \tilde{h}}$, where $\sum_{ \tilde{h} = 1}^{n_{ \tilde{h}}} W_{r_{i}, \tilde{h}} = 1$ and, by construction, $W_{r_{i}, 1} \geq \ldots \geq W_{r_{i}, n_{\tilde{h}}}$.

For simplicity, `methsim` assumes that all $n_{\tilde{h}}$ realisations have the same parameters (discussed below; __TODO: Discuss implications of this assumption somewhere below__).

The weights are interpreted as the proportion of $\{ Z_{h, i} \}_{h = 1}^{h = n_{h}}$ with the methylation pattern given by $Z_{\tilde{h}, i}$ for the $i^{th}$ methylation locus[^weights]. We write the matrix of weights for all methylation loci in the genome as $\mathbf{W} = \{ \mathbf{W_{r_{i}, \cdot}} \}_{i = 1}^{N_{loci}} = \{ \mathbf{W_{\cdot, \tilde{h}}} \}_{\tilde{h} = 1}^{n_{\tilde{h}}}$. In regions of low within-sample heterogeneity, such as those unmethylated regions with $B_{i} \approx 0$, this means we want $\mathbf{W_{r_{i}, \cdot}} \approx (1, 0, \ldots, 0)$, whereas in regions of high within-sample heterogeneity, such as partially methylated regions, we want a more dispersed $\mathbf{W_{r_{i}}, \cdot}$.

[^weights]: Note that these weights are defined at the region-level so all loci in the region have the same weights.

A simulated methylome is therefore described by $\{ \mathbf{{Z_{\tilde{h}}}} \}_{\tilde{h} = 1}^{n_{\tilde{h}}}$ and $\mathbf{W}$. For example, $B_{i} = \sum_{\tilde{h} = 1}^{n_{\tilde{h}}} W_{r_{i}, \tilde{h}} \times Z_{i, \tilde{h}}$.

I now describe each step in this process in greater detail.

\subsubsection{Segmenting the methylome}

- [x] DONE

`methsim` uses the the R/Bioconductor package, `MethylSeekR` \citep{Burger:2013kq}, to segment the input methylome into these regions of similarity. Initially published to discover regulatory motifs from bisulfite-sequencing data, `MethylSeekR` segments the genome into unmethylated regions (\emph{UMR}s), lowly-methylated regions (\emph{LMR}s) and partially methylated regions[^pmds] (\emph{PMR}s). It does this using a two-stage algorithm applied to the $\beta$-values from a sample:

[^pmds]: Partially methylated regions are also commonly known as partially methylated domains, but I refer to them as regions for consistency with UMRs and LMRs.

1. Identify partially methylated regions. A summary statistic, $\alpha$, which is based on the $\beta$-values in a sliding window of 100 CpGs, is used to identify PMRs. Briefly, a two-state hidden Markov model is fit to the $\alpha$ values to identify PMRs and non-PMRs.
2. Identify UMRs and LMRs. The PMRs are masked from the genome and simple heuristics are used to identify UMRs and LMRs based on the average $\beta$-values in a window and the number of CpGs in the window.

`methsim` uses the functionality of `MethylSeekR` to partition the genome into "regions of similarity", namely `UMR`, `LMR`, `PMR` or `other`[^other]. While a tailored algorithm may improve the segmentation process, the result produced by `MethylSeekR` is a reasonable approximation to segmenting into "regions of similarity". This can be seen in __FIGURES__  (__TODO: Figures showing that `MethylSeekR` gives a reasonable segmentation for the purposes of `methsim`)__). It should be noted that `methsim` also performs some post-processing of the `MethylSeekR` output so that the resulting segmentation is disjoint and exhaustive.

[^other]: `other` roughly corresponds to 'mostly methylated regions', although \citet{Burger:2013kq} do not describe these regions as such.

\subsubsection{Simulating $\{ \mathbf{{Z_{\tilde{h}}}} \}_{\tilde{h} = 1}^{n_{\tilde{h}}}$}

- [x] DONE

`methsim` uses a two-state ($1 =$ methylated or $0 =$ unmethylated), one-step Markov chain to model $\mathbf{{Z_{\tilde{h}}}}$. This allows the incorporation of within-fragment co-methylation into the simulation. The choice of a one-step process is one of computational simplicity, although I will show that this is a reasonable approximation of within-fragment co-methylation.

In Chapter \ref{chap:co-methylation} I showed that the strength of within-fragment co-methylation varies as a function of the intra-pair distance and by the genomic context. I also showed that for a pair of CpGs separated by given IPD, and in the same genomic context, that there is a large amount of variability in the strength of within-fragment co-methylation.

For these reasons I allow the transition probabilities to vary with $i$ and $r_{i}$, that is, I allow the Markov chain to be spatially inhomogeneous. In particular, I allow the transition probabilities, $p_{i^{(h, j)}, a, b} = Pr(Z_{h, i + 1, j} = b | Z_{h, i, j} = a)$, to depend on the intra-pair distance between the $i^{th}$ and $(i + 1)^{th}$ locus and on the region for this pair of loci[^context], $p_{i^{(h, j)}, a, b} = Pr(Z_{h, i + 1, j} | Z_{h, i, j}, r_{i}, IPD(i, i + 1))$, where $r_{i}$ is the region type (UMR, LMR, PMR or MMR) of the $i^{th}$ methylation locus.

[^context]: Actually, it only depends on the region of the $i^{th}$ locus. Most pairs of loci, $(i, i + 1)$, will lie in the same region. For pairs that span the boundary of two different regions I have arbitrarily chosen to use the region of the $i^{th}$ locus.

The above-described model, being spatially inhomogeneous, is not particularly amenable to analytical calculations. However, it is very simple to simulate realisations from this model, requiring only a single loop over the set of loci on each chromosome. For a chromosome containing $n$ methylation loci there are $n - 1$ transition matrices, $P_{i^{(h)}}$, to estimate or otherwise assign. Again, I emphasise that `methsim` uses the same set of parameters for all $n_{\tilde{h}}$ "pseudo-haplotypes". Therefore, only one set of transition matrices need to be assigned.

`methsim` assigns these transition probabilities based on $\beta_{i | r_{i}}$ and $LOR_{(i, i+1) | r_{i}, IPD(i, i+1)}$, the empirical distributions of $\beta_{i}$ and $LOR_{IPD(i, i + 1)}$ conditional on the region type of the $i^{th}$ locus, $r_{i}$, and the intra-pair distance of the the $i^{th}$ and $(i + 1)^{th}$ locus, $IPD(i, i+1)$, respectively. These empirical distributions are shown for the \emph{ADS} sample in Figures \ref{fig:ADS_beta_by_pm_boxplots} and \ref{fig:ADS_lor_by_pm_quantile_plots}.

\begin{figure}[!htb]
\centering
\includegraphics[width=\textwidth]{../figures/ADS_beta_by_pm_boxplots.pdf}
\caption{$\beta_{i | r_{i}}$, the distribution of $\beta$-values for CpGs in each region type, for the \emph{ADS} sample. Only CpGs with at least $10\times$ sequencing coverage are used. UMR = unmethylated region; LMR = lowly methylated region; PMR = partially methylated region; other = any other region.}
\label{fig:ADS_beta_by_pm_boxplots}
\end{figure}

\begin{figure}[!htb]
\centering
\includegraphics[width=\textwidth]{../figures/ADS_lor_by_pm_quantile_plots.pdf}
\caption{$LOR_{(i, i + 1) | r_{i}}$, the distribution of $\log_{2}$ odds-ratio estimates of within-fragment co-methylation for CpGs in each region type, for the \emph{ADS} sample. Only 2-tuples with at least $10\times$ sequencing coverage are used. UMR = unmethylated region; LMR = lowly methylated region; PMR = partially methylated region; other = any other region.}
\label{fig:ADS_lor_by_pm_quantile_plots}
\end{figure}

These same distributions for all the _EPISCOPE_, _Lister_, _Seisenberger_ and _Ziller_ samples are shown in Figures \ref{fig:EPISCOPE_beta_by_pm_boxplots}, \ref{fig:EPISCOPE_lor_by_pm_quantile_plots}, \ref{fig:Lister_beta_by_pm_boxplots}, \ref{fig:Lister_lor_by_pm_quantile_plots},
\ref{fig:Seisenberger_beta_by_pm_boxplots}, \ref{fig:Seisenberger_lor_by_pm_quantile_plots},
\ref{fig:Ziller_merged_beta_by_pm_boxplots} and \ref{fig:Ziller_merged_lor_by_pm_quantile_plots}.

\begin{figure}[!htb]
\centering
\includegraphics[width=\textwidth]{../figures/EPISCOPE_beta_by_pm_boxplots.pdf}
\caption{$\beta_{i | r_{i}}$, the distribution of $\beta$-values for CpGs in each region type, for the \emph{EPISCOPE} dataset. Only CpGs with at least $10\times$ sequencing coverage are used. UMR = unmethylated region; LMR = lowly methylated region; PMR = partially methylated region; other = any other region.}
\label{fig:EPISCOPE_beta_by_pm_boxplots}
\end{figure}

\begin{figure}[!htb]
\centering
\includegraphics[width=\textwidth]{../figures/EPISCOPE_lor_by_pm_quantile_plots.pdf}
\caption{$LOR_{(i, i + 1) | r_{i}}$, the distribution of $\log_{2}$ odds-ratio estimates of within-fragment co-methylation for CpGs in each region type, for the \emph{EPISCOPE} sample. Only 2-tuples with at least $10\times$ sequencing coverage are used. UMR = unmethylated region; LMR = lowly methylated region; PMR = partially methylated region; other = any other region.}
\label{fig:EPISCOPE_lor_by_pm_quantile_plots}
\end{figure}

\begin{figure}[!htb]
\centering
\includegraphics[width=\textwidth]{../figures/Lister_beta_by_pm_boxplots.pdf}
\caption{$\beta_{i | r_{i}}$, the distribution of $\beta$-values for CpGs in each region type, for the \emph{Lister} dataset. Only CpGs with at least $10\times$ sequencing coverage are used. UMR = unmethylated region; LMR = lowly methylated region; PMR = partially methylated region; other = any other region.}
\label{fig:Lister_beta_by_pm_boxplots}
\end{figure}

\begin{figure}[!htb]
\centering
\includegraphics[width=\textwidth]{../figures/Lister_lor_by_pm_quantile_plots.pdf}
\caption{$LOR_{(i, i + 1) | r_{i}}$, the distribution of $\log_{2}$ odds-ratio estimates of within-fragment co-methylation for CpGs in each region type, for the \emph{Lister} sample. Only 2-tuples with at least $10\times$ sequencing coverage are used. UMR = unmethylated region; LMR = lowly methylated region; PMR = partially methylated region; other = any other region.}
\label{fig:Lister_lor_by_pm_quantile_plots}
\end{figure}

\begin{figure}[!htb]
\centering
\includegraphics[width=\textwidth]{../figures/Seisenberger_beta_by_pm_boxplots.pdf}
\caption{$\beta_{i | r_{i}}$, the distribution of $\beta$-values for CpGs in each region type, for the \emph{Seisenberger} dataset. Only CpGs with at least $10\times$ sequencing coverage are used. UMR = unmethylated region; LMR = lowly methylated region; PMR = partially methylated region; other = any other region.}
\label{fig:Seisenberger_beta_by_pm_boxplots}
\end{figure}

\begin{figure}[!htb]
\centering
\includegraphics[width=\textwidth]{../figures/Seisenberger_lor_by_pm_quantile_plots.pdf}
\caption{$LOR_{(i, i + 1) | r_{i}}$, the distribution of $\log_{2}$ odds-ratio estimates of within-fragment co-methylation for CpGs in each region type, for the \emph{Seisenberger} sample. Only 2-tuples with at least $10\times$ sequencing coverage are used. UMR = unmethylated region; LMR = lowly methylated region; PMR = partially methylated region; other = any other region.}
\label{fig:Seisenberger_lor_by_pm_quantile_plots}
\end{figure}

\begin{figure}[!htb]
\centering
\includegraphics[width=\textwidth]{../figures/Ziller_merged_beta_by_pm_boxplots.pdf}
\caption{$\beta_{i | r_{i}}$, the distribution of $\beta$-values for CpGs in each region type, for the \emph{Ziller} dataset. Only CpGs with at least $10\times$ sequencing coverage are used. UMR = unmethylated region; LMR = lowly methylated region; PMR = partially methylated region; other = any other region.}
\label{fig:Ziller_merged_beta_by_pm_boxplots}
\end{figure}

\begin{figure}[!htb]
\centering
\includegraphics[width=\textwidth]{../figures/Ziller_merged_lor_by_pm_quantile_plots.pdf}
\caption{$LOR_{(i, i + 1) | r_{i}}$, the distribution of $\log_{2}$ odds-ratio estimates of within-fragment co-methylation for CpGs in each region type, for the \emph{Ziller} sample. Only 2-tuples with at least $10\times$ sequencing coverage are used. UMR = unmethylated region; LMR = lowly methylated region; PMR = partially methylated region; other = any other region.}
\label{fig:Ziller_merged_lor_by_pm_quantile_plots}
\end{figure}

`methsim` assigns these probabilities as follows:

1. For each region, $r_{i}$, sample $B_{r_{i}}$ from $\beta_{i | r_{i}}$. This can be interpreted as $Pr(Z_{r_{i}} = 1)$, the marginal probability that a methylation locus in region $r_{i}$ is methylated, or, alternatively, as the average level of methylation for all methylation loci in the region[^extreme_B].
2. For every pair of methylation loci, $(i, i + 1)$, $1 \leq i \leq (n - 1)$, sample $\Psi_{(i, i + 1)}$ from $LOR_{(i, i+1) | r_{i}, IPD(i, i+1)}$. This can be interpreted as the odds that the $(i + 1)^{th}$ methylation locus has the same methylation state as the $i^{th}$ methylation locus.

[^extreme_B]: If $B_{r_{i}} = 0$ (resp. $B_{r_{i}} = 1$), `methsim` adds (resp. subtracts) a small value $\epsilon$. The default value is $\epsilon = 0.01$. The boundary values of $0$ and $1$ are common in $\beta_{i | r_{i}}$ and, were these retained, would result in methylomes with large regions that are uniformly unmethylated (resp. methylated).

The second step requires further explanation. If there are no observations with the required $IPD$-region combination, then, by definition, `methsim` is unable to sample from the appropriate empirical distribution[^min_n]. Instead, when this occurs, `methsim` samples from a user-specified parametric distribution. __RESULTS SECTION__ explores different choices of this distribution __AND SHOWS WHAT__. The default option is a zero-mean Gaussian distribution, chosen so that "on average" the methylation state at adjacent CpGs with large $IPD$s are independent of one another. The default standard deviation of this Gaussian distribution is $2$; this value, while fairly arbitrary, reflects the median standard deviation of the $LOR_{(i, i+1) | r_{i}, IPD(i, i+1)}$ distributions for each sample, which is remarkably constant across $r_{i}$ and samples[^ipd_stability] (__TODO: Format and reference table__).

[^min_n]: By default, `methsim` requires at least 100 observations in the  $LOR_{(i, i+1) | r_{i}, IPD(i, i+1)}$ to try to avoid sampling from sparse empirical distributions.

[^ipd_stability]: Figures \ref{fig:EPISCOPE_lor_by_pm_quantile_plots}, \ref{fig:Lister_lor_by_pm_quantile_plots}, \ref{fig:Seisenberger_lor_by_pm_quantile_plots} and \ref{fig:Ziller_merged_lor_by_pm_quantile_plots} show that the variability of these distributions is fairly constant across $IPD$ for a given $r_{i}$ and sample.

```{r, eval = TRUE, echo = FALSE}
knitr::kable(readRDS("../data/median_sd_LOR_distributions.rds"), align = "c", digits = 2)
```

Now that we have $B_{r_{i}}$ and $\Psi_{i, i + 1}$ for $1 \leq i \leq n - 1$ we can simulate realisations of $\mathbf{Z_{\tilde{h}}}$. To begin the process (i.e., when $i = 1$) $Z_{\tilde{h}, 1}$, is simulated from the "marginal" distribution, which is Bernoulli($B_{r_{1}}$). Then, we loop over $i = 2, \ldots, n - 1$ to simulate the remainder of the process. At each step we compute the transition probability, $p_{(i + 1) | i} = Pr(Z_{\tilde{h}, i + 1} = 1 | Z_{\tilde{h}, i}) = \frac{Pr(Z_{tilde{h}, i}, Z_{tilde{h}, i + 1} = 1)}{Pr(Z_{tilde{h}})}$, from $B_{r_{i}}, B_{r_{i + 1}}, \Psi_{(i, i + 1)}$ and the observed (simulated) value at the previous iteration, $Z_{\tilde{h}, i}$. Then, $Z_{\tilde{h}, i + 1}$ is simulated from the Bernoulli($p_{(i + 1) | i}$) distribution.

In order to compute the transition probability, $p_{(i + 1) | i}$, `methsim` first uses the iterative proportional fitting algorithm \citet[see]{Speed:2005tg}[for an overview of the algorithm] to construct the joint probability matrix,

\begin{equation*}
P = \bordermatrix{~ & Pr(Z_{i+1} = 0) = 1 - B_{r_{i+1}} & Pr(Z_{i+1} = 1) = B_{r_{i+1}} \cr
Pr(Z_{i} = 0) = 1 - B_{r_{i}} & Pr(Z_{i} = 0, Z_{i + 1} = 0) & Pr(Z_{i} = 0, Z_{i + 1} = 1) \cr
Pr(Z_{i} = 1) = B_{r_{i}} & Pr(Z_{i} = 1, Z_{i + 1} = 0) & Pr(Z_{i} = 1, Z_{i + 1} = 1) \cr}
\end{equation*}

Iterative proportional fitting is a general method "for constructing tables of numbers satisfying certain constraints" \citep{Speed:2005tg}. `methsim` uses iterative proportional fitting to construct the _unique_ $2 \times 2$ array (the joint probability matrix) with specified margins (the marginal probabilities of each methylation state at each locus, $B_{r_{i}} = Pr(Z_{i} = 1), B_{r_{i + 1}} = Pr(Z_{i + 1} = 1)$) and the specified cross-product ratio (the odds-ratio, $\psi_{i, i + 1} = \log_{2}(\Psi_{i, i + 1})$).

Once $P$ is formed, the appropriate element of it is divided by the appropriate marginal probability to obtain the transition probability. An example of this procedure is illustrative.

Let $\Psi_{i, i + 1} = 2$, and therefore $\psi = 2^{2} = 4$, and $B_{r_{i}} = B_{r_{i + 1}} = 0.7$. To begin the iterative proportional fitting algorithm, form the matrix $P^{(0)} = \Bigl(\begin{smallmatrix}
\psi & 1 \\
   1 & 1
\end{smallmatrix} \Bigr).$ This matrix has the desired cross-ratio, $\psi$, but not the desired row and column margins. At each iteration of the algorithm, iterative proportional fitting adjusts the rows and columns such that the cross-ratio remains $\psi$ while the row and column margins converge towards the desired values. The algorithm continues in this manner, forming a series of $2 \times 2$ tables that converge pointwise (and uniquely) to a $2 \times 2$ table $P^{(*)} \equiv P$. For our example, $P^{(*)} = \Bigl(\begin{smallmatrix}
0.122 & 0.178 \\
0.178 & 0.522
\end{smallmatrix} \Bigr)$, to 3 decimal places of precision. It is easily verified that $P^{(*)}$ has the desired cross-ratio and marginal sums.

In `methsim`, the routine to simulate $\{ \mathbf{{Z_{\tilde{h}}}} \}_{\tilde{h} = 1}^{n_{\tilde{h}}}$ is available as the `simulate,SimulateMethylomeParam-method`. It is mostly written in C++ using the wonderful Rcpp package \citep{Eddelbuettel:2011to, Eddelbuettel:2013if}, which greatly speeds up the running time.

\subsubsection{Simulating $\mathbf{W}$}

- [x] DONE

For each region we need to assign the vector of weights, $\mathbf{W_{r_{i}, \cdot}}$. To do so, `methsim` uses a heuristic procedure. For higher-order m-tuples, $m = 3$ or $4$ is recommended, `methsim` computes the relative frequency of the $2^{m}$ possible methylation patterns at each m-tuple with sufficient sequencing coverage. `methsim` then samples from this distribution, stratified by region-type; call these empirical distributions $w_{(i, \ldots, i + m - 1) | r_{(i, \ldots, i + m - 1)}}$. Figure \ref{fig:ADS_pattern_freqs_by_pm_boxplots} shows elementwise boxplots of $w_{(i, \ldots, i + m - 1) | r_{(i, \ldots, i + m - 1)}}$, that is, it shows the distributions $w_{i | r_{(i, \ldots, i + m - 1)}}$ for $i = 1, \ldots, 2^{m}$.

\begin{figure}[!htb]
\centering
\includegraphics[width=\textwidth]{../figures/ADS_pattern_freqs_by_pm_boxplots.pdf}
\caption{$w_{(i, \ldots, i + m - 1) | r_{(i, \ldots, i + m - 1)}}, i = 1, \ldots, 2^{m} = 8$, the distribution of the relative frequencies for all CpG 3-tuples in each region type, for the \emph{ADS} sample. Only CpGs with at least $5\times$ sequencing coverage are used. UMR = unmethylated region; LMR = lowly methylated region; PMR = partially methylated region; other = any other region.}
\label{fig:ADS_pattern_freqs_by_pm_boxplots}
\end{figure}

These same distributions for all the _EPISCOPE_, _Lister_, _Seisenberger_ and _Ziller_ samples are shown in Figures \ref{fig:EPISCOPE_pattern_freqs_by_pm_boxplots}, \ref{fig:Lister_pattern_freqs_by_pm_boxplots}, \ref{fig:Seisenberger_pattern_freqs_by_pm_boxplots} and  \ref{fig:Ziller_merged_pattern_freqs_by_pm_boxplots} (__TODO: Are these other examples necessary?__).

\begin{figure}[!htb]
\centering
\includegraphics[width=\textwidth]{../figures/EPISCOPE_pattern_freqs_by_pm_boxplots.pdf}
\caption{$w_{(i, \ldots, i + m - 1) | r_{(i, \ldots, i + m - 1)}}, i = 1, \ldots, 2^{m} = 8$, the distribution of the relative frequencies for all CpG 3-tuples in each region type, for the \emph{EPISCOPE} dataset. Only CpGs with at least $5\times$ sequencing coverage are used. UMR = unmethylated region; LMR = lowly methylated region; PMR = partially methylated region; other = any other region.}
\label{fig:EPISCOPE_pattern_freqs_by_pm_boxplots}
\end{figure}

\begin{figure}[!htb]
\centering
\includegraphics[width=\textwidth]{../figures/Lister_pattern_freqs_by_pm_boxplots.pdf}
\caption{$w_{(i, \ldots, i + m - 1) | r_{(i, \ldots, i + m - 1)}}, i = 1, \ldots, 2^{m} = 8$, the distribution of the relative frequencies for all CpG 3-tuples in each region type, for the \emph{Lister} dataset. Only CpGs with at least $5\times$ sequencing coverage are used. UMR = unmethylated region; LMR = lowly methylated region; PMR = partially methylated region; other = any other region.}
\label{fig:Lister_pattern_freqs_by_pm_boxplots}
\end{figure}

\begin{figure}[!htb]
\centering
\includegraphics[width=\textwidth]{../figures/Seisenberger_pattern_freqs_by_pm_boxplots.pdf}
\caption{$w_{(i, \ldots, i + m - 1) | r_{(i, \ldots, i + m - 1)}}, i = 1, \ldots, 2^{m} = 8$, the distribution of the relative frequencies for all CpG 3-tuples in each region type, for the \emph{Seisenberger} dataset. Only CpGs with at least $5\times$ sequencing coverage are used. UMR = unmethylated region; LMR = lowly methylated region; PMR = partially methylated region; other = any other region.}
\label{fig:Seisenberger_pattern_freqs_by_pm_boxplots}
\end{figure}

\begin{figure}[!htb]
\centering
\includegraphics[width=\textwidth]{../figures/Ziller_merged_pattern_freqs_by_pm_boxplots.pdf}
\caption{$w_{(i, \ldots, i + m - 1) | r_{(i, \ldots, i + m - 1)}}, i = 1, \ldots, 2^{m} = 8$, the distribution of the relative frequencies for all CpG 3-tuples in each region type, for the \emph{Ziller} dataset. Only CpGs with at least $5\times$ sequencing coverage are used. UMR = unmethylated region; LMR = lowly methylated region; PMR = partially methylated region; other = any other region.}
\label{fig:Ziller_merged_pattern_freqs_by_pm_boxplots}
\end{figure}

From Figures \ref{fig:EPISCOPE_pattern_freqs_by_pm_boxplots}, \ref{fig:Lister_pattern_freqs_by_pm_boxplots}, \ref{fig:Seisenberger_pattern_freqs_by_pm_boxplots} and \ref{fig:Ziller_merged_pattern_freqs_by_pm_boxplots} we see that, unsurprisingly, in UMRs that a single methylation pattern dominates, the completely unmethylated pattern. Outside of UMRs, we see that while there is generally a dominant methylation pattern at each m-tuple, there may be $2-4$ otheres that are observed at non-negligible frequencies. PMRs are the most heterogeneous and LMRs and other regions are quite similar.

The above procedure has several obvious limitations. Firstly, we are restricted to analying m-tuples within individual reads and read-pairs. Secondly, as $m$ increases, and similarly as $IPD$ increases, there are fewer m-tuples with sufficient sequencing coverage. This means our data are heavily biased towards regions of high CpG density, which typically correspond to UMRs. Additionally, we know that m-tuples with small $IPDs$ are far more likely to be co-methylated than those with larger $IPDs$ and so we are biased towards sampling entirely unmethylated and entirely methylated methylation patterns. It is difficult to know how to adjust for these biases, which are unavoidable given current sequencing technologies. These estimates very likely underestimate the degree of within-sample heterogeneity and, consequently, weights based on sampling from the empirical distributions, $w_{(i, \ldots, i + m - 1) | r_{(i, \ldots, i + m - 1)}}$, likely result in overly homogeneous simulated methylomes.

A second, more subtle, bias is due to sequencing errors, bisulfite-conversion errors and alignment errors. Using the simple relative frequency estimation procedure will result in errors like these being inferred as (rare) methylation patterns. `MPFE` ([http://bioconductor.org/packages/release/bioc/html/MPFE.html](http://bioconductor.org/packages/release/bioc/html/MPFE.html)) is an R package that uses an estimation procedure designed to deal with some of these errors. However, `MPFE` is intended for analysing amplicon bisulfite-sequencing data with high sequencing coverages, and it's more computationally demanding estimation procedure does not scale to whole-genome data.

With all the above limitations in mind, `methsim` still uses the simple strategy of sampling the $\mathbf{W}$ from $w_{(i, \ldots, i + m - 1) | r_{(i, \ldots, i + m - 1)}}$. At present I do not have a better alternative and, as we shall see, the current procedure still procedure still produced realistic simulated data. In the __RESULTS SECTION__ I also explore a simulation that simulates $n_{\bar{h}}$ independent realisations of $\mathbf{Z}$ where $n_{\tilde{h}} < n_{\bar{h}} \\l n_{h}$ and uniformly sets $\mathbf{W_{r_{i}, \cdot}} = (\frac{1}{n_{\bar{h}}}, \ldots, \frac{1}{n_{\bar{h}}})$.

\subsection{Simulating reads}

- [x] DONE

Once we have our 'true' methylome, we want to simulate an assay of this sample. In theory this could be any type of methylation assay, but here we focus on simulating whole-genome bisulfite-sequencing data.

`methsim` uses a simple method for simulating bisulfite-sequencing. The user specifies the desired read length and average sequencing coverage[^read_length], the error rate ($\epsilon$, which includes both sequencing error and bisulfite-conversion error), and provides the simulated 'true' methylome.

The number of reads required is given by $n_{reads} = \frac{\text{average sequencing coverage}}{\text{read length}} \times \text{size of genome}$. The number of reads per chromosome is assigned proportional to the chromosome length. Then, the start of each read is sampled uniformly across the chromosome respective chromosome.

[^read_length]: Currently only single-end data are supported. This isn't a big issue. Most paired-end bisulfite-sequencing datasets have overlapping mates, and so paired-end data can be crudely simulated by simply doubling the read length. Also, all reads must have the same length, i.e., no simulation of read trimming. I don't consider this feature a priority, but it could easily be implemented by generating read lengths from a probability distribution, e.g., Poisson.

For all reads that overlap a methylation locus in the 'true' methylome, a pseudo-haplotype is sampled with probabilities given by $W_{i, \cdot}$. The methylation pattern of the sampled pseudo-haplotype is then copied, with probability $\epsilon$ that the methylation state at each locus is switched.

The result is an object storing the genomic co-ordinates and methylation patterns of every read that contains a methylation locus.

\subsection{Construct the output}

- [x] DONE

While the `SimulatedBS` object contains all the data on the methylation states of the reads, it is not always the most convenient or efficient format with which to work. For example, many downstream analyses only make use of $\beta$-values, so we might want to summarise our simulated data this way. `methsim` provides the `asMethPat()` function to convert the `SimulatedBS` object to a `MethPat` object containing methylation patterns at m-tuples of a given size; all the functionality of the `MethylationTuples` package is then available to the user, such as computing $\beta$-values with the `methLevel()` method.

\subsection{Simulating multiple samples}

- [x] DONE

The above describes how to simulate a single whole-genome bisulfite-sequencing sample. If we wanted to simulate six independent samples then we could simply run this procedure six times, with a different `SimulateMethylomeParam` object for each realisation. However, we will generally be interested in simulating experiments where the samples are not  is some relationship amongst the samples. For example, suppose we want to simulate a two-group experiment with three samples per group. Further, suppose we want to include differentially methylated regions between the two groups. How do we simulate such an experiment?

To simulate these type of experiments, we need to be able to control the variation between the `SimulatedBS` objects. In this example, we probably want the three samples in each group to be more similar to each other than to the three samples in the other group. Essentially, we need to be able to control the within-group and between-group variation. `methsim` controls this variation through the workflow show in __FIGURE__. If we want two samples two be very different from one another, then we would use two different `SimulateMethylomeParam` objects at the beginning. However, if we want to simulate two technical replicates, then we would use the same `SimulateMethylomeParam`, `SimulatedMethylome` and `SimulateBSParam` objects; the two samples have identical 'true' methylomes and they only diverge at the 'sequencing' step of the simulation.

We might also consider more subtle ways of introducing variation. For example, we might use the same partitioning of the methylome and the same average methylation level for each region but vary the strength of co-methylation. This could be achieved by altering the $LOR_{(i, i+1) | r_{i}, IPD(i, i+1)}$ distributions between two `SimulateMethylomeParam` objects while holding the rest constant. Or we might vary the weights, $W$, to simulate more or less heterogeneous samples.

In __RESULTS SECTION__ I explore how much variation is created using these different techniques.

\subsubsection{Simulating differential methylation}

- [ ] DONE

An obvious application of `methsim` is to simulate data to be used in a study comparing methods for identifying differential methylation. This requires a particularly controllable method for inducing differences in `SimulatedMethylome` objects. The user wants to provide a set of methylation loci, either individual loci or as regions, and the expected difference in methylation levels at these regions.

__Scattered thoughts below__

1. Individual loci: Want to specify $B_{i} = \sum_{\tilde{h} = 1}^{n_{\tilde{h}}} W_{r_{i}, \tilde{h}} \times Z_{i, \tilde{h}}$, i.e., fix $\mathbf{W_{i, \cdot}}$ and $\mathbf{Z_{i, \cdot}}$.
2. Regionwise: Want to specify $mean(B_{r_{i}})$ (i.e., average $B$ across all loci in $r_{i}$). Could do this by applying (1) to all loci in the region or by specifying $B_{r_{i}}$ and hoping it just works out in the simulation.

Perhaps both could be achieved by holding $Z$ fixed and varying $W$? Is that like assuming that the heterogeneity is causing the DMRs (or will it be interpreted that way)? Would tweaking $W$ to guarantee a particular $B$ be a computationally feasible strategy? Or would it be easier to specify the desired $B$ and simulate the region (this wouldn't guarantee $B$ is the desired value but should equal it "on average")?


\subsection{Implementation}

- [x] DONE

`methsim` is written in R and builds on the `MethylationTuples` package described in __CHAPTER__, as well as several R packages available on Bioconductor and CRAN (see `DESCRIPTION` for a complete list. __TODO: Add citations of all packages used in thesis__).

`methsim` makes extensive use of the S4 object system. The most important classes defined by `methsim` are `SimulateMethylomeParam`, `SimulatedMethylome`, `SimulateBSParam` and `SimulatedBS`. The most important methods are the `simulate()` methods defined for the `SimulateMethylomeParam` and `SimulateBSParam` classes. __FIGURE__ illustrates the `methsim` workflow to create simulate whole-genome bisulfite-sequencing data.

Because `methsim` simulates data by sampling from empirical distributions of key statistics, it requires that the user supplies data from a relevant sample. To help a new user get started, `methsim` includes data for the _ADS_, _ADS-adipose_ and _ADS-iPSC_ samples in the form of `SimulateMethylomeParam` objects. Alternatively, the user may process the BAM file for their sample with `methtuple` and then use the helper functions in `methsim` to construct a `SimulateMethylomeParam` object.

The next step is to simulate a 'true' methylome by applying the `simulate()` method to the `SimulateMethylomeParam` object. This returns a `SimulatedMethylome`, a schematic of which is __SHOWN BELOW__. The `SimulatedMethylome` object forms part of the `SimulateBSParam` object from which a `SimualtedMethylome` is created by calling the `simulate()` method.

\subsubsection{Performance}

- [x] DONE

Much of `methsim` is written in C++, using the wonderful Rcpp package \citep{Eddelbuettel:2011to, Eddelbuettel:2013if}, which greatly speeds up the running time of key procedures. Furthermore, several steps of the simulation can be run in parallel for each chromosome. Parallelism is implemented via the `BiocParallel` Bioconductor package \citep{BiocParallel}.

The `SimulatedMethylome` and `SimulatedBS` objects are large in memory. This is to some extent unavoidable; we are simulating entire methylomes and whole-genome sequencing experiments. Some improvements in memory usage are possible by alternative implementations of key classes, however, these are generally at the expense of more complicated code and longer running times of associated methods. For example, long and repetitive vectors may be efficiently represented by run length encodings, however, it is slower to extract elements from these than simple numeric vectors.

Shown below are the running times and memory usage involved in simulating a $30\times$ whole-genome bisulfite-sequencing sample using the _ADS_ `SimulateMethylomeParam` object included in `methsim`. Using eight CPU cores in parallel, it takes approximately seven minutes to simulate the underlying 'true' methylome and a further nine minutes to simulate a single whole-genome bisulfite-sequencing assay of this methylome.

These results are indicative of simulating high-coverage whole-genome bisulfite-sequencing for human-sized genomes. Simulations using a smaller genome or lower sequencing coverage will have faster running times and lower memory usage.

```r
# Load necessary packages
suppressPackageStartupMessages(library(methsim))
suppressPackageStartupMessages(library(BiocParallel))
suppressPackageStartupMessages(library(pryr))

# Use up to 8 cores in parallel computations.
BPPARAM <- MulticoreParam(workers = 8L)

# Set seed for reproducibility.
set.seed(666)

# ADS: the SimulateMethylomeParam object.
pryr::object_size(ADS)
#> 22.5 MB

# sm: the SimulatedMethylome object
system.time(sm <- simulate(ADS, BPPARAM = BPPARAM))
#>    user  system elapsed
#> 839.505  19.612 417.271
pryr::object_size(sm)
#> 2.93 GB

# sbsp: the SimulatedBSParam object.
# NOTE: Basically all the memory used by the SimulatedBSParam object is the
# SimulatedMethylome object.
sbsp <- SimulateBSParam(SimulatedMethylome = sm,
                        aveCov = 30,
                        errorRate = 0.01,
                        target = NULL)
pryr::object_size(sbsp)
#> 2.93 GB

# sbs: the SimulatedBS object.
system.time(sbs <- simulate(sbsp, BPPARAM = BPPARAM))
#>     user   system  elapsed
#> 2716.503  174.591  548.122
pryr::object_size(sbs)
#> 10.1 GB

# methpat: the MethylationTuples::MethPat object of one-tuples.
# NOTE: information is lost when a SimulatedBS object is coerced to a MethPat
# object.
system.time(methpat <- asMethPat(sbs,
                                 sampleName = "sim_ADS",
                                 size = 1L,
                                 BPPARAM = BPPARAM))
#>    user  system elapsed
#> 557.900  34.779  48.944
pryr::object_size(methpat)
#> 451 MB
```

\section{Results}

-  __TODO: Figure showing 'rest of genome' is MMRs. More generally, a figure showing the average methylation level in each of `UMR`, `LMR`, `PMR` and `other`.


\subsection{Case study}

- [ ] DONE

\section{Conclusions}

- [ ] DONE

\subsubsection{Comparison to other methods}

\citet{Lacey:2013iy} supply R scripts to implement their simulation method ([http://rrbs-sim.r-forge.r-project.org/](http://rrbs-sim.r-forge.r-project.org/)). __TODO: Try these scripts, see how long they take__. The running times reported in the original publication are shown in __TABLE__ for a single simulation. __TODO: To simulate a single RRBS dataset would take HOW LONG__. The simulation scripts could be made more user-friendly by creating an R package, rather than supplying the software as a collection of scripts, and would benefit from some documentation.

\begin{table}[h]
\centering
\begin{tabular}{@{}lll@{}}
\toprule
Number of samples & Number of sites & Running time (seconds) \\ \midrule
8                 & 1000            & \textless 1            \\
8                 & 5000            & 13                     \\
8                 & 10000           &                        \\ \bottomrule
\end{tabular}
\caption{Reported running times for simulation methodology of \citet{Lacey:2013iy}. Simulations were run on "on a desktop with a 3.4 GHz processor and 10 GB RAM running 64-bit Windows 7".}
\label{tab:Lacey_runtime}
\end{table}

__TODO: Compare `methsim` to `WGBSSuite`.


# TODO:

- [ ] Decide on a variable for sequencing depth; $m_{i, j} + u_{i, j}$ is kinda confusing. Perhaps $d_{i, j} / D_{i, j}$?
