# Thesis outline

## Thesis title

Statistical analysis of high-throughput assays for studying DNA methylation

## Research questions

# Chapter outlines

For each chapter I outline my plans and add checkboxes for what is and is not drafted.

## Preamble

Do last.

## Introduction

- [x] Chapter overview
- [ ] DNA
  - [ ] basic structure, mitotic replication, DNA -> nucleosomes -> chromatin -> chromosomes -> genome, DNA -> RNA -> protein.
- [ ] Epigenetics
- [ ] DNA methylation
  - Mostly done.
- [ ] Assays for studying DNA methylation
  - Mostly done.
- [ ] WGBS
- [ ] Outline of thesis

## Analysis of bisulfite-sequencing data

This chapter is mostly complete. Need to re-work the intro a little to be more general than differential methylation experiments.

- [x] Chapter overview
- [x] Data quality control checks
- [x] Read mapping and post-processing of mapped reads
- [x] Methylation calling
- [ ] Analysis

## A statistical framework for analysing bisulfite sequencing data

I should emphasise that this chapter is also useful because this framework is lacking, which makes it difficult to compare methods.

This chapter should focus on the statistical framework and not the bioinformatics problems (these should go in the previous chapter) and so this chapter requires some re-working.

- [x] Chapter overview
- [ ] One sample
- [ ] _n_ samples
- [ ] Parameter estimation
- [ ] Statistical properties of $\beta$
- [ ] Downstream analyses

## Co-methylation review

__UP TO HERE__

## Co-methylation

## Simulation model

This chapter is a good candidate to write during the bootcamp.

## Using simulation model

## Analysis of Emma Whitelaw's data

## Concluding remarks

## Bibliography

## Appendices

## Statistical model of methylC-seq data
## Standard analysis of methylC-seq data
## m-tuples
## Correlation of $\beta$-values
## Simulation model: Part I
## Simulation model: Part II
## Data analysis - EPISCOPE
## Concluding Remarks
## Bibliography
## Appendices
